<!DOCTYPE html><html lang="zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0,viewport-fit=cover"><title>生信小记：使用Seurat包处理多模态数据 | Nahida的净善宫</title><meta name="author" content="Nahida"><meta name="copyright" content="Nahida"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="在一次实验中，我们可以对同一个细胞测量不同分子层面的数据（如基因表达、蛋白质表达、表观遗传修饰、空间分布等等），所测得的数据我们称之为多模态数据（multimodal data）。例如，CITE-seq可以同时检测单细胞的转录组RNA和表面蛋白表达水平；10X Multiome可以同时测量单细胞的转录组RNA和ATAC（染色质可及性）。本篇文章将以一个包括8617个脐带血单个核细胞（CBMC）的C">
<meta property="og:type" content="article">
<meta property="og:title" content="生信小记：使用Seurat包处理多模态数据">
<meta property="og:url" content="http://example.com/2025/03/31/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E4%BD%BF%E7%94%A8Seurat%E5%8C%85%E5%A4%84%E7%90%86%E5%A4%9A%E6%A8%A1%E6%80%81%E6%95%B0%E6%8D%AE/index.html">
<meta property="og:site_name" content="Nahida的净善宫">
<meta property="og:description" content="在一次实验中，我们可以对同一个细胞测量不同分子层面的数据（如基因表达、蛋白质表达、表观遗传修饰、空间分布等等），所测得的数据我们称之为多模态数据（multimodal data）。例如，CITE-seq可以同时检测单细胞的转录组RNA和表面蛋白表达水平；10X Multiome可以同时测量单细胞的转录组RNA和ATAC（染色质可及性）。本篇文章将以一个包括8617个脐带血单个核细胞（CBMC）的C">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="http://example.com/images/Seurat.png">
<meta property="article:published_time" content="2025-03-31T15:25:21.000Z">
<meta property="article:modified_time" content="2025-07-10T07:29:46.421Z">
<meta property="article:author" content="Nahida">
<meta property="article:tag" content="生信分析">
<meta property="article:tag" content="R语言">
<meta property="article:tag" content="Seurat">
<meta property="article:tag" content="多模态数据">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://example.com/images/Seurat.png"><script type="application/ld+json">{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "生信小记：使用Seurat包处理多模态数据",
  "url": "http://example.com/2025/03/31/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E4%BD%BF%E7%94%A8Seurat%E5%8C%85%E5%A4%84%E7%90%86%E5%A4%9A%E6%A8%A1%E6%80%81%E6%95%B0%E6%8D%AE/",
  "image": "http://example.com/images/Seurat.png",
  "datePublished": "2025-03-31T15:25:21.000Z",
  "dateModified": "2025-07-10T07:29:46.421Z",
  "author": [
    {
      "@type": "Person",
      "name": "Nahida",
      "url": "http://example.com/"
    }
  ]
}</script><link rel="shortcut icon" href="/img/avatar.jpg"><link rel="canonical" href="http://example.com/2025/03/31/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E4%BD%BF%E7%94%A8Seurat%E5%8C%85%E5%A4%84%E7%90%86%E5%A4%9A%E6%A8%A1%E6%80%81%E6%95%B0%E6%8D%AE/index.html"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css"><script>
    (() => {
      
    const saveToLocal = {
      set: (key, value, ttl) => {
        if (!ttl) return
        const expiry = Date.now() + ttl * 86400000
        localStorage.setItem(key, JSON.stringify({ value, expiry }))
      },
      get: key => {
        const itemStr = localStorage.getItem(key)
        if (!itemStr) return undefined
        const { value, expiry } = JSON.parse(itemStr)
        if (Date.now() > expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return value
      }
    }

    window.btf = {
      saveToLocal,
      getScript: (url, attr = {}) => new Promise((resolve, reject) => {
        const script = document.createElement('script')
        script.src = url
        script.async = true
        Object.entries(attr).forEach(([key, val]) => script.setAttribute(key, val))
        script.onload = script.onreadystatechange = () => {
          if (!script.readyState || /loaded|complete/.test(script.readyState)) resolve()
        }
        script.onerror = reject
        document.head.appendChild(script)
      }),
      getCSS: (url, id) => new Promise((resolve, reject) => {
        const link = document.createElement('link')
        link.rel = 'stylesheet'
        link.href = url
        if (id) link.id = id
        link.onload = link.onreadystatechange = () => {
          if (!link.readyState || /loaded|complete/.test(link.readyState)) resolve()
        }
        link.onerror = reject
        document.head.appendChild(link)
      }),
      addGlobalFn: (key, fn, name = false, parent = window) => {
        if (!false && key.startsWith('pjax')) return
        const globalFn = parent.globalFn || {}
        globalFn[key] = globalFn[key] || {}
        globalFn[key][name || Object.keys(globalFn[key]).length] = fn
        parent.globalFn = globalFn
      }
    }
  
      
      const activateDarkMode = () => {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      const activateLightMode = () => {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }

      btf.activateDarkMode = activateDarkMode
      btf.activateLightMode = activateLightMode

      const theme = saveToLocal.get('theme')
    
          theme === 'dark' ? activateDarkMode() : theme === 'light' ? activateLightMode() : null
        
      
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        document.documentElement.classList.toggle('hide-aside', asideStatus === 'hide')
      }
    
      
    const detectApple = () => {
      if (/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)) {
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
  
    })()
  </script><script>const GLOBAL_CONFIG = {
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  highlight: {"plugin":"highlight.js","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false,"highlightFullpage":false,"highlightMacStyle":false},
  copy: {
    success: '复制成功',
    error: '复制失败',
    noSupport: '浏览器不支持'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  dateSuffix: {
    just: '刚刚',
    min: '分钟前',
    hour: '小时前',
    day: '天前',
    month: '个月前'
  },
  copyright: undefined,
  lightbox: 'null',
  Snackbar: undefined,
  infinitegrid: {
    js: 'https://cdn.jsdelivr.net/npm/@egjs/infinitegrid/dist/infinitegrid.min.js',
    buttonText: '加载更多'
  },
  isPhotoFigcaption: false,
  islazyloadPlugin: false,
  isAnchor: false,
  percent: {
    toc: true,
    rightside: false,
  },
  autoDarkmode: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: '生信小记：使用Seurat包处理多模态数据',
  isHighlightShrink: false,
  isToc: true,
  pageType: 'post'
}</script><meta name="generator" content="Hexo 7.3.0"><link rel="alternate" href="/atom.xml" title="Nahida的净善宫" type="application/atom+xml">
</head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img text-center"><img src="/img/avatar.jpg" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="site-data text-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">95</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">55</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">22</div></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> 时间轴</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> 标签</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> 分类</span></a></div><div class="menus_item"><a class="site-page" href="/photos/"><i class="fa-fw fas fa-images"></i><span> 相册</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> 友链</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url(/images/Seurat.png);"><nav id="nav"><span id="blog-info"><a class="nav-site-title" href="/"><img class="site-icon" src="/image/test.gif" alt="Logo"><span class="site-name">Nahida的净善宫</span></a><a class="nav-page-title" href="/"><span class="site-name">生信小记：使用Seurat包处理多模态数据</span></a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> 时间轴</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> 标签</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> 分类</span></a></div><div class="menus_item"><a class="site-page" href="/photos/"><i class="fa-fw fas fa-images"></i><span> 相册</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> 友链</span></a></div></div><div id="toggle-menu"><span class="site-page"><i class="fas fa-bars fa-fw"></i></span></div></div></nav><div id="post-info"><h1 class="post-title">生信小记：使用Seurat包处理多模态数据</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">发表于</span><time class="post-meta-date-created" datetime="2025-03-31T15:25:21.000Z" title="发表于 2025-03-31 23:25:21">2025-03-31</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">更新于</span><time class="post-meta-date-updated" datetime="2025-07-10T07:29:46.421Z" title="更新于 2025-07-10 15:29:46">2025-07-10</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/">生信分析</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title=""><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">浏览量:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="container post-content" id="article-container"><p>在一次实验中，我们可以对同一个细胞测量不同分子层面的数据（如基因表达、蛋白质表达、表观遗传修饰、空间分布等等），所测得的数据我们称之为多模态数据（multimodal data）。例如，CITE-seq可以同时检测单细胞的转录组RNA和表面蛋白表达水平；10X Multiome可以同时测量单细胞的转录组RNA和ATAC（染色质可及性）。本篇文章将以一个包括8617个脐带血单个核细胞（CBMC）的CITE-seq结果（包含转录组RNA测序和表面蛋白丰度估计值）的数据集，介绍如何使用Seurat包对多模态数据进行分析。</p>
<h1 id="导入数据，创建Seurat对象，并加入RNA和表面蛋白数据"><a href="#导入数据，创建Seurat对象，并加入RNA和表面蛋白数据" class="headerlink" title="导入数据，创建Seurat对象，并加入RNA和表面蛋白数据"></a>导入数据，创建Seurat对象，并加入RNA和表面蛋白数据</h1><p><a href="">官方教程</a>中介绍的方法如下：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line">library(Seurat)</span><br><span class="line">library(ggplot2)</span><br><span class="line">library(patchwork)</span><br><span class="line"></span><br><span class="line"># 注意：这份数据同时存在人和小鼠的细胞，分别冠有&quot;HUMAN_&quot;和&quot;MOUSE_&quot;的前缀。</span><br><span class="line"></span><br><span class="line"># 导入RNA数据</span><br><span class="line">cbmc.rna &lt;- as.sparse(read.csv(file = &quot;/brahms/shared/vignette-data/GSE100866_CBMC_8K_13AB_10X-RNA_umi.csv.gz&quot;,</span><br><span class="line">    sep = &quot;,&quot;, header = TRUE, row.names = 1))</span><br><span class="line"></span><br><span class="line"># 在本例中，我们将删除除了100个表达水平最高的小鼠基因之外的所有小鼠基因，并且移除人源基因前面的HUMAN_标记。</span><br><span class="line">cbmc.rna &lt;- CollapseSpeciesExpressionMatrix(cbmc.rna)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"># 载入表面蛋白丰度数据</span><br><span class="line">cbmc.adt &lt;- as.sparse(read.csv(file = &quot;/brahms/shared/vignette-data/GSE100866_CBMC_8K_13AB_10X-ADT_umi.csv.gz&quot;,</span><br><span class="line">    sep = &quot;,&quot;, header = TRUE, row.names = 1))</span><br><span class="line"></span><br><span class="line"># 在多模态数据中，不同类型的数据往往相互对应；更直白的讲，一个细胞中的RNA和表面蛋白丰度应该相互匹配，不能出现一个细胞只有RNA数据，而表面蛋白丰度缺失；或反之的情况。</span><br><span class="line"># all.equal()函数即可以对数据进行检查，具体来说，通过对比两列细胞的barcodes是否一致进行检查</span><br><span class="line"># 若返回true，说明数据没问题，如果返回false，则说明数据可能出错了</span><br><span class="line">all.equal(colnames(cbmc.rna), colnames(cbmc.adt))</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">cbmc &lt;- CreateSeuratObject(counts = cbmc.rna) #基于转录组RNA数据，创建Seurat对象</span><br><span class="line"># create a new assay to store ADT information</span><br><span class="line">adt_assay &lt;- CreateAssay5Object(counts = cbmc.adt) #新建一个assay，用于存储表面蛋白丰度数据</span><br><span class="line">cbmc[[&quot;ADT&quot;]] &lt;- adt_assay #将新建的assay加入到Seurat对象中</span><br></pre></td></tr></table></figure>

<p>然而，由于官方所给的下载链接为<code>ftp</code>协议（需要进行一系列复杂的设置才能正常下载……然鹅我到现在也没有下载成，伟大如Deepseek也在ftp面前落败），本文章改用下述方法直接下载并加载Seurat数据集。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">library(Seurat)</span><br><span class="line">library(patchwork)</span><br><span class="line">library(ggplot2)</span><br><span class="line">library(SeuratData)</span><br><span class="line">InstallData(&quot;cbmc&quot;)</span><br><span class="line">data(&quot;cbmc&quot;)</span><br><span class="line">Assays(cbmc) #查看cbmc中存在多少个assay</span><br><span class="line">cbmc = UpdateSeuratObject(object = cbmc) #更新Seurat对象结构，否则后续会报错（好像是什么没有images之类的）</span><br><span class="line">all.equal(colnames(cbmc[[&quot;RNA&quot;]]), colnames(cbmc[[&quot;ADT&quot;]])) #检查数据，具体如前述。</span><br></pre></td></tr></table></figure>
<ul>
<li>运行结果：</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">Assays(cbmc)</span><br><span class="line">[1] &quot;RNA&quot; &quot;ADT&quot;</span><br><span class="line">&gt; cbmc = UpdateSeuratObject(object = cbmc)</span><br><span class="line">Validating object structure</span><br><span class="line">Updating object slots</span><br><span class="line">Ensuring keys are in the proper structure</span><br><span class="line">Warning: Assay RNA changing from Assay to Assay</span><br><span class="line">Warning: Assay ADT changing from Assay to Assay</span><br><span class="line">Ensuring keys are in the proper structure</span><br><span class="line">Ensuring feature names don&#x27;t have underscores or pipes</span><br><span class="line">Updating slots in RNA</span><br><span class="line">Updating slots in ADT</span><br><span class="line">Validating object structure for Assay ‘RNA’</span><br><span class="line">Validating object structure for Assay ‘ADT’</span><br><span class="line">Object representation is consistent with the most current Seurat version</span><br><span class="line">&gt; all.equal(colnames(cbmc[[&quot;RNA&quot;]]), colnames(cbmc[[&quot;ADT&quot;]]))</span><br><span class="line">[1] TRUE</span><br></pre></td></tr></table></figure>

<p>对于导入好的数据，我们可以：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"># 用`rownames()`函数查看我们所测定的特征（基因）列表：</span><br><span class="line">rownames(cbmc[[&quot;ADT&quot;]])</span><br><span class="line"></span><br><span class="line"># 切换Assay的默认值，默认使用RNA或ADT数据进行分析：</span><br><span class="line">DefaultAssay(cbmc)  # 查看默认Assay</span><br><span class="line">DefaultAssay(cbmc) &lt;- &quot;ADT&quot; #修改默认Assay为ADT</span><br><span class="line">DefaultAssay(cbmc)  # 再次查看默认Assay</span><br><span class="line"></span><br><span class="line"># 当然，我们可以通过直接在函数中手动调整Assay参数来实现对不同Assay的分析。</span><br></pre></td></tr></table></figure>

<ul>
<li>运行结果：</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">&gt; Assays(cbmc)</span><br><span class="line">[1] &quot;RNA&quot; &quot;ADT&quot;</span><br><span class="line">&gt; DefaultAssay(cbmc)</span><br><span class="line">[1] &quot;RNA&quot;</span><br><span class="line">&gt; DefaultAssay(cbmc) &lt;- &quot;ADT&quot;</span><br><span class="line">&gt; DefaultAssay(cbmc)</span><br><span class="line">[1] &quot;ADT&quot;</span><br></pre></td></tr></table></figure>
<h1 id="根据转录组测序结果进行细胞聚类"><a href="#根据转录组测序结果进行细胞聚类" class="headerlink" title="根据转录组测序结果进行细胞聚类"></a>根据转录组测序结果进行细胞聚类</h1><p>这一步和<a target="_blank" rel="noopener" href="https://nahida23333.github.io/2025/03/28/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9ASeurat%E7%9A%84%E5%B8%B8%E8%A7%81%E5%88%86%E6%9E%90%E5%B7%A5%E4%BD%9C%E6%B5%81%E7%A8%8B%EF%BC%9A%E4%BB%A5PBMC-3K%E6%95%B0%E6%8D%AE%E9%9B%86%E4%B8%BA%E4%BE%8B/">《Seurat的常见分析工作流程：以PBMC_3K数据集为例》</a>这篇文章中所提到的其实是一样的，但要记得先把上一步被我们切走的默认Assay值切回来（欸嘿！）。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">DefaultAssay(cbmc) &lt;- &quot;RNA&quot;</span><br><span class="line">cbmc &lt;- NormalizeData(cbmc)</span><br><span class="line">cbmc &lt;- FindVariableFeatures(cbmc)</span><br><span class="line">cbmc &lt;- ScaleData(cbmc)</span><br><span class="line">cbmc &lt;- RunPCA(cbmc, verbose = FALSE)</span><br><span class="line">cbmc &lt;- FindNeighbors(cbmc, dims = 1:30)</span><br><span class="line">cbmc &lt;- FindClusters(cbmc, resolution = 0.8, verbose = FALSE)</span><br><span class="line">cbmc &lt;- RunUMAP(cbmc, dims = 1:30)</span><br><span class="line">DimPlot(cbmc, label = TRUE)</span><br></pre></td></tr></table></figure>

<p><img src="/images/multimodal_practice/eb9995d9-c976-4ac9-8a18-01ac033e27fd.png"></p>
<h1 id="多模态数据的并排可视化"><a href="#多模态数据的并排可视化" class="headerlink" title="多模态数据的并排可视化"></a>多模态数据的并排可视化</h1><p>如果想研究某一个基因在不同细胞中的转录水平和蛋白质表达水平的对应关系（或者相互验证某一基因的表达情况），我们可以通过将多模态数据并排可视化。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">cbmc &lt;- NormalizeData(cbmc, normalization.method = &quot;CLR&quot;, margin = 2, assay = &quot;ADT&quot;)</span><br><span class="line">DefaultAssay(cbmc) &lt;- &quot;ADT&quot;</span><br><span class="line">p1 &lt;- FeaturePlot(cbmc, &quot;CD19&quot;, cols = c(&quot;lightgrey&quot;, &quot;darkgreen&quot;)) + ggtitle(&quot;CD19 protein&quot;)</span><br><span class="line">DefaultAssay(cbmc) &lt;- &quot;RNA&quot;</span><br><span class="line">p2 &lt;- FeaturePlot(cbmc, &quot;CD19&quot;) + ggtitle(&quot;CD19 RNA&quot;)</span><br><span class="line">p1 | p2 #横排显示p1和p2</span><br></pre></td></tr></table></figure>
<p><img src="/images/multimodal_practice/c5f5b52e-77ce-464a-ac48-a617840ac9b4.png"></p>
<ul>
<li>标记了一处问题：为什么ADT数据使用CLR进行标准化？（查了很多资料没有搞懂，先放放）</li>
</ul>
<p>在不同的Assay中，特征名可能重名。比如，在RNA和ADT这两个Assay中都有一列CD19的数据，那么可能会冲突而影响分析。为了避免这种情况，每一个Assay均有其唯一标识符，称为“Key”。默认情况下，RNA的Key为“<code>rna_</code>”，ADT的Key为“<code>adt_</code>”。我们也可以通过<code>key()</code>查询不同Assay的唯一标识符，在后续过程中通过明确key参数来指定对哪一个Assay的数据进行处理，而不用频繁地切换默认Assay。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">Key(cbmc[[&quot;RNA&quot;]])</span><br><span class="line">Key(cbmc[[&quot;ADT&quot;]])</span><br></pre></td></tr></table></figure>
<ul>
<li>运行结果：</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">&gt; Key(cbmc[[&quot;RNA&quot;]])</span><br><span class="line">[1] &quot;rna_&quot;</span><br><span class="line">&gt; Key(cbmc[[&quot;ADT&quot;]])</span><br><span class="line">[1] &quot;adt_&quot;</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">p1 &lt;- FeaturePlot(cbmc, &quot;adt_CD19&quot;, cols = c(&quot;lightgrey&quot;, &quot;darkgreen&quot;)) + ggtitle(&quot;CD19 protein&quot;)</span><br><span class="line">p2 &lt;- FeaturePlot(cbmc, &quot;rna_CD19&quot;) + ggtitle(&quot;CD19 RNA&quot;)</span><br><span class="line">p1 | p2</span><br></pre></td></tr></table></figure>
<p>输出的图象同上，这里不再重复。</p>
<h1 id="鉴定不同细胞集群的表面蛋白标记"><a href="#鉴定不同细胞集群的表面蛋白标记" class="headerlink" title="鉴定不同细胞集群的表面蛋白标记"></a>鉴定不同细胞集群的表面蛋白标记</h1><p>我们知道，CD19是B细胞的标志，通过绘制小提琴图我们可以识别B细胞集群（细胞集群6）。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">VlnPlot(cbmc, &quot;adt_CD19&quot;)</span><br></pre></td></tr></table></figure>
<p><img src="/images/multimodal_practice/da042531-8d37-479b-b423-a024f59e2146.png"></p>
<p>通过<code>FindMarkers()</code>函数，我们可以找出该细胞簇中差异表达的RNA和蛋白质。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">adt_markers &lt;- FindMarkers(cbmc, ident.1 = 6, assay = &quot;ADT&quot;)</span><br><span class="line">rna_markers &lt;- FindMarkers(cbmc, ident.1 = 6, assay = &quot;RNA&quot;)</span><br><span class="line">head(adt_markers)</span><br><span class="line">head(rna_markers)</span><br></pre></td></tr></table></figure>
<ul>
<li>运行结果：</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">&gt; head(adt_markers)</span><br><span class="line">               p_val avg_log2FC pct.1 pct.2     p_val_adj</span><br><span class="line">CD19   1.764302e-215  2.7444758     1     1 1.764302e-214</span><br><span class="line">CD45RA 1.581574e-121  0.5840419     1     1 1.581574e-120</span><br><span class="line">CD14   1.216541e-102 -1.0143010     1     1 1.216541e-101</span><br><span class="line">CD4    1.116804e-100 -1.6788367     1     1  1.116804e-99</span><br><span class="line">CD3     6.247313e-78 -1.5884973     1     1  6.247313e-77</span><br><span class="line">CD56    3.131803e-30  0.2355264     1     1  3.131803e-29</span><br><span class="line">&gt; head(rna_markers)</span><br><span class="line">      p_val avg_log2FC pct.1 pct.2 p_val_adj</span><br><span class="line">IGHM      0   6.660187 0.977 0.044         0</span><br><span class="line">CD79A     0   6.748356 0.965 0.045         0</span><br><span class="line">TCL1A     0   7.428099 0.904 0.028         0</span><br><span class="line">CD79B     0   5.525568 0.944 0.089         0</span><br><span class="line">IGHD      0   7.811884 0.857 0.015         0</span><br><span class="line">MS4A1     0   7.523215 0.851 0.016         0</span><br></pre></td></tr></table></figure>

<h1 id="多模态数据的其他可视化方法"><a href="#多模态数据的其他可视化方法" class="headerlink" title="多模态数据的其他可视化方法"></a>多模态数据的其他可视化方法</h1><p>我们可以通过绘制散点图来可视化不同表面蛋白之间、RNA和蛋白质之间的表达关系。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">FeatureScatter(cbmc, feature1 = &quot;adt_CD19&quot;, feature2 = &quot;adt_CD3&quot;)</span><br></pre></td></tr></table></figure>
<p><img src="/images/multimodal_practice/30ba3d2a-c356-4bf1-b610-1d42a085814c.png"></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">FeatureScatter(cbmc, feature1 = &quot;adt_CD3&quot;, feature2 = &quot;rna_CD3E&quot;)</span><br></pre></td></tr></table></figure>
<p><img src="/images/multimodal_practice/d14ceda2-b5ba-474e-a00a-1e0123fd2f1b.png"></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">FeatureScatter(cbmc, feature1 = &quot;adt_CD4&quot;, feature2 = &quot;adt_CD8&quot;)</span><br></pre></td></tr></table></figure>
<p><img src="/images/multimodal_practice/a6588e64-cba4-4bc4-8112-235af7e409a9.png"></p>
<p>由于蛋白质的数量和RNA相比超级无敌巨多，所以如果没有标准化……横纵轴的数量级会变得很恐怖：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">FeatureScatter(cbmc, feature1 = &quot;adt_CD4&quot;, feature2 = &quot;adt_CD8&quot;, slot = &quot;counts&quot;)</span><br></pre></td></tr></table></figure>
<p><img src="/images/multimodal_practice/cc6db391-f22b-4f66-9c7a-e840f3354ba6.png"></p>
<h1 id="加载通过10X-Multiome获得的多模态数据"><a href="#加载通过10X-Multiome获得的多模态数据" class="headerlink" title="加载通过10X Multiome获得的多模态数据"></a>加载通过10X Multiome获得的多模态数据</h1><p>这里我们利用<a target="_blank" rel="noopener" href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/pbmc_10k_protein_v3">10X Multiome提供的PBMC数据集</a>进行演示。需要下载的文件为<code>Feature / cell matrix (filtered)</code>。我们下载的多模态数据中包含以下两种数据：</p>
<ul>
<li>Gene Expression：转录组RNA测序数据。</li>
<li>Antibody Capture：ADT数据。</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line"># 加载数据</span><br><span class="line">pbmc10k.data &lt;- Read10X(data.dir = &quot;C:/Users/86150/Downloads/pbmc_10k_protein_v3_filtered_feature_bc_matrix/filtered_feature_bc_matrix/&quot;)</span><br><span class="line"># 处理抗体名称：利用gsub()函数去除ADT数据中抗体名称的后缀，如将`CD19_control_TotalSeqB`调整为`CD19`，将`CD3_TotalSeqB`调整为`CD3`。</span><br><span class="line">rownames(x = pbmc10k.data[[&quot;Antibody Capture&quot;]]) &lt;- gsub(pattern = &quot;_[control_]*TotalSeqB&quot;, replacement = &quot;&quot;, x = rownames(x = pbmc10k.data[[&quot;Antibody Capture&quot;]]))</span><br><span class="line"></span><br><span class="line"># 以转录组RNA数据为基础，创建Seurat对象。</span><br><span class="line">pbmc10k &lt;- CreateSeuratObject(counts = pbmc10k.data[[&quot;Gene Expression&quot;]], min.cells = 3, min.features = 200)</span><br><span class="line"># 默认使用LogNormalize对RNA数据进行标准化。</span><br><span class="line">pbmc10k &lt;- NormalizeData(pbmc10k)</span><br><span class="line"># 向Seurat对象中添加ADT数据。</span><br><span class="line">pbmc10k[[&quot;ADT&quot;]] &lt;- CreateAssayObject(pbmc10k.data[[&quot;Antibody Capture&quot;]][, colnames(x = pbmc10k)])</span><br><span class="line"># 用CLR对ADT数据进行标准化。</span><br><span class="line">pbmc10k &lt;- NormalizeData(pbmc10k, assay = &quot;ADT&quot;, normalization.method = &quot;CLR&quot;)</span><br><span class="line"></span><br><span class="line"># 可视化</span><br><span class="line">plot1 &lt;- FeatureScatter(pbmc10k, feature1 = &quot;adt_CD19&quot;, feature2 = &quot;adt_CD3&quot;, pt.size = 1)</span><br><span class="line">plot2 &lt;- FeatureScatter(pbmc10k, feature1 = &quot;adt_CD4&quot;, feature2 = &quot;adt_CD8a&quot;, pt.size = 1)</span><br><span class="line">plot3 &lt;- FeatureScatter(pbmc10k, feature1 = &quot;adt_CD3&quot;, feature2 = &quot;CD3E&quot;, pt.size = 1)</span><br><span class="line">(plot1 + plot2 + plot3) &amp; NoLegend()</span><br></pre></td></tr></table></figure>
<p><img src="/images/multimodal_practice/50ec13cf-7e73-49d0-92ef-3c65e86c5dd6.png"></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">plot &lt;- FeatureScatter(cbmc, feature1 = &quot;adt_CD19&quot;, feature2 = &quot;adt_CD3&quot;) + NoLegend() + theme(axis.title = </span><br><span class="line">    element_text(size = 18), # 设置坐标轴标题字体大小</span><br><span class="line">    legend.text = element_text(size = 18) # 设置图例文本字体大小（但已被NoLegend()移除）</span><br><span class="line">)</span><br><span class="line">plot</span><br></pre></td></tr></table></figure>
<p><img src="/images/multimodal_practice/e9f9c9e0-a9ca-402d-8a0c-0acf0ba9a083.png"></p>
<p><a href="">官方教程</a>中还利用<code>ggsave()</code>进行了保存。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ggsave(filename = &quot;../output/images/citeseq_plot.jpg&quot;, height = 7, width = 12, plot = plot, quality = 50)</span><br></pre></td></tr></table></figure>

<h1 id="完整代码"><a href="#完整代码" class="headerlink" title="完整代码"></a>完整代码</h1><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br></pre></td><td class="code"><pre><span class="line">library(Seurat)</span><br><span class="line">library(patchwork)</span><br><span class="line">library(ggplot2)</span><br><span class="line">library(SeuratData)</span><br><span class="line">InstallData(&quot;cbmc&quot;)</span><br><span class="line">data(&quot;cbmc&quot;)</span><br><span class="line"></span><br><span class="line">Assays(cbmc)</span><br><span class="line">cbmc = UpdateSeuratObject(object = cbmc)</span><br><span class="line">all.equal(colnames(cbmc[[&quot;RNA&quot;]]), colnames(cbmc[[&quot;ADT&quot;]]))</span><br><span class="line"></span><br><span class="line">rownames(cbmc[[&quot;ADT&quot;]])</span><br><span class="line"></span><br><span class="line">DefaultAssay(cbmc)</span><br><span class="line">DefaultAssay(cbmc) &lt;- &quot;ADT&quot;</span><br><span class="line">DefaultAssay(cbmc)</span><br><span class="line"></span><br><span class="line">DefaultAssay(cbmc) &lt;- &quot;RNA&quot;</span><br><span class="line">cbmc &lt;- NormalizeData(cbmc)</span><br><span class="line">cbmc &lt;- FindVariableFeatures(cbmc)</span><br><span class="line">cbmc &lt;- ScaleData(cbmc)</span><br><span class="line">cbmc &lt;- RunPCA(cbmc, verbose = FALSE)</span><br><span class="line">cbmc &lt;- FindNeighbors(cbmc, dims = 1:30)</span><br><span class="line">cbmc &lt;- FindClusters(cbmc, resolution = 0.8, verbose = FALSE)</span><br><span class="line">cbmc &lt;- RunUMAP(cbmc, dims = 1:30)</span><br><span class="line">DimPlot(cbmc, label = TRUE)</span><br><span class="line"></span><br><span class="line">cbmc &lt;- NormalizeData(cbmc, normalization.method = &quot;CLR&quot;, margin = 2, assay = &quot;ADT&quot;)</span><br><span class="line">DefaultAssay(cbmc) &lt;- &quot;ADT&quot;</span><br><span class="line">p1 &lt;- FeaturePlot(cbmc, &quot;CD19&quot;, cols = c(&quot;lightgrey&quot;, &quot;darkgreen&quot;)) + ggtitle(&quot;CD19 protein&quot;)</span><br><span class="line">DefaultAssay(cbmc) &lt;- &quot;RNA&quot;</span><br><span class="line">p2 &lt;- FeaturePlot(cbmc, &quot;CD19&quot;) + ggtitle(&quot;CD19 RNA&quot;)</span><br><span class="line">p1 | p2</span><br><span class="line"></span><br><span class="line">Key(cbmc[[&quot;RNA&quot;]])</span><br><span class="line">Key(cbmc[[&quot;ADT&quot;]])</span><br><span class="line">p1 &lt;- FeaturePlot(cbmc, &quot;adt_CD19&quot;, cols = c(&quot;lightgrey&quot;, &quot;darkgreen&quot;)) + ggtitle(&quot;CD19 protein&quot;)</span><br><span class="line">p2 &lt;- FeaturePlot(cbmc, &quot;rna_CD19&quot;) + ggtitle(&quot;CD19 RNA&quot;)</span><br><span class="line">p1 | p2</span><br><span class="line"></span><br><span class="line">VlnPlot(cbmc, &quot;adt_CD19&quot;)</span><br><span class="line"></span><br><span class="line">adt_markers &lt;- FindMarkers(cbmc, ident.1 = 6, assay = &quot;ADT&quot;)</span><br><span class="line">rna_markers &lt;- FindMarkers(cbmc, ident.1 = 6, assay = &quot;RNA&quot;)</span><br><span class="line">head(adt_markers)</span><br><span class="line">head(rna_markers)</span><br><span class="line"></span><br><span class="line">FeatureScatter(cbmc, feature1 = &quot;adt_CD19&quot;, feature2 = &quot;adt_CD3&quot;)</span><br><span class="line">FeatureScatter(cbmc, feature1 = &quot;adt_CD3&quot;, feature2 = &quot;rna_CD3E&quot;)</span><br><span class="line">FeatureScatter(cbmc, feature1 = &quot;adt_CD4&quot;, feature2 = &quot;adt_CD8&quot;)</span><br><span class="line">FeatureScatter(cbmc, feature1 = &quot;adt_CD4&quot;, feature2 = &quot;adt_CD8&quot;, slot = &quot;counts&quot;)</span><br><span class="line"></span><br><span class="line">pbmc10k.data &lt;- Read10X(data.dir = &quot;C:/Users/86150/Downloads/pbmc_10k_protein_v3_filtered_feature_bc_matrix/filtered_feature_bc_matrix/&quot;)</span><br><span class="line">rownames(x = pbmc10k.data[[&quot;Antibody Capture&quot;]]) &lt;- gsub(pattern = &quot;_[control_]*TotalSeqB&quot;, replacement = &quot;&quot;, x = rownames(x = pbmc10k.data[[&quot;Antibody Capture&quot;]]))</span><br><span class="line"></span><br><span class="line">pbmc10k &lt;- CreateSeuratObject(counts = pbmc10k.data[[&quot;Gene Expression&quot;]], min.cells = 3, min.features = 200)</span><br><span class="line">pbmc10k &lt;- NormalizeData(pbmc10k)</span><br><span class="line">pbmc10k[[&quot;ADT&quot;]] &lt;- CreateAssayObject(pbmc10k.data[[&quot;Antibody Capture&quot;]][, colnames(x = pbmc10k)])</span><br><span class="line">pbmc10k &lt;- NormalizeData(pbmc10k, assay = &quot;ADT&quot;, normalization.method = &quot;CLR&quot;)</span><br><span class="line"></span><br><span class="line">plot1 &lt;- FeatureScatter(pbmc10k, feature1 = &quot;adt_CD19&quot;, feature2 = &quot;adt_CD3&quot;, pt.size = 1)</span><br><span class="line">plot2 &lt;- FeatureScatter(pbmc10k, feature1 = &quot;adt_CD4&quot;, feature2 = &quot;adt_CD8a&quot;, pt.size = 1)</span><br><span class="line">plot3 &lt;- FeatureScatter(pbmc10k, feature1 = &quot;adt_CD3&quot;, feature2 = &quot;CD3E&quot;, pt.size = 1)</span><br><span class="line">(plot1 + plot2 + plot3) &amp; NoLegend()</span><br><span class="line"></span><br><span class="line">plot &lt;- FeatureScatter(cbmc, feature1 = &quot;adt_CD19&quot;, feature2 = &quot;adt_CD3&quot;) + NoLegend() + theme(axis.title = element_text(size = 18), legend.text = element_text(size = 18))</span><br></pre></td></tr></table></figure></article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta"><i class="fas fa-circle-user fa-fw"></i>文章作者: </span><span class="post-copyright-info"><a href="http://example.com">Nahida</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta"><i class="fas fa-square-arrow-up-right fa-fw"></i>文章链接: </span><span class="post-copyright-info"><a href="http://example.com/2025/03/31/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E4%BD%BF%E7%94%A8Seurat%E5%8C%85%E5%A4%84%E7%90%86%E5%A4%9A%E6%A8%A1%E6%80%81%E6%95%B0%E6%8D%AE/">http://example.com/2025/03/31/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E4%BD%BF%E7%94%A8Seurat%E5%8C%85%E5%A4%84%E7%90%86%E5%A4%9A%E6%A8%A1%E6%80%81%E6%95%B0%E6%8D%AE/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta"><i class="fas fa-circle-exclamation fa-fw"></i>版权声明: </span><span class="post-copyright-info">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0</a> 许可协议。转载请注明来源 <a href="http://example.com" target="_blank">Nahida的净善宫</a>！</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/">生信分析</a><a class="post-meta__tags" href="/tags/R%E8%AF%AD%E8%A8%80/">R语言</a><a class="post-meta__tags" href="/tags/Seurat/">Seurat</a><a class="post-meta__tags" href="/tags/%E5%A4%9A%E6%A8%A1%E6%80%81%E6%95%B0%E6%8D%AE/">多模态数据</a></div><div class="post-share"><div class="social-share" data-image="/images/Seurat.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><a class="pagination-related" href="/2025/03/31/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E4%BD%BF%E7%94%A8SCTransform%E8%BF%9B%E8%A1%8C%E6%A0%87%E5%87%86%E5%8C%96/" title="生信小记：使用SCTransform进行标准化"><img class="cover" src="/images/Seurat.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="info"><div class="info-1"><div class="info-item-1">上一篇</div><div class="info-item-2">生信小记：使用SCTransform进行标准化</div></div><div class="info-2"><div class="info-item-1">...</div></div></div></a><a class="pagination-related" href="/2025/04/03/%E3%80%90%E5%8C%BB%E5%AD%A6%E5%BD%B1%E5%83%8F%E5%AD%A6%E3%80%91%E6%80%BB%E8%AE%BA/" title="【医学影像学】总论"><div class="cover" style="background: var(--default-bg-color)"></div><div class="info text-right"><div class="info-1"><div class="info-item-1">下一篇</div><div class="info-item-2">【医学影像学】总论</div></div><div class="info-2"><div class="info-item-1">医学影像学基本概念利用X线、CT、MRI、超声、核素、DSA等等多种影像技术所形成的影像对临床疾病进行诊断或影像引导治疗的学科。 学科划分 放射影像学：X线、CT、MRI、介入放射学 核医学：核素显像、PET-CT（PET-MRI）、SPECT、放射性核素治疗 超声医学：超声诊断、超声引导治疗、超声射频治疗  X线X线的产生在 X 线管内，高速行进的电子流被靶物质的原子阻挡而发生能量转换，其中约1%的能量转变为 X 线。 X线的本质波长很短的电磁波，以光速沿直线前进，波长为0.01nm10nm。X线诊断用0.031nm0.008nm（40~150kV）。 X线的特性： 穿透性：X线能穿过可见光不能穿过的物体。管电压越高，X线波长越短，穿透力越强。X线穿过时会被衰减。穿透能力也与所穿过的物质的厚度、密度相关，这是X线的成像基础。 荧光效应（透视基础）：X线能激发荧光物质（如钨酸钙）发出可见光。 感光效应：X线使溴化银感光形成潜影：溴化银（Ag+）→...</div></div></div></a></nav><div class="relatedPosts"><div class="headline"><i class="fas fa-thumbs-up fa-fw"></i><span>相关推荐</span></div><div class="relatedPosts-list"><a class="pagination-related" href="/2025/09/20/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E6%98%A0%E5%B0%84%E5%92%8C%E6%B3%A8%E9%87%8A%E6%9F%A5%E8%AF%A2%E6%95%B0%E6%8D%AE%E9%9B%86/" title="生信小记：映射和注释查询数据集"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-09-20</div><div class="info-item-2">生信小记：映射和注释查询数据集</div></div><div class="info-2"><div class="info-item-1">在前面的教程中，我们比较系统地学习了处理单细胞测序数据的分析流程。分析流程严谨而精细，可以帮助我们获得很多有价值的信息。然而，对于长期专注于某一领域的研究者来说，每拿到一批数据都要走一遍这样的分析流程，其计算量大、耗费的时间精力多、效率低下，不太能满足研究者的效率需求。为了提高效率，研究者们开辟出了一条更为方便的捷径，即使用多个可靠的数据集构建一个整合的参考框架，在拿到新的数据之后不再需要从头开始分析，只需要通过算法将数据映射到参考数据集上，即可自动获得较高质量的注释。这个过程很像人工智能大模型的开发，即使用高质量的标注标注数据​“训练”​​一个模型，学习其中的模式、特征和关联，让其获得充分的知识储备和逻辑分析能力，在遇到新问题时可以利用内部知识进行推理并生成答案。 数据集的预处理我们先引入示例数据集，这份数据集包括由四种不同测序技术得到的人胰岛细胞数据，即CelSeq（GSE81076）、CelSeq2（GSE85241）、Fluidigm...</div></div></div></a><a class="pagination-related" href="/2025/08/20/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9Ascvi-tools%E7%9A%84%E5%AE%89%E8%A3%85/" title="生信小记：scvi_tools的安装"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-08-20</div><div class="info-item-2">生信小记：scvi_tools的安装</div></div><div class="info-2"><div class="info-item-1">使用scvi算法进行单细胞测序数据的整合，一方面需要reticulate包，另一方面则需要安装在conda环境中的scvi-tools包及其依赖项。前者可以通过install.packages()语句从CRAN中下载，而本文章将介绍如何从零开始安装scvi-tools。 参考教程如下：  Anaconda安装教程（2025年保姆级超详解）【附安装包+环境玩转指南】   安装Anaconda打开Anaconda官网，点击右上角的“Free Download”，注册并登录后即可下载。 下载完成后打开安装包，可见如下界面。点击Next &gt;。 点击I Agree。 点击All users (requires admin privileges)。 选择存储位置时，强烈不建议存储在C盘，因为它安装完占用的空间不小。更改完存储路径后点击Next &gt;。 全勾选上再点Next...</div></div></div></a><a class="pagination-related" href="/2025/07/09/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9ASeurat-v5%E4%B8%AD%E7%9A%84%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90/" title="生信小记：Seurat_v5中的整合分析"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-07-09</div><div class="info-item-2">生信小记：Seurat_v5中的整合分析</div></div><div class="info-2"><div class="info-item-1">前言单细胞测序中，实验和数据的收集往往不是一次性完成的，而是分为多个批次进行的。在不同批次之间，操作者、操作流程等诸多因素存在差异，可能给实验引入一些与真实生物学差异无关的系统性偏差，称为 批次效应 。 单细胞测序数据的整合通常是scRNA-seq测序的重要步骤之一。通过匹配不同数据集之间的共享细胞类型和状态，我们可以消除批次效应所引入的偏差，从而提高了统计功效。我们常使用的整合数据的方法有多种，包括Seurat早期版本发布的基于锚点的整合流程，其他团队发布的整合算法（如Harmony、scVI等）。Seurat...</div></div></div></a><a class="pagination-related" href="/2025/04/29/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%84%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90%E4%BB%8B%E7%BB%8D/" title="生信小记：单细胞转录组数据整合分析介绍"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-04-29</div><div class="info-item-2">生信小记：单细胞转录组数据整合分析介绍</div></div><div class="info-2"><div class="info-item-1">...</div></div></div></a><a class="pagination-related" href="/2025/03/31/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E4%BD%BF%E7%94%A8SCTransform%E8%BF%9B%E8%A1%8C%E6%A0%87%E5%87%86%E5%8C%96/" title="生信小记：使用SCTransform进行标准化"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-03-31</div><div class="info-item-2">生信小记：使用SCTransform进行标准化</div></div><div class="info-2"><div class="info-item-1">...</div></div></div></a><a class="pagination-related" href="/2025/03/29/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9ASeurat%E5%8C%85%E7%9A%84%E6%95%B0%E6%8D%AE%E5%8F%AF%E8%A7%86%E5%8C%96%E6%96%B9%E6%B3%95/" title="生信小记：Seurat包的数据可视化方法"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-03-29</div><div class="info-item-2">生信小记：Seurat包的数据可视化方法</div></div><div class="info-2"><div class="info-item-1">在这篇教程中，我们会沿用上一篇教程处理好的数据来进行演示。 先做一些前期准备，安装并加载好R包，下载并加载好数据（或者直接运用上一篇教程中处理好的数据）。 12345678910111213141516171819202122install.packages(c(&quot;Seurat&quot;,&quot;patchwork&quot;,&quot;ggplot2&quot;,&quot;SeuratData&quot;))#下载SeuratData包时报错显示：package ‘SeuratData’ is not available for this version of...</div></div></div></a></div></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info text-center"><div class="avatar-img"><img src="/img/avatar.jpg" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info-name">Nahida</div><div class="author-info-description">智慧树上智慧果，智慧树下你和我</div><div class="site-data"><a href="/archives/"><div class="headline">文章</div><div class="length-num">95</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">55</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">22</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/xxxxxx"><i class="fab fa-github"></i><span>Follow Me</span></a></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>公告</span></div><div class="announcement_content">暂无公告哦~</div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>目录</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%AF%BC%E5%85%A5%E6%95%B0%E6%8D%AE%EF%BC%8C%E5%88%9B%E5%BB%BASeurat%E5%AF%B9%E8%B1%A1%EF%BC%8C%E5%B9%B6%E5%8A%A0%E5%85%A5RNA%E5%92%8C%E8%A1%A8%E9%9D%A2%E8%9B%8B%E7%99%BD%E6%95%B0%E6%8D%AE"><span class="toc-number">1.</span> <span class="toc-text">导入数据，创建Seurat对象，并加入RNA和表面蛋白数据</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%A0%B9%E6%8D%AE%E8%BD%AC%E5%BD%95%E7%BB%84%E6%B5%8B%E5%BA%8F%E7%BB%93%E6%9E%9C%E8%BF%9B%E8%A1%8C%E7%BB%86%E8%83%9E%E8%81%9A%E7%B1%BB"><span class="toc-number">2.</span> <span class="toc-text">根据转录组测序结果进行细胞聚类</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%A4%9A%E6%A8%A1%E6%80%81%E6%95%B0%E6%8D%AE%E7%9A%84%E5%B9%B6%E6%8E%92%E5%8F%AF%E8%A7%86%E5%8C%96"><span class="toc-number">3.</span> <span class="toc-text">多模态数据的并排可视化</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E9%89%B4%E5%AE%9A%E4%B8%8D%E5%90%8C%E7%BB%86%E8%83%9E%E9%9B%86%E7%BE%A4%E7%9A%84%E8%A1%A8%E9%9D%A2%E8%9B%8B%E7%99%BD%E6%A0%87%E8%AE%B0"><span class="toc-number">4.</span> <span class="toc-text">鉴定不同细胞集群的表面蛋白标记</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%A4%9A%E6%A8%A1%E6%80%81%E6%95%B0%E6%8D%AE%E7%9A%84%E5%85%B6%E4%BB%96%E5%8F%AF%E8%A7%86%E5%8C%96%E6%96%B9%E6%B3%95"><span class="toc-number">5.</span> <span class="toc-text">多模态数据的其他可视化方法</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%8A%A0%E8%BD%BD%E9%80%9A%E8%BF%8710X-Multiome%E8%8E%B7%E5%BE%97%E7%9A%84%E5%A4%9A%E6%A8%A1%E6%80%81%E6%95%B0%E6%8D%AE"><span class="toc-number">6.</span> <span class="toc-text">加载通过10X Multiome获得的多模态数据</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%AE%8C%E6%95%B4%E4%BB%A3%E7%A0%81"><span class="toc-number">7.</span> <span class="toc-text">完整代码</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>最新文章</span></div><div class="aside-list"><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2025/09/25/%E3%80%90%E7%9C%BC%E7%A7%91%E5%AD%A6%E3%80%91%E7%BB%93%E8%86%9C%E7%96%BE%E7%97%85/" title="【眼科学】结膜疾病">【眼科学】结膜疾病</a><time datetime="2025-09-25T13:58:52.000Z" title="发表于 2025-09-25 21:58:52">2025-09-25</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2025/09/25/%E3%80%90%E7%9C%BC%E7%A7%91%E5%AD%A6%E3%80%91%E8%A7%92%E8%86%9C%E7%96%BE%E7%97%85/" title="【眼科学】角膜疾病">【眼科学】角膜疾病</a><time datetime="2025-09-25T11:39:15.000Z" title="发表于 2025-09-25 19:39:15">2025-09-25</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2025/09/25/%E3%80%90%E7%9C%BC%E7%A7%91%E5%AD%A6%E3%80%91%E7%9C%BC%E7%A7%91%E5%AD%A6%E5%9F%BA%E7%A1%80/" title="【眼科学】眼科学基础">【眼科学】眼科学基础</a><time datetime="2025-09-25T08:23:24.000Z" title="发表于 2025-09-25 16:23:24">2025-09-25</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2025/09/20/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E6%98%A0%E5%B0%84%E5%92%8C%E6%B3%A8%E9%87%8A%E6%9F%A5%E8%AF%A2%E6%95%B0%E6%8D%AE%E9%9B%86/" title="生信小记：映射和注释查询数据集"><img src="/images/Seurat.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="生信小记：映射和注释查询数据集"/></a><div class="content"><a class="title" href="/2025/09/20/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E6%98%A0%E5%B0%84%E5%92%8C%E6%B3%A8%E9%87%8A%E6%9F%A5%E8%AF%A2%E6%95%B0%E6%8D%AE%E9%9B%86/" title="生信小记：映射和注释查询数据集">生信小记：映射和注释查询数据集</a><time datetime="2025-09-20T04:39:57.000Z" title="发表于 2025-09-20 12:39:57">2025-09-20</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2025/09/14/%E3%80%90%E5%A6%87%E4%BA%A7%E5%AD%A6%E3%80%91%E5%A6%8A%E5%A8%A0%E7%94%9F%E7%90%86/" title="【妇产学】妊娠生理">【妇产学】妊娠生理</a><time datetime="2025-09-14T01:55:34.000Z" title="发表于 2025-09-14 09:55:34">2025-09-14</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2019 - 2025 By Nahida</div><div class="framework-info"><span>框架 </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo 7.3.0</a><span class="footer-separator">|</span><span>主题 </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly 5.3.5</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="阅读模式"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="日间和夜间模式切换"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="单栏和双栏切换"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside-config" type="button" title="设置"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="目录"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="回到顶部"><span class="scroll-percent"></span><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><div class="js-pjax"><script>(() => {
  const loadMathjax = () => {
    if (!window.MathJax) {
      window.MathJax = {
        tex: {
          inlineMath: [['$', '$'], ['\\(', '\\)']],
          tags: 'none',
        },
        chtml: {
          scale: 1.1
        },
        options: {
          enableMenu: true,
          renderActions: {
            findScript: [10, doc => {
              for (const node of document.querySelectorAll('script[type^="math/tex"]')) {
                const display = !!node.type.match(/; *mode=display/)
                const math = new doc.options.MathItem(node.textContent, doc.inputJax[0], display)
                const text = document.createTextNode('')
                node.parentNode.replaceChild(text, node)
                math.start = {node: text, delim: '', n: 0}
                math.end = {node: text, delim: '', n: 0}
                doc.math.push(math)
              }
            }, '']
          }
        }
      }

      const script = document.createElement('script')
      script.src = 'https://cdn.jsdelivr.net/npm/mathjax/es5/tex-mml-chtml.min.js'
      script.id = 'MathJax-script'
      script.async = true
      document.head.appendChild(script)
    } else {
      MathJax.startup.document.state(0)
      MathJax.texReset()
      MathJax.typesetPromise()
    }
  }

  btf.addGlobalFn('encrypt', loadMathjax, 'mathjax')
  window.pjax ? loadMathjax() : window.addEventListener('load', loadMathjax)
})()</script></div><script type="text/javascript" src="/js/memos/waterfall.min.js"></script><script type="text/javascript" src="/js/memos/imgStatus.min.js"></script><script type="text/javascript" src="/js/memos/lately.min.js"></script><script type="text/javascript" src="/js/memos/photo.js"></script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>