<!DOCTYPE html><html lang="zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0,viewport-fit=cover"><title>生信小记：Seurat的常见分析工作流程：以PBMC_3K数据集为例 | Nahida的净善宫</title><meta name="author" content="Nahida"><meta name="copyright" content="Nahida"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="更新日志2025&#x2F;4&#x2F;26 更正了“确认数据集维度”中，对肘部图意义解释的错误。 修改了“执行非线性降维”引语中，对于PCA降维不足之处的表述。 在“创建Seurat对象”中，新增对于读入文件的解释。 修正了一些排版错误。   在网上找了好几篇Seurat包的教程发现对新手都不太友好……遂自己写了一篇…… 创建Seurat对象我们将对10X Genomics免费提供的PBMC数">
<meta property="og:type" content="article">
<meta property="og:title" content="生信小记：Seurat的常见分析工作流程：以PBMC_3K数据集为例">
<meta property="og:url" content="http://example.com/2025/03/28/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9ASeurat%E7%9A%84%E5%B8%B8%E8%A7%81%E5%88%86%E6%9E%90%E5%B7%A5%E4%BD%9C%E6%B5%81%E7%A8%8B%EF%BC%9A%E4%BB%A5PBMC-3K%E6%95%B0%E6%8D%AE%E9%9B%86%E4%B8%BA%E4%BE%8B/index.html">
<meta property="og:site_name" content="Nahida的净善宫">
<meta property="og:description" content="更新日志2025&#x2F;4&#x2F;26 更正了“确认数据集维度”中，对肘部图意义解释的错误。 修改了“执行非线性降维”引语中，对于PCA降维不足之处的表述。 在“创建Seurat对象”中，新增对于读入文件的解释。 修正了一些排版错误。   在网上找了好几篇Seurat包的教程发现对新手都不太友好……遂自己写了一篇…… 创建Seurat对象我们将对10X Genomics免费提供的PBMC数">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="http://example.com/images/Seurat.png">
<meta property="article:published_time" content="2025-03-28T02:02:28.000Z">
<meta property="article:modified_time" content="2025-07-10T07:30:03.225Z">
<meta property="article:author" content="Nahida">
<meta property="article:tag" content="生信分析">
<meta property="article:tag" content="R语言">
<meta property="article:tag" content="Seurat">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://example.com/images/Seurat.png"><script type="application/ld+json">{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "生信小记：Seurat的常见分析工作流程：以PBMC_3K数据集为例",
  "url": "http://example.com/2025/03/28/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9ASeurat%E7%9A%84%E5%B8%B8%E8%A7%81%E5%88%86%E6%9E%90%E5%B7%A5%E4%BD%9C%E6%B5%81%E7%A8%8B%EF%BC%9A%E4%BB%A5PBMC-3K%E6%95%B0%E6%8D%AE%E9%9B%86%E4%B8%BA%E4%BE%8B/",
  "image": "http://example.com/images/Seurat.png",
  "datePublished": "2025-03-28T02:02:28.000Z",
  "dateModified": "2025-07-10T07:30:03.225Z",
  "author": [
    {
      "@type": "Person",
      "name": "Nahida",
      "url": "http://example.com/"
    }
  ]
}</script><link rel="shortcut icon" href="/img/avatar.jpg"><link rel="canonical" href="http://example.com/2025/03/28/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9ASeurat%E7%9A%84%E5%B8%B8%E8%A7%81%E5%88%86%E6%9E%90%E5%B7%A5%E4%BD%9C%E6%B5%81%E7%A8%8B%EF%BC%9A%E4%BB%A5PBMC-3K%E6%95%B0%E6%8D%AE%E9%9B%86%E4%B8%BA%E4%BE%8B/index.html"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css"><script>
    (() => {
      
    const saveToLocal = {
      set: (key, value, ttl) => {
        if (!ttl) return
        const expiry = Date.now() + ttl * 86400000
        localStorage.setItem(key, JSON.stringify({ value, expiry }))
      },
      get: key => {
        const itemStr = localStorage.getItem(key)
        if (!itemStr) return undefined
        const { value, expiry } = JSON.parse(itemStr)
        if (Date.now() > expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return value
      }
    }

    window.btf = {
      saveToLocal,
      getScript: (url, attr = {}) => new Promise((resolve, reject) => {
        const script = document.createElement('script')
        script.src = url
        script.async = true
        Object.entries(attr).forEach(([key, val]) => script.setAttribute(key, val))
        script.onload = script.onreadystatechange = () => {
          if (!script.readyState || /loaded|complete/.test(script.readyState)) resolve()
        }
        script.onerror = reject
        document.head.appendChild(script)
      }),
      getCSS: (url, id) => new Promise((resolve, reject) => {
        const link = document.createElement('link')
        link.rel = 'stylesheet'
        link.href = url
        if (id) link.id = id
        link.onload = link.onreadystatechange = () => {
          if (!link.readyState || /loaded|complete/.test(link.readyState)) resolve()
        }
        link.onerror = reject
        document.head.appendChild(link)
      }),
      addGlobalFn: (key, fn, name = false, parent = window) => {
        if (!false && key.startsWith('pjax')) return
        const globalFn = parent.globalFn || {}
        globalFn[key] = globalFn[key] || {}
        globalFn[key][name || Object.keys(globalFn[key]).length] = fn
        parent.globalFn = globalFn
      }
    }
  
      
      const activateDarkMode = () => {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      const activateLightMode = () => {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }

      btf.activateDarkMode = activateDarkMode
      btf.activateLightMode = activateLightMode

      const theme = saveToLocal.get('theme')
    
          theme === 'dark' ? activateDarkMode() : theme === 'light' ? activateLightMode() : null
        
      
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        document.documentElement.classList.toggle('hide-aside', asideStatus === 'hide')
      }
    
      
    const detectApple = () => {
      if (/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)) {
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
  
    })()
  </script><script>const GLOBAL_CONFIG = {
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  highlight: {"plugin":"highlight.js","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false,"highlightFullpage":false,"highlightMacStyle":false},
  copy: {
    success: '复制成功',
    error: '复制失败',
    noSupport: '浏览器不支持'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  dateSuffix: {
    just: '刚刚',
    min: '分钟前',
    hour: '小时前',
    day: '天前',
    month: '个月前'
  },
  copyright: undefined,
  lightbox: 'null',
  Snackbar: undefined,
  infinitegrid: {
    js: 'https://cdn.jsdelivr.net/npm/@egjs/infinitegrid/dist/infinitegrid.min.js',
    buttonText: '加载更多'
  },
  isPhotoFigcaption: false,
  islazyloadPlugin: false,
  isAnchor: false,
  percent: {
    toc: true,
    rightside: false,
  },
  autoDarkmode: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: '生信小记：Seurat的常见分析工作流程：以PBMC_3K数据集为例',
  isHighlightShrink: false,
  isToc: true,
  pageType: 'post'
}</script><meta name="generator" content="Hexo 7.3.0"><link rel="alternate" href="/atom.xml" title="Nahida的净善宫" type="application/atom+xml">
</head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img text-center"><img src="/img/avatar.jpg" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="site-data text-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">83</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">48</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">20</div></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> 时间轴</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> 标签</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> 分类</span></a></div><div class="menus_item"><a class="site-page" href="/photos/"><i class="fa-fw fas fa-images"></i><span> 相册</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> 友链</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url(/images/Seurat.png);"><nav id="nav"><span id="blog-info"><a class="nav-site-title" href="/"><img class="site-icon" src="/image/test.gif" alt="Logo"><span class="site-name">Nahida的净善宫</span></a><a class="nav-page-title" href="/"><span class="site-name">生信小记：Seurat的常见分析工作流程：以PBMC_3K数据集为例</span></a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> 时间轴</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> 标签</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> 分类</span></a></div><div class="menus_item"><a class="site-page" href="/photos/"><i class="fa-fw fas fa-images"></i><span> 相册</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> 友链</span></a></div></div><div id="toggle-menu"><span class="site-page"><i class="fas fa-bars fa-fw"></i></span></div></div></nav><div id="post-info"><h1 class="post-title">生信小记：Seurat的常见分析工作流程：以PBMC_3K数据集为例</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">发表于</span><time class="post-meta-date-created" datetime="2025-03-28T02:02:28.000Z" title="发表于 2025-03-28 10:02:28">2025-03-28</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">更新于</span><time class="post-meta-date-updated" datetime="2025-07-10T07:30:03.225Z" title="更新于 2025-07-10 15:30:03">2025-07-10</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/">生信分析</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title=""><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">浏览量:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="container post-content" id="article-container"><h1 id="更新日志"><a href="#更新日志" class="headerlink" title="更新日志"></a>更新日志</h1><h2 id="2025-4-26"><a href="#2025-4-26" class="headerlink" title="2025&#x2F;4&#x2F;26"></a>2025&#x2F;4&#x2F;26</h2><ol>
<li>更正了“确认数据集维度”中，对肘部图意义解释的错误。</li>
<li>修改了“执行非线性降维”引语中，对于PCA降维不足之处的表述。</li>
<li>在“创建Seurat对象”中，新增对于读入文件的解释。</li>
<li>修正了一些排版错误。</li>
</ol>
<hr>
<p>在网上找了好几篇Seurat包的教程发现对新手都不太友好……遂自己写了一篇……</p>
<h1 id="创建Seurat对象"><a href="#创建Seurat对象" class="headerlink" title="创建Seurat对象"></a>创建Seurat对象</h1><p>我们将对10X Genomics免费提供的PBMC数据集进行分析，包括2700个由Illumina NextSeq测序平台测序的单细胞。我们的原始数据包括3个文件：<code>barcodes.tsv</code>、<code>genes.tsv</code>、<code>matrix.mtx</code>。分别用于存储每个单细胞的细胞条形码、基因的标识符和名称、存储基因表达数据。</p>
<h2 id="数据的读取"><a href="#数据的读取" class="headerlink" title="数据的读取"></a>数据的读取</h2><p><code>Read10X()</code>函数可以从10X cellranger pipeline的输出中读取数据，并形成一个唯一的分子识别（UMI）计数矩阵。矩阵中的每个值代表所测序的每个细胞中的每个特征（如基因）。对于h5文件格式的文件，可以用<code>Read10X_h5()</code>读取。 </p>
<blockquote>
<p>10X Cell Ranger Pipeline 是 10X Genomics 公司开发的一套标准化生物信息分析工具，专门用于处理其单细胞测序平台（如 Chromium）生成的数据。它的核心功能是将原始的测序数据（fastq 文件）转化为可用于下游分析的基因表达矩阵。</p>
</blockquote>
<blockquote>
<p>分子识别（UMI）计数矩阵：在PCR扩增中，同一mRNA分子经PCR扩增后，会产生多个重度的测序片段，如不加干预会导致测得的mRNA表达量偏高。为了避免这种情况，我们在逆转录生成cDNA时会在每个mRNA分子上加上唯一的UMI序列和细胞Barcode（细胞条形码）。这时同一个mRNA经扩增后产生的多个重复序列的UMI完全相同，在数据分析时可以借此消除扩增重复而产生的偏差。</p>
</blockquote>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">#install.packages(&quot;dplyr&quot;)</span><br><span class="line">#install.packages(&quot;Seurat&quot;)</span><br><span class="line">#install.packages(&quot;patchwork&quot;)</span><br><span class="line">#安装R包的过程，如已安装可省略</span><br><span class="line"></span><br><span class="line">library(dplyr)</span><br><span class="line">library(Seurat)</span><br><span class="line">library(patchwork)</span><br><span class="line"></span><br><span class="line">pbmc.data &lt;- Read10X(data.dir = &quot;C:/Users/86150/Downloads/pbmc3k_filtered_gene_bc_matrices/filtered_gene_bc_matrices/hg19&quot;)</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<h2 id="Seurat对象的创建"><a href="#Seurat对象的创建" class="headerlink" title="Seurat对象的创建"></a>Seurat对象的创建</h2><p>在这一步中，我们将创建一个Seurat对象，用于储存我们读入的原始数据和分析结果（如PCA或聚类结果）。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pbmc &lt;- CreateSeuratObject(counts = pbmc.data, project = &quot;pbmc3k&quot;, min.cells = 3,min.features = 200)</span><br><span class="line">pbmc</span><br></pre></td></tr></table></figure>

<ul>
<li>输出结果</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">&gt; pbmc &lt;- CreateSeuratObject(counts = pbmc.data, project = &quot;pbmc3k&quot;, min.cells = 3,min.features = 200)</span><br><span class="line">Warning: Feature names cannot have underscores (&#x27;_&#x27;), replacing with dashes (&#x27;-&#x27;)</span><br><span class="line">&gt; pbmc</span><br><span class="line">An object of class Seurat </span><br><span class="line">13714 features across 2700 samples within 1 assay </span><br><span class="line">Active assay: RNA (13714 features, 0 variable features)</span><br><span class="line"> 1 layer present: counts</span><br></pre></td></tr></table></figure>

<h1 id="标准预处理工作流程"><a href="#标准预处理工作流程" class="headerlink" title="标准预处理工作流程"></a>标准预处理工作流程</h1><h2 id="质量控制（QC）和选择细胞进行进一步分析"><a href="#质量控制（QC）和选择细胞进行进一步分析" class="headerlink" title="质量控制（QC）和选择细胞进行进一步分析"></a>质量控制（QC）和选择细胞进行进一步分析</h2><p>单细胞测序计数高度复杂，在测序过程中，细胞死亡或破裂、空微珠（未捕获细胞的微珠，仅包括少量环境RNA）、双胞体或多胞体（两个或多个细胞被包裹到同一微滴中）、环境RNA污染等，都会为测序引入噪声和偏差。因此，我们需要对原始数据进行质量控制（QC）。</p>
<p>常用的质量控制指标包括：</p>
<ul>
<li>细胞独特基因数量：低质量细胞中表达的基因数量很少，空微珠仅有少数环境RNA污染，因此检测到的基因数量偏低。同时，双胞体或多胞体中，单个微珠中包含了两个及两个以上的细胞，因此检测到的基因数量会偏高。</li>
<li>细胞内分子总数：细胞内分子总数和细胞独特基因数量密切相关，因此也可以借此筛除不合格的数据。</li>
<li>线粒体基因比例：即线粒体基因UMI占总UMI的比例。在死亡或破裂的细胞中，线粒体RNA泄露而更容易被捕获，其线粒体基因比例偏高；而在空微珠中，仅有少量污染的环境RNA，因此其线粒体基因比例偏低。</li>
</ul>
<p>在这里，我们采用上述指标作为QC指标。其中，我们以“MT-”开头的所有基因的集合作为一组线粒体基因，利用<code>PercentageFeatureSet()</code>函数计算线粒体基因比例。然后，我们将QC指标进行可视化，借此筛选细胞。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">#计算线粒体基因比例</span><br><span class="line">pbmc[[&quot;percent.mt&quot;]] &lt;- PercentageFeatureSet(pbmc, pattern = &quot;^MT-&quot;)</span><br><span class="line">#可视化，自左向右为细胞独特基因数量、细胞内分子总数和线粒体基因比例。</span><br><span class="line">VlnPlot(pbmc, features = c(&quot;nFeature_RNA&quot;, &quot;nCount_RNA&quot;, &quot;percent.mt&quot;), ncol =3)</span><br></pre></td></tr></table></figure>
<p><img src="/images/PBMC3k/5d17b893-835a-45ef-977c-b9314b586731.png"></p>
<p>接下来，我们利用<code>FeatureScatter()</code>函数，生成并组合nCount_RNA和percent.mt，nCount_RNA和nFeature_RNA的散点图，有利于我们从多维视角，更精准地识别低质量细胞并确定合理的过滤标准。</p>
<table>
<thead>
<tr>
<th align="center">模式特征</th>
<th align="center">可能原因</th>
</tr>
</thead>
<tbody><tr>
<td align="center">低nCount_RNA和高percent.mt</td>
<td align="center">细胞破裂或死亡</td>
</tr>
<tr>
<td align="center">高nCount_RNA和高nFeature_RNA</td>
<td align="center">双胞体或多胞体</td>
</tr>
<tr>
<td align="center">低nCount_RNA和低nFeature_RNA</td>
<td align="center">空微珠或环境RNA污染</td>
</tr>
</tbody></table>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">plot1 &lt;- FeatureScatter(pbmc, feature1 = &quot;nCount_RNA&quot;, feature2 = &quot;percent.mt&quot;)</span><br><span class="line">plot2 &lt;- FeatureScatter(pbmc, feature1 = &quot;nCount_RNA&quot;, feature2 = &quot;nFeature_RNA&quot;)</span><br><span class="line">plot1 + plot2</span><br></pre></td></tr></table></figure>
<p><img src="/images/PBMC3k/350f9624-71bf-4a93-8e1b-a7dfcdd544fd.png"></p>
<p>最后，我们根据下述标准，过滤低质量细胞。</p>
<ul>
<li>基因数量（特征计数）超过2500或少于200。</li>
<li>线粒体计数大于5%。</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">pbmc &lt;- subset(pbmc, subset = nFeature_RNA &gt; 200 &amp; nFeature_RNA &lt; 2500 &amp; percent.mt &lt; 5)</span><br></pre></td></tr></table></figure>

<h1 id="数据标准化"><a href="#数据标准化" class="headerlink" title="数据标准化"></a>数据标准化</h1><p>由于测序技术的缺陷，每个细胞测序深度的差异难以保证完全相同。测序深度高的细胞所检测到的总UMI数（可以理解为总mRNA数）偏多，而测序深度浅的细胞所检测到的总UMI数偏少，而这种差异往往会掩盖真正的生物学表达差异。为了消除这种技术性偏差，一般情况下我们会使用一种全局缩放标准化方法”LogNormalize”（先叫它“对数标准化法”叭），该方法的原理是：将每个细胞的基因表达值除以该细胞基因的总表达量，再乘以缩放因子数（一般为10000），最后对结果进行对数转换得到标准化的表达量。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pbmc &lt;- NormalizeData(pbmc, normalization.method = &quot;LogNormalize&quot;, scale.factor = 10000)</span><br><span class="line">#pbmc &lt;- NormalizeData(pbmc) #简写版</span><br></pre></td></tr></table></figure>

<p>对数标准化法通过消除细胞间总UMI数的差异，消除了测序深浅对分析结果造成的偏差。然而，细胞间总UMI数的差异并不一定（或者说并不全是）由测序深度的差异造成的，也有可能是细胞间“天然”的表达量差异。而忽略这种“天然”差异、一刀切地消除细胞间总UMI数的差异，也会对分析结果造成影响，产生偏差。为了更精准地分离上述偏差，可以使用Seurat包中的“SCTransform”法进行数据标准化，在本文不进一步展开。</p>
<h1 id="识别高变基因"><a href="#识别高变基因" class="headerlink" title="识别高变基因"></a>识别高变基因</h1><p>高变基因（Highly Variable Genes，HVGs）是指不同细胞之间表达差异显著的一类基因（这里需要和高表达基因作区分，高变基因侧重于“方差大”，而高表达基因则侧重于“均值大”）。鉴定高变基因有利于更加突出单细胞数据集中的生物信号。</p>
<p>在进行筛选之前，我们需要先理顺在前面步骤已经处理的和尚未处理的偏差。在质量控制这一步中，我们清除了低质量细胞，排除了破裂或死亡细胞、双胞体&#x2F;多胞体等不合格细胞的影响；在数据标准化这一步中，我们消除了技术限制所带来的不同细胞间总UMI数的差异。尽管如此，由于技术限制，我们很难保证在单个细胞内不同mRNA均得到了同等水平的扩增，这也会对分析产生影响，尤其对于低表达基因，测序技术的偏差往往远大于其真实表达水平的差异。而对于中高表达基因，其真实表达水平的差异往往会远大于测序技术所带来的偏差，此时后者可以忽略。因此我们需要在消除上述技术偏差的前提下筛选高变基因。常用的方法有vst法、mean.var.plot法、dispersion法等，它们都是基于均值-方差关系（即表达量-表达差异关系）进行筛选的方法。此处选择vst法进行筛选。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">pbmc &lt;- FindVariableFeatures(pbmc, selection.method = &quot;vst&quot;, nfeatures = 2000)</span><br><span class="line">top10 &lt;- head(VariableFeatures(pbmc), 10) #确定前10个高变基因</span><br><span class="line"></span><br><span class="line">#可视化</span><br><span class="line">plot1 &lt;- VariableFeaturePlot(pbmc)</span><br><span class="line">plot2 &lt;- LabelPoints(plot = plot1, points = top10, repel = TRUE)</span><br><span class="line">plot1 + plot2 #在运行这一步之前建议将右下角的plots窗口调至最大，否则很容易报错……也可以用ggsave()函数保存后打开查看</span><br></pre></td></tr></table></figure>

<p><img src="/images/PBMC3k/66818c4a-912c-4649-b4ab-6b7cde29a739.png"></p>
<h1 id="数据缩放（Z-score标准化）"><a href="#数据缩放（Z-score标准化）" class="headerlink" title="数据缩放（Z-score标准化）"></a>数据缩放（Z-score标准化）</h1><p>为了理解数据缩放的必要性，首先我们需要对PCA降维算法具有一些初步了解。</p>
<p>在测序之后，我们可以得到一组有n个基因表达水平的信息，这些基因之间相互影响，而我们需要找出这些基因之间的相互关系，确定不同细胞基因表达模式的不同。如果只有2个基因，我们会考虑使用一个二维的坐标系来表示不同的细胞，x轴、y轴分别表示两个基因的表达水平，而坐标系中的每个点则代表每一个细胞。如果两个点的距离越近，则说明这两个细胞的基因表达模式更加相近，两个细胞越有可能是同一种类的细胞（这时的图像比较直观）；如果有3个基因，那么就可以用三维的坐标系（还能用）；但测序结果中往往有许许多多个基因，这时候很难以图像的形式直观反映（许许多多维的图……怎么想都想不出来叭）。这时候我们需要对数据进行降维，去繁化简，将高维数据的特征反映在低维的图像上，从而为分析带来便利。</p>
<p>PCA算法的作用在于将一个$n$维的数据降低至$k$维的数据（$n＞k$），使得数据更加直观化。这$k$维中的每一维，我们称为主成分（PC），其反映了最能区分不同种类细胞的最主要差异（可以理解为，最能区分不同种类细胞的基因表达模式）。举个通俗的例子来说，一个城市的经济水平可以通过GDP、人均收入、消费水平、三大产业的比例等等许多指标反映（$n$维），这时候如果我们需要比较$m$个城市的经济水平，那我们手头上就有一个$n×m$的表格，一眼望过去难以直观对比不同城市的发展情况。通过PCA降维分析，我们总结出了几个最关键的因素：整体经济规模（对应GDP、人均收入、消费水平）、产业结构差异（对应三大产业分别的占比），等等等等（$k$维）。这一些我们总结出来的“最关键因素”也就是主成分。通过这种降维操作，我们将一个$n×m$的表格简化成为一个$n×k$的表格，比较起来就相对简单多了。我们将其对应至单细胞测序分析中，“城市经济水平”即对应“细胞基因表达模式”，“GDP、人均收入、消费水平、三大产业的比例”即对应“各个基因的表达水平”，“整体经济规模”“产业结构差异”则对应不同的基因表达水平的特定组合，这一组合可以反映不同细胞的总体基因表达模式的重要特征。</p>
<p>PCA的原理在此处不详细展开，但其核心为<strong>协方差</strong>的计算，这是我们现在需要关注的部分。协方差反映两个或多个因素之间的关联强度，简化版的协方差的计算公式为： $\text{Cov}(X,Y)&#x3D;\frac{1}{n-1}\sum_{i&#x3D;1}^{n}(x_i-\bar{X})(y_i-\bar{Y})$ 。在一个细胞中，不同基因的表达水平的数量级可能存在巨大差异，而这种数量级差异会对分析结果造成巨大影响。</p>
<p>举例来说，假设我们现在有<code>X</code>、<code>Y</code>两个基因的表达水平信息，如下：</p>
<table>
<thead>
<tr>
<th align="center">基因<code>X</code></th>
<th align="center">基因<code>Y</code></th>
</tr>
</thead>
<tbody><tr>
<td align="center">5</td>
<td align="center">15</td>
</tr>
<tr>
<td align="center">10</td>
<td align="center">30</td>
</tr>
<tr>
<td align="center">15</td>
<td align="center">45</td>
</tr>
</tbody></table>
<p>不难看出，两个基因的表达量呈完全正相关关系，此时直接计算两个基因之间的协方差，结果为25。</p>
<p>如果拿到的<code>X</code>、<code>Y</code>两个基因的表达水平信息如下：</p>
<table>
<thead>
<tr>
<th align="center">基因<code>X</code></th>
<th align="center">基因<code>Y</code></th>
</tr>
</thead>
<tbody><tr>
<td align="center">5</td>
<td align="center">100</td>
</tr>
<tr>
<td align="center">10</td>
<td align="center">200</td>
</tr>
<tr>
<td align="center">15</td>
<td align="center">300</td>
</tr>
</tbody></table>
<p>此时两个基因的表达水平仍然呈正相关关系，但如果直接计算两个基因之间的协方差，结果为500。</p>
<p>由上可见，两种情况中两个基因的关联强度都是一样强的（完全正相关）。但由于第二种情况中<code>Y</code>基因表达水平的数量级远远大于第一种情况，协方差的计算结果也出现了巨大的差异。这时如果直接进行比较，反而会得到“第二种情况中两种基因的关联强度远远大于第一种情况”的错误结论。因此，为了纠正这种情况，我们需要将数据“缩放”至同一数量级，在此基础上再进行降维分析。</p>
<p>Z-score标准化可以消除这种差异，其可以将每个变量的数据转换为均值为0、标准差为1的分布：$z&#x3D;\frac{x-μ}{σ}$，其中$μ$为变量的均值，$σ$为变量的标准差。对上述两种情况的数据进行标准化之后，其结果一致如下：</p>
<table>
<thead>
<tr>
<th align="center">基因<code>X</code>（标准化后）</th>
<th align="center">基因<code>Y</code>（标准化后）</th>
</tr>
</thead>
<tbody><tr>
<td align="center">-1</td>
<td align="center">-1</td>
</tr>
<tr>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr>
<td align="center">1</td>
<td align="center">1</td>
</tr>
</tbody></table>
<p>此时计算两个基因之间的协方差，结果均为1，表明它们具有相同的相关性模式。（有没有发现，标准化后的协方差其实就等于原始数据的相关系数！）</p>
<p>在Seurat包中，<code>ScaleData()</code>函数可以实现对数据的Z-score标准化，即①使每个基因的表达值偏倚，使得细胞间的平均表达为0；②对每个基因的表达值进行缩放，使得细胞间的方差为1。借此可以在保留协方差（即不同基因表达量之间的关系）的同时，将不同数量级表达水平的基因转化到同一水平，从而消除高表达基因的主导作用。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">all.genes &lt;- rownames(pbmc)</span><br><span class="line">pbmc &lt;- ScaleData(pbmc, features = all.genes)</span><br></pre></td></tr></table></figure>

<h1 id="执行线性降维"><a href="#执行线性降维" class="headerlink" title="执行线性降维"></a>执行线性降维</h1><p>接下来，我们会对缩放后的数据进行PCA分析。默认情况下只会输入先前确定的HVGs，如需选择不同的子集，可以使用argument进行定义。</p>
<p>对于PCA分析的结果，Seurat包提供了多种可视化的方式，包括<code>VizDimReduction()</code>、<code>DimPlot()</code>和<code>DimHeatmap()</code>等。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">pbmc &lt;- RunPCA(pbmc, features = VariableFeatures(object = pbmc))</span><br><span class="line">#检验并可视化PCA</span><br><span class="line">print(pbmc[[&quot;pca&quot;]], dims = 1:5, nfeatures = 5)</span><br></pre></td></tr></table></figure>
<ul>
<li>运行结果</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br></pre></td><td class="code"><pre><span class="line">&gt; pbmc &lt;- RunPCA(pbmc, features = VariableFeatures(object = pbmc))</span><br><span class="line">PC_ 1 </span><br><span class="line">Positive:  CST3, TYROBP, LST1, AIF1, FTL, FTH1, LYZ, FCN1, S100A9, TYMP </span><br><span class="line">	   FCER1G, CFD, LGALS1, S100A8, CTSS, LGALS2, SERPINA1, IFITM3, SPI1, CFP </span><br><span class="line">	   PSAP, IFI30, SAT1, COTL1, S100A11, NPC2, GRN, LGALS3, GSTP1, PYCARD </span><br><span class="line">Negative:  MALAT1, LTB, IL32, IL7R, CD2, B2M, ACAP1, CD27, STK17A, CTSW </span><br><span class="line">	   CD247, GIMAP5, AQP3, CCL5, SELL, TRAF3IP3, GZMA, MAL, CST7, ITM2A </span><br><span class="line">	   MYC, GIMAP7, HOPX, BEX2, LDLRAP1, GZMK, ETS1, ZAP70, TNFAIP8, RIC3 </span><br><span class="line">PC_ 2 </span><br><span class="line">Positive:  CD79A, MS4A1, TCL1A, HLA-DQA1, HLA-DQB1, HLA-DRA, LINC00926, CD79B, HLA-DRB1, CD74 </span><br><span class="line">	   HLA-DMA, HLA-DPB1, HLA-DQA2, CD37, HLA-DRB5, HLA-DMB, HLA-DPA1, FCRLA, HVCN1, LTB </span><br><span class="line">	   BLNK, P2RX5, IGLL5, IRF8, SWAP70, ARHGAP24, FCGR2B, SMIM14, PPP1R14A, C16orf74 </span><br><span class="line">Negative:  NKG7, PRF1, CST7, GZMB, GZMA, FGFBP2, CTSW, GNLY, B2M, SPON2 </span><br><span class="line">	   CCL4, GZMH, FCGR3A, CCL5, CD247, XCL2, CLIC3, AKR1C3, SRGN, HOPX </span><br><span class="line">	   TTC38, APMAP, CTSC, S100A4, IGFBP7, ANXA1, ID2, IL32, XCL1, RHOC </span><br><span class="line">PC_ 3 </span><br><span class="line">Positive:  HLA-DQA1, CD79A, CD79B, HLA-DQB1, HLA-DPB1, HLA-DPA1, CD74, MS4A1, HLA-DRB1, HLA-DRA </span><br><span class="line">	   HLA-DRB5, HLA-DQA2, TCL1A, LINC00926, HLA-DMB, HLA-DMA, CD37, HVCN1, FCRLA, IRF8 </span><br><span class="line">	   PLAC8, BLNK, MALAT1, SMIM14, PLD4, P2RX5, IGLL5, LAT2, SWAP70, FCGR2B </span><br><span class="line">Negative:  PPBP, PF4, SDPR, SPARC, GNG11, NRGN, GP9, RGS18, TUBB1, CLU </span><br><span class="line">	   HIST1H2AC, AP001189.4, ITGA2B, CD9, TMEM40, PTCRA, CA2, ACRBP, MMD, TREML1 </span><br><span class="line">	   NGFRAP1, F13A1, SEPT5, RUFY1, TSC22D1, MPP1, CMTM5, RP11-367G6.3, MYL9, GP1BA </span><br><span class="line">PC_ 4 </span><br><span class="line">Positive:  HLA-DQA1, CD79B, CD79A, MS4A1, HLA-DQB1, CD74, HIST1H2AC, HLA-DPB1, PF4, SDPR </span><br><span class="line">	   TCL1A, HLA-DRB1, HLA-DPA1, HLA-DQA2, PPBP, HLA-DRA, LINC00926, GNG11, SPARC, HLA-DRB5 </span><br><span class="line">	   GP9, AP001189.4, CA2, PTCRA, CD9, NRGN, RGS18, CLU, TUBB1, GZMB </span><br><span class="line">Negative:  VIM, IL7R, S100A6, IL32, S100A8, S100A4, GIMAP7, S100A10, S100A9, MAL </span><br><span class="line">	   AQP3, CD2, CD14, FYB, LGALS2, GIMAP4, ANXA1, CD27, FCN1, RBP7 </span><br><span class="line">	   LYZ, S100A11, GIMAP5, MS4A6A, S100A12, FOLR3, TRABD2A, AIF1, IL8, IFI6 </span><br><span class="line">PC_ 5 </span><br><span class="line">Positive:  GZMB, NKG7, S100A8, FGFBP2, GNLY, CCL4, CST7, PRF1, GZMA, SPON2 </span><br><span class="line">	   GZMH, S100A9, LGALS2, CCL3, CTSW, XCL2, CD14, CLIC3, S100A12, RBP7 </span><br><span class="line">	   CCL5, MS4A6A, GSTP1, FOLR3, IGFBP7, TYROBP, TTC38, AKR1C3, XCL1, HOPX </span><br><span class="line">Negative:  LTB, IL7R, CKB, VIM, MS4A7, AQP3, CYTIP, RP11-290F20.3, SIGLEC10, HMOX1 </span><br><span class="line">	   LILRB2, PTGES3, MAL, CD27, HN1, CD2, GDI2, CORO1B, ANXA5, TUBA1B </span><br><span class="line">	   FAM110A, ATP1A1, TRADD, PPA1, CCDC109B, ABRACL, CTD-2006K23.1, WARS, VMO1, FYB </span><br></pre></td></tr></table></figure>

<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">&gt; #可视化PCA</span><br><span class="line">&gt; print(pbmc[[&quot;pca&quot;]], dims = 1:5, nfeatures = 5)</span><br><span class="line">PC_ 1 </span><br><span class="line">Positive:  CST3, TYROBP, LST1, AIF1, FTL </span><br><span class="line">Negative:  MALAT1, LTB, IL32, IL7R, CD2 </span><br><span class="line">PC_ 2 </span><br><span class="line">Positive:  CD79A, MS4A1, TCL1A, HLA-DQA1, HLA-DQB1 </span><br><span class="line">Negative:  NKG7, PRF1, CST7, GZMB, GZMA </span><br><span class="line">PC_ 3 </span><br><span class="line">Positive:  HLA-DQA1, CD79A, CD79B, HLA-DQB1, HLA-DPB1 </span><br><span class="line">Negative:  PPBP, PF4, SDPR, SPARC, GNG11 </span><br><span class="line">PC_ 4 </span><br><span class="line">Positive:  HLA-DQA1, CD79B, CD79A, MS4A1, HLA-DQB1 </span><br><span class="line">Negative:  VIM, IL7R, S100A6, IL32, S100A8 </span><br><span class="line">PC_ 5 </span><br><span class="line">Positive:  GZMB, NKG7, S100A8, FGFBP2, GNLY </span><br><span class="line">Negative:  LTB, IL7R, CKB, VIM, MS4A7 </span><br></pre></td></tr></table></figure>

<p>运用<code>VimDimLoadings()</code>进行可视化：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">VizDimLoadings(pbmc, dims = 1:2, reduction = &quot;pca&quot;)</span><br></pre></td></tr></table></figure>
<p><img src="/images/PBMC3k/63d37bf9-d9db-46fa-8ff1-2c982bbbab4d.png"></p>
<p>运用<code>DimPlot()</code>进行可视化：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">DimPlot(pbmc, reduction = &quot;pca&quot;) + NoLegend()</span><br></pre></td></tr></table></figure>
<p><img src="/images/PBMC3k/7e4ef846-01f8-4e06-aa11-3f2cd283937e.png"></p>
<p>运用<code>DimHeatmap()</code>进行可视化：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)</span><br></pre></td></tr></table></figure>
<p><img src="/images/PBMC3k/5b8b49e9-3e72-488f-8997-9d43bfd0a338.png"></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)</span><br></pre></td></tr></table></figure>
<p><img src="/images/PBMC3k/e8dd52c6-b107-482d-a672-d553cecb4f17.png"></p>
<h1 id="确认数据集的维度"><a href="#确认数据集的维度" class="headerlink" title="确认数据集的维度"></a>确认数据集的维度</h1><p>在上一步中，我们通过PCA降维算法得到了若干个主成分。但如果我们选择全部的主成分进行下一步分析，一方面会导致计算量增大，不利于大数据集的分析；另一方面会导致过度细分细胞簇，不够直观。但反之，如果我们选择的主成分过少，则会导致细胞分簇的精度下降，不足以体现出不同种类细胞的生物学差异。因此，我们需要平衡二者，选择一个适当的主成分数量进行进一步分型。</p>
<p>肘部图可以通过计算并比较在选择不同数量的主成分进行聚类分簇后，每个细胞距离其所属簇中心的距离平方和（有资料称其为惯性、SSE），来确定选择主成分最佳数量。SSE数值越大，说明拟合的精度越低；反之亦然。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ElbowPlot(pbmc)</span><br></pre></td></tr></table></figure>
<p><img src="/images/PBMC3k/e628012a-65b9-4a83-bb9a-3f23d5647188.png"></p>
<p>通过上图我们不难看出，选择的主成分数量小于10之前，随着主成分数量的增加，SSE下降的非常快，说明精度随主成分数量的增加而迅速提升；但当选择的主成分数量超过10，SSE的下降速度反而变缓，说明性价比下降。因此，选择10个左右的主成分进行下一步分析比较适合。</p>
<p>值得一提的是，除了肘部图之外，我们还可以利用<code>JackStrawPlot()</code>函数对主成分进行筛选。但对于较大的数据集，其计算过程往往需要很长时间，因此肘部图仍然是最效率的选择。（为了得到下面这个图，我的电脑跑了三分钟，而肘部图几乎秒出……）</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">pbmc &lt;- JackStraw(pbmc, num.replicate = 100)</span><br><span class="line">pbmc &lt;- ScoreJackStraw(pbmc, dims = 1:20)</span><br><span class="line">JackStrawPlot(pbmc, dims = 1:15)</span><br></pre></td></tr></table></figure>
<p><img src="/images/PBMC3k/fff0fb76-be33-4e56-900c-05018665cfaf.png"></p>
<h1 id="细胞聚类"><a href="#细胞聚类" class="headerlink" title="细胞聚类"></a>细胞聚类</h1><p>在上面几步中，我们确认了不同细胞集群基因表达模式的主要变异方向，将$n×m$的表格降维到了$n×k$的表格。但是，我们还没有对降维后的数据进行分析，还没有确认“哪些细胞属于哪一类”。在这一步中，我们将会将降维后的数据划分为若干组（簇），每组代表一种潜在的细胞类型或状态，也就是细胞聚类。</p>
<p>细胞聚类分为领域图构建和聚类两步，前者通常用KNN算法（$K$-Nearest Neighbors）进行；而聚类则有多种方法，Seurat包中默认使用Louvain算法。大致原理如下：</p>
<ul>
<li>KNN算法（$K$-Nearest Neighbors）：通过测量降维后数据中每个细胞间的距离，找到距离每个细胞最近的$K$个细胞（这时我们称这些细胞为“邻居”），借此可以构建一个邻域图（其实就是一个拓扑图）。</li>
<li>Louvain算法：在通过KNN算法构建领域图的基础上，根据细胞距离的远近将所有划分为不同的“社区（community）”（即不同的细胞集群）。</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">pbmc &lt;- FindNeighbors(pbmc, dims = 1:10) #KNN算法</span><br><span class="line">pbmc &lt;- FindClusters(pbmc, resolution = 0.5) #Louvain算法</span><br><span class="line">head(Idents(pbmc), 5) #了解前5个细胞分别属于哪些集群</span><br></pre></td></tr></table></figure>

<ul>
<li>运行结果：</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">&gt; pbmc &lt;- FindNeighbors(pbmc, dims = 1:10)</span><br><span class="line">Computing nearest neighbor graph</span><br><span class="line">Computing SNN</span><br><span class="line">&gt; pbmc &lt;- FindClusters(pbmc, resolution = 0.5)</span><br><span class="line">Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span><br><span class="line"></span><br><span class="line">Number of nodes: 2638	</span><br><span class="line">Number of edges: 95927	</span><br><span class="line"></span><br><span class="line">Running Louvain algorithm...</span><br><span class="line">0%   10   20   30   40   50   60   70   80   90   100%</span><br><span class="line">[----|----|----|----|----|----|----|----|----|----|</span><br><span class="line">**************************************************|</span><br><span class="line">Maximum modularity in 10 random starts: 0.8728</span><br><span class="line">Number of communities: 9</span><br><span class="line">Elapsed time: 0 seconds</span><br><span class="line">&gt; head(Idents(pbmc), 5)</span><br><span class="line">AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACATTGATCAGC-1 AAACCGTGCTTCCG-1 AAACCGTGTATGCG-1 </span><br><span class="line">               2                3                2                1                6 </span><br><span class="line">Levels: 0 1 2 3 4 5 6 7 8</span><br></pre></td></tr></table></figure>

<h1 id="执行非线性降维（UMAP-t-SNE）"><a href="#执行非线性降维（UMAP-t-SNE）" class="headerlink" title="执行非线性降维（UMAP&#x2F;$t$-SNE）"></a>执行非线性降维（UMAP&#x2F;$t$-SNE）</h1><p>PCA降维可以对细胞进行初步分群，但其分群还不够精确直观。PCA没有很直观的可视化方法将不同细胞簇很好地分开来。以下图为例，我们只能选取两个主成分进行可视化，但就像简单粗暴地把一个纸团拍扁，尽管保留了整体的轮廓，但有可能因为纸团在压缩过程中发生重叠，丢失了哪一边是顶部、哪一边是底部的信息。因此为了更好地鉴别各个细胞群，我们需要更加精细地处理这些数据。<br><img src="/images/PBMC3k/25e225b7-051f-44c7-b2b2-2f9d688d9954.png"></p>
<p>在未经处理的高维数据中，我们想象用一个点来代表一个细胞。在高维空间中，同一种类细胞所代表的点之间的距离就会更近；非同类细胞之间的距离将会更远。依此原理，我们可以通过计算两两细胞间的距离，并根据点与点之间的距离生成一个二维图像。在高维空间中距离相近的两个点，在二维图像中距离也拉近；而在高维空间中距离较远的两个点，在二维图像中距离也会拉远，这就可以更好地反映出不同细胞群的区别。这就是UMAP和$t$-SNE算法的基本原理。</p>
<p>UMAP算法和$t$-SNE算法的思路总体相同，但存在一些细节上的差异。$t$-SNE算法会计算所有点与点之间的距离，并将其转化为$t$分布，结果就是在高维数据中两个距离较近的点，在降维后的数据中距离将会更近；而在高维数据中两个距离较远的点，在降维后的数据中距离也将拉远，会更有利于不同细胞集群的区分；而UMAP算法则只会计算距离某个点最近的$k$个点的距离（很像KNN算法），并且也不会对计算得到的距离进行$t$分布转化，因此通过UMAP算法降维后的图中，两点间距离会更有可比性。</p>
<p>此处以UMAP算法为例进行降维：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pbmc &lt;- RunUMAP(pbmc, dims = 1:10)</span><br><span class="line">DimPlot(pbmc, reduction = &quot;umap&quot;)</span><br></pre></td></tr></table></figure>
<p><img src="/images/PBMC3k/5ed1c3ce-8162-464f-99ab-d1944882c09a.png"></p>
<p>此时可以用下面的代码保存该结果，在需要时不需要重新运行上面的计算（计算量太大了……）。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">#检查output目录是否存在，若不存在则创建</span><br><span class="line">if (!dir.exists(&quot;C:/Users/86150/Downloads/pbmc3k_filtered_gene_bc_matrices/output&quot;)) &#123;</span><br><span class="line">  dir.create(&quot;C:/Users/86150/Downloads/pbmc3k_filtered_gene_bc_matrices/output&quot;)</span><br><span class="line">&#125;</span><br><span class="line">#保存</span><br><span class="line">saveRDS(pbmc, file = &quot;C:/Users/86150/Downloads/pbmc3k_filtered_gene_bc_matrices/output/pbmc_tutorial.rds&quot;)</span><br><span class="line"></span><br><span class="line">#重新加载</span><br><span class="line">library(Seurat) #加载Seurat包</span><br><span class="line">pbmc_restored &lt;- readRDS(&quot;C:/Users/86150/Downloads/pbmc3k_filtered_gene_bc_matrices/output/pbmc_tutorial.rds&quot;) #读取RDS文件</span><br><span class="line">DimPlot(pbmc_restored, reduction = &quot;umap&quot;) </span><br></pre></td></tr></table></figure>
<h1 id="鉴定差异表达基因（DEGs）"><a href="#鉴定差异表达基因（DEGs）" class="headerlink" title="鉴定差异表达基因（DEGs）"></a>鉴定差异表达基因（DEGs）</h1><p>Seurat包可以通过差异表达分析寻找不同细胞簇之间的表达基因，默认情况下其将会寻找所有上调和下调的基因。<code>FindAllMarkers()</code>函数会对所有细胞集群自动进行差异表达分析，我们也可以利用<code>FindMarkers()</code>函数对特定的细胞集群寻找。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">cluster2.markers &lt;- FindMarkers(pbmc, ident.1 = 2) #和其他所有细胞簇进行对比，寻找细胞簇2的差异表达基因</span><br><span class="line">head(cluster2.markers, n = 5) #查看前5个差异表达基因</span><br></pre></td></tr></table></figure>
<ul>
<li>运行结果</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">&gt; head(cluster2.markers, n = 5)</span><br><span class="line">            p_val avg_log2FC pct.1 pct.2    p_val_adj</span><br><span class="line">IL32 2.892340e-90  1.3070772 0.947 0.465 3.966555e-86</span><br><span class="line">LTB  1.060121e-86  1.3312674 0.981 0.643 1.453850e-82</span><br><span class="line">CD3D 8.794641e-71  1.0597620 0.922 0.432 1.206097e-66</span><br><span class="line">IL7R 3.516098e-68  1.4377848 0.750 0.326 4.821977e-64</span><br><span class="line">LDHB 1.642480e-67  0.9911924 0.954 0.614 2.252497e-63</span><br></pre></td></tr></table></figure>
<ul>
<li>结果解读：<ul>
<li>p_val：原始p值（未校正的多重假设检验结果）</li>
<li>avg_log2FC：目标细胞簇与其他细胞簇的平均log2倍变化（正值表示上调）</li>
<li>pct.1：目标细胞簇中表达该基因的细胞比例</li>
<li>pct.2：其他细胞簇中表达该基因的细胞比例</li>
<li>p_val_adj：（利用Bonferroni或BH方法）校正后的p值</li>
</ul>
</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">cluster5.markers &lt;- FindMarkers(pbmc, ident.1 = 5, ident.2 = c(0, 3)) #和细胞簇0和细胞簇3对比，寻找细胞簇5的差异基因</span><br><span class="line">head(cluster5.markers, n = 5) #查看前5个差异基因</span><br></pre></td></tr></table></figure>
<ul>
<li>运行结果</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">&gt; head(cluster5.markers, n = 5)</span><br><span class="line">                      p_val avg_log2FC pct.1 pct.2     p_val_adj</span><br><span class="line">FCGR3A        8.246578e-205   6.794969 0.975 0.040 1.130936e-200</span><br><span class="line">IFITM3        1.677613e-195   6.192558 0.975 0.049 2.300678e-191</span><br><span class="line">CFD           2.401156e-193   6.015172 0.938 0.038 3.292945e-189</span><br><span class="line">CD68          2.900384e-191   5.530330 0.926 0.035 3.977587e-187</span><br><span class="line">RP11-290F20.3 2.513244e-186   6.297999 0.840 0.017 3.446663e-182</span><br></pre></td></tr></table></figure>


<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">#和其他细胞簇对比，寻找每一个细胞簇的标志基因（设置仅寻找上调基因）</span><br><span class="line">pbmc.markers &lt;- FindAllMarkers(pbmc, only.pos = TRUE) #留意only.pos = TRUE，默认情况下上调基因和下调基因都会找出来，需要只寻找上调基因的时候需要特别设置</span><br><span class="line">pbmc.markers %&gt;%</span><br><span class="line">  group_by(cluster) %&gt;% #不同的细胞簇分组列出</span><br><span class="line">  dplyr::filter(avg_log2FC &gt; 1) #进一步筛选高变基因</span><br></pre></td></tr></table></figure>
<ul>
<li>运行结果</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">&gt; pbmc.markers %&gt;%</span><br><span class="line">+   group_by(cluster) %&gt;%</span><br><span class="line">+   dplyr::filter(avg_log2FC &gt; 1)</span><br><span class="line"># A tibble: 7,019 × 7</span><br><span class="line"># Groups:   cluster [9]</span><br><span class="line">       p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene     </span><br><span class="line">       &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt;    </span><br><span class="line"> 1 3.75e-112       1.21 0.912 0.592 5.14e-108 0       LDHB     </span><br><span class="line"> 2 9.57e- 88       2.40 0.447 0.108 1.31e- 83 0       CCR7     </span><br><span class="line"> 3 1.15e- 76       1.06 0.845 0.406 1.58e- 72 0       CD3D     </span><br><span class="line"> 4 1.12e- 54       1.04 0.731 0.4   1.54e- 50 0       CD3E     </span><br><span class="line"> 5 1.35e- 51       2.14 0.342 0.103 1.86e- 47 0       LEF1     </span><br><span class="line"> 6 1.94e- 47       1.20 0.629 0.359 2.66e- 43 0       NOSIP    </span><br><span class="line"> 7 2.81e- 44       1.53 0.443 0.185 3.85e- 40 0       PIK3IP1  </span><br><span class="line"> 8 6.27e- 43       1.99 0.33  0.112 8.60e- 39 0       PRKCQ-AS1</span><br><span class="line"> 9 1.16e- 40       2.70 0.2   0.04  1.59e- 36 0       FHIT     </span><br><span class="line">10 1.34e- 34       1.96 0.268 0.087 1.84e- 30 0       MAL      </span><br><span class="line"># ℹ 7,009 more rows</span><br><span class="line"># ℹ Use `print(n = ...)` to see more rows</span><br></pre></td></tr></table></figure>

<p>利用ROC检验，可以筛选出有统计学意义且生物学意义显著的差异表达基因。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">cluster0.markers &lt;- FindMarkers(pbmc, ident.1 = 0, logfc.threshold = 0.25, test.use = &quot;roc&quot;, only.pos = TRUE)</span><br><span class="line">#和其他细胞簇对比，寻找细胞簇0的标志基因（设置仅寻找上调基因）</span><br><span class="line">#要求基因在细胞簇0中的表达量（log2倍数变化）至少比对照组高0.25</span><br><span class="line">#使用ROC检验，返回值为1则代表该基因可以完美区分细胞簇0和其他细胞簇；返回值为0则代表该基因完全不能区分</span><br></pre></td></tr></table></figure>

<p>我们可以用<code>VinPlot()</code>函数和<code>FeaturePlot()</code>函数来使得差异基因可视化。其中前者可以显示某个或某些基因在不同细胞集群的表达概率分布，后者可以在PCA或UMAP图上可视化表达特征基因。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">VlnPlot(pbmc, features = c(&quot;MS4A1&quot;, &quot;CD79A&quot;))</span><br></pre></td></tr></table></figure>
<p><img src="/images/PBMC3k/dbca2fef-f403-4e7d-941f-d9a5e771f143.png"></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">VlnPlot(pbmc, features = c(&quot;NKG7&quot;, &quot;PF4&quot;), slot = &quot;counts&quot;, log = TRUE)</span><br><span class="line"># 使用未标准化的原始数据的对数进行可视化……</span><br></pre></td></tr></table></figure>
<p><img src="/images/PBMC3k/0b9bd5f6-28ba-4c58-88a0-f83f03a568c9.png"></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">FeaturePlot(pbmc, features = c(&quot;MS4A1&quot;, &quot;GNLY&quot;, &quot;CD3E&quot;, &quot;CD14&quot;, &quot;FCER1A&quot;, &quot;FCGR3A&quot;, &quot;LYZ&quot;, &quot;PPBP&quot;, &quot;CD8A&quot;))</span><br></pre></td></tr></table></figure>
<p><img src="/images/PBMC3k/9cbb9020-3496-4b48-9113-2cbf5027418c.png"></p>
<p>通过<code>DoHeatmap()</code>函数，我们可以生成一个不同细胞簇标志基因在所有细胞簇表达水平差异的热图。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">pbmc.markers %&gt;%</span><br><span class="line">  group_by(cluster) %&gt;%</span><br><span class="line">  dplyr::filter(avg_log2FC &gt; 1) %&gt;%	#筛选log2FC&gt;1的基因</span><br><span class="line">  slice_head(n = 10) %&gt;%            #筛选每个细胞集群前10个特征基因</span><br><span class="line">  ungroup() -&gt; top10</span><br><span class="line">DoHeatmap(pbmc, features = top10$gene) + NoLegend()</span><br></pre></td></tr></table></figure>
<p><img src="/images/PBMC3k/7f306674-ee41-499f-a50c-a15af612ab0e.png"></p>
<h1 id="注释细胞类型"><a href="#注释细胞类型" class="headerlink" title="注释细胞类型"></a>注释细胞类型</h1><p>在上一步中，我们发掘出了每一个细胞簇的标志基因。基于现有的研究成果，我们可以将标志基因和细胞类型对应起来（如下表），然后给我们的图象进行标注。</p>
<table>
<thead>
<tr>
<th align="center">细胞簇</th>
<th align="center">标志基因</th>
<th align="center">细胞类型</th>
</tr>
</thead>
<tbody><tr>
<td align="center">0</td>
<td align="center">IL7R, CCR7</td>
<td align="center">Naive CD4+ T</td>
</tr>
<tr>
<td align="center">1</td>
<td align="center">CD14, LYZ</td>
<td align="center">CD14+ Mono</td>
</tr>
<tr>
<td align="center">2</td>
<td align="center">IL7R, S100A4</td>
<td align="center">Memory CD4+</td>
</tr>
<tr>
<td align="center">3</td>
<td align="center">MS4A1</td>
<td align="center">B</td>
</tr>
<tr>
<td align="center">4</td>
<td align="center">CD8A</td>
<td align="center">CD8+ T</td>
</tr>
<tr>
<td align="center">5</td>
<td align="center">FCGR3A, MS4A7</td>
<td align="center">FCGR3A+ Mono</td>
</tr>
<tr>
<td align="center">6</td>
<td align="center">GNLY, NKG7</td>
<td align="center">NK</td>
</tr>
<tr>
<td align="center">7</td>
<td align="center">FCER1A, CST3</td>
<td align="center">DC</td>
</tr>
<tr>
<td align="center">8</td>
<td align="center">PPBP</td>
<td align="center">Platelet</td>
</tr>
</tbody></table>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">new.cluster.ids &lt;- c(&quot;Naive CD4 T&quot;, &quot;CD14+ Mono&quot;, &quot;Memory CD4 T&quot;, &quot;B&quot;, &quot;CD8 T&quot;, &quot;FCGR3A+ Mono&quot;, &quot;NK&quot;, &quot;DC&quot;, &quot;Platelet&quot;)</span><br><span class="line">names(new.cluster.ids) &lt;- levels(pbmc)</span><br><span class="line">pbmc &lt;- RenameIdents(pbmc, new.cluster.ids)</span><br><span class="line">DimPlot(pbmc, reduction = &quot;umap&quot;, label = TRUE, pt.size = 0.5) + NoLegend()</span><br></pre></td></tr></table></figure>
<p><img src="/images/PBMC3k/fc85338f-8024-45c6-8874-afd678b8e825.png"></p>
<p>导出最终的图片：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">#install.packages(&quot;ggplot2&quot;)</span><br><span class="line">library(ggplot2)</span><br><span class="line">plot &lt;- DimPlot(pbmc, reduction = &quot;umap&quot;, label = TRUE, label.size = 4.5) + xlab(&quot;UMAP 1&quot;) + ylab(&quot;UMAP 2&quot;) + theme(axis.title = element_text(size = 18), legend.text = element_text(size = 18)) + guides(colour = guide_legend(override.aes = list(size = 10)))</span><br><span class="line">#检查output/images目录是否存在，若不存在则创建</span><br><span class="line">if (!dir.exists(&quot;C:/Users/86150/Downloads/pbmc3k_filtered_gene_bc_matrices/output/images&quot;)) &#123;</span><br><span class="line">  dir.create(&quot;C:/Users/86150/Downloads/pbmc3k_filtered_gene_bc_matrices/output/images&quot;)</span><br><span class="line">&#125;</span><br><span class="line">ggsave(filename = &quot;C:/Users/86150/Downloads/pbmc3k_filtered_gene_bc_matrices/output/images/pbmc3k_umap.jpg&quot;, height = 7, width = 12, plot = plot, quality = 50)</span><br></pre></td></tr></table></figure>

<p>保存运算结果：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">saveRDS(pbmc, file = &quot;C:/Users/86150/Downloads/pbmc3k_filtered_gene_bc_matrices/output/pbmc3k_final.rds&quot;)</span><br></pre></td></tr></table></figure>

<p>下班收工！！</p>
<h1 id="完整代码"><a href="#完整代码" class="headerlink" title="完整代码"></a>完整代码</h1><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br></pre></td><td class="code"><pre><span class="line">#install.packages(&quot;dplyr&quot;)</span><br><span class="line">#install.packages(&quot;Seurat&quot;)</span><br><span class="line">#install.packages(&quot;patchwork&quot;)</span><br><span class="line"></span><br><span class="line">library(dplyr)</span><br><span class="line">library(Seurat)</span><br><span class="line">library(patchwork)</span><br><span class="line"></span><br><span class="line">pbmc.data &lt;- Read10X(data.dir = &quot;C:/Users/86150/Downloads/pbmc3k_filtered_gene_bc_matrices/filtered_gene_bc_matrices/hg19&quot;)</span><br><span class="line"></span><br><span class="line">pbmc &lt;- CreateSeuratObject(counts = pbmc.data, project = &quot;pbmc3k&quot;, min.cells = 3,min.features = 200)</span><br><span class="line">pbmc</span><br><span class="line"></span><br><span class="line">#计算线粒体基因比例</span><br><span class="line">pbmc[[&quot;percent.mt&quot;]] &lt;- PercentageFeatureSet(pbmc, pattern = &quot;^MT-&quot;)</span><br><span class="line">#可视化</span><br><span class="line">VlnPlot(pbmc, features = c(&quot;nFeature_RNA&quot;, &quot;nCount_RNA&quot;, &quot;percent.mt&quot;), ncol =3)</span><br><span class="line"></span><br><span class="line">plot1 &lt;- FeatureScatter(pbmc, feature1 = &quot;nCount_RNA&quot;, feature2 = &quot;percent.mt&quot;)</span><br><span class="line">plot2 &lt;- FeatureScatter(pbmc, feature1 = &quot;nCount_RNA&quot;, feature2 = &quot;nFeature_RNA&quot;)</span><br><span class="line">plot1 + plot2</span><br><span class="line"></span><br><span class="line">pbmc &lt;- subset(pbmc, subset = nFeature_RNA &gt; 200 &amp; nFeature_RNA &lt; 2500 &amp; percent.mt &lt; 5)</span><br><span class="line"></span><br><span class="line">pbmc &lt;- NormalizeData(pbmc, normalization.method = &quot;LogNormalize&quot;, scale.factor = 10000)</span><br><span class="line">#pbmc &lt;- NormalizeData(pbmc) #简写版</span><br><span class="line"></span><br><span class="line">pbmc &lt;- FindVariableFeatures(pbmc, selection.method = &quot;vst&quot;, nfeatures = 2000)</span><br><span class="line"></span><br><span class="line">#确定前10个高变基因</span><br><span class="line">top10 &lt;- head(VariableFeatures(pbmc), 10)</span><br><span class="line"></span><br><span class="line"># 可视化</span><br><span class="line">plot1 &lt;- VariableFeaturePlot(pbmc)</span><br><span class="line">plot2 &lt;- LabelPoints(plot = plot1, points = top10, repel = TRUE)</span><br><span class="line">plot1 + plot2</span><br><span class="line"></span><br><span class="line">all.genes &lt;- rownames(pbmc)</span><br><span class="line">pbmc &lt;- ScaleData(pbmc, features = all.genes)</span><br><span class="line"></span><br><span class="line">pbmc &lt;- RunPCA(pbmc, features = VariableFeatures(object = pbmc))</span><br><span class="line"></span><br><span class="line">#检验并可视化PCA</span><br><span class="line">print(pbmc[[&quot;pca&quot;]], dims = 1:5, nfeatures = 5)</span><br><span class="line"></span><br><span class="line">VizDimLoadings(pbmc, dims = 1:2, reduction = &quot;pca&quot;)</span><br><span class="line">DimPlot(pbmc, reduction = &quot;pca&quot;) + NoLegend()</span><br><span class="line">DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)</span><br><span class="line">DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)</span><br><span class="line"></span><br><span class="line">ElbowPlot(pbmc)</span><br><span class="line">pbmc &lt;- JackStraw(pbmc, num.replicate = 100)</span><br><span class="line">pbmc &lt;- ScoreJackStraw(pbmc, dims = 1:20)</span><br><span class="line">JackStrawPlot(pbmc, dims = 1:15)</span><br><span class="line"></span><br><span class="line">pbmc &lt;- FindNeighbors(pbmc, dims = 1:10)</span><br><span class="line">pbmc &lt;- FindClusters(pbmc, resolution = 0.5)</span><br><span class="line">#获取前5个集群的集落ID</span><br><span class="line">head(Idents(pbmc), 5)</span><br><span class="line"></span><br><span class="line">pbmc &lt;- RunUMAP(pbmc, dims = 1:10)</span><br><span class="line">DimPlot(pbmc, reduction = &quot;umap&quot;)</span><br><span class="line"></span><br><span class="line">#检查output目录是否存在，若不存在则创建</span><br><span class="line">if (!dir.exists(&quot;C:/Users/86150/Downloads/pbmc3k_filtered_gene_bc_matrices/output&quot;)) &#123;</span><br><span class="line">  dir.create(&quot;C:/Users/86150/Downloads/pbmc3k_filtered_gene_bc_matrices/output&quot;)</span><br><span class="line">&#125;</span><br><span class="line">saveRDS(pbmc, file = &quot;C:/Users/86150/Downloads/pbmc3k_filtered_gene_bc_matrices/output/pbmc_tutorial.rds&quot;)</span><br><span class="line"></span><br><span class="line">pbmc_restored &lt;- readRDS(&quot;C:/Users/86150/Downloads/pbmc3k_filtered_gene_bc_matrices/output/pbmc_tutorial.rds&quot;) #读取RDS文件</span><br><span class="line">DimPlot(pbmc_restored, reduction = &quot;umap&quot;)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">cluster2.markers &lt;- FindMarkers(pbmc, ident.1 = 2)</span><br><span class="line">head(cluster2.markers, n = 5)</span><br><span class="line"></span><br><span class="line">cluster5.markers &lt;- FindMarkers(pbmc, ident.1 = 5, ident.2 = c(0, 3))</span><br><span class="line">head(cluster5.markers, n = 5)</span><br><span class="line"></span><br><span class="line">#和其他细胞簇对比，寻找每一个细胞簇的标志基因（设置仅寻找上调基因）</span><br><span class="line">pbmc.markers &lt;- FindAllMarkers(pbmc, only.pos = TRUE)</span><br><span class="line">pbmc.markers %&gt;%</span><br><span class="line">  group_by(cluster) %&gt;%</span><br><span class="line">  dplyr::filter(avg_log2FC &gt; 1)</span><br><span class="line"></span><br><span class="line">cluster0.markers &lt;- FindMarkers(pbmc, ident.1 = 0, logfc.threshold = 0.25, test.use = &quot;roc&quot;, only.pos = TRUE)</span><br><span class="line">#和其他细胞簇对比，寻找细胞簇0的标志基因（设置仅寻找上调基因）</span><br><span class="line">#要求基因在细胞簇0中的表达量（log2倍数变化）至少比对照组高0.25</span><br><span class="line">#使用ROC检验，返回值为1则代表该基因可以完美区分细胞簇0和其他细胞簇；返回值为0则代表该基因完全不能区分</span><br><span class="line"></span><br><span class="line">VlnPlot(pbmc, features = c(&quot;MS4A1&quot;, &quot;CD79A&quot;))</span><br><span class="line"></span><br><span class="line">VlnPlot(pbmc, features = c(&quot;NKG7&quot;, &quot;PF4&quot;), slot = &quot;counts&quot;, log = TRUE)</span><br><span class="line"># 使用未标准化的原始数据的对数进行可视化</span><br><span class="line"></span><br><span class="line">FeaturePlot(pbmc, features = c(&quot;MS4A1&quot;, &quot;GNLY&quot;, &quot;CD3E&quot;, &quot;CD14&quot;, &quot;FCER1A&quot;, &quot;FCGR3A&quot;, &quot;LYZ&quot;, &quot;PPBP&quot;, &quot;CD8A&quot;))</span><br><span class="line"></span><br><span class="line">pbmc.markers %&gt;%</span><br><span class="line">  group_by(cluster) %&gt;%</span><br><span class="line">  dplyr::filter(avg_log2FC &gt; 1) %&gt;% #筛选log2FC&gt;1的基因</span><br><span class="line">  slice_head(n = 10) %&gt;%            #筛选每个细胞集群前10个特征基因</span><br><span class="line">  ungroup() -&gt; top10</span><br><span class="line">DoHeatmap(pbmc, features = top10$gene) + NoLegend()</span><br><span class="line"></span><br><span class="line">new.cluster.ids &lt;- c(&quot;Naive CD4 T&quot;, &quot;CD14+ Mono&quot;, &quot;Memory CD4 T&quot;, &quot;B&quot;, &quot;CD8 T&quot;, &quot;FCGR3A+ Mono&quot;, &quot;NK&quot;, &quot;DC&quot;, &quot;Platelet&quot;)</span><br><span class="line">names(new.cluster.ids) &lt;- levels(pbmc)</span><br><span class="line">pbmc &lt;- RenameIdents(pbmc, new.cluster.ids)</span><br><span class="line">DimPlot(pbmc, reduction = &quot;umap&quot;, label = TRUE, pt.size = 0.5) + NoLegend()</span><br><span class="line"></span><br><span class="line">#install.packages(&quot;ggplot2&quot;)</span><br><span class="line">library(ggplot2)</span><br><span class="line">plot &lt;- DimPlot(pbmc, reduction = &quot;umap&quot;, label = TRUE, label.size = 4.5) + xlab(&quot;UMAP 1&quot;) + ylab(&quot;UMAP 2&quot;) + theme(axis.title = element_text(size = 18), legend.text = element_text(size = 18)) + guides(colour = guide_legend(override.aes = list(size = 10)))</span><br><span class="line">#检查output/images目录是否存在，若不存在则创建</span><br><span class="line">if (!dir.exists(&quot;C:/Users/86150/Downloads/pbmc3k_filtered_gene_bc_matrices/output/images&quot;)) &#123;</span><br><span class="line">  dir.create(&quot;C:/Users/86150/Downloads/pbmc3k_filtered_gene_bc_matrices/output/images&quot;)</span><br><span class="line">&#125;</span><br><span class="line">ggsave(filename = &quot;C:/Users/86150/Downloads/pbmc3k_filtered_gene_bc_matrices/output/images/pbmc3k_umap.jpg&quot;, height = 7, width = 12, plot = plot, quality = 50)</span><br><span class="line"></span><br><span class="line">saveRDS(pbmc, file = &quot;C:/Users/86150/Downloads/pbmc3k_filtered_gene_bc_matrices/output/pbmc3k_final.rds&quot;)</span><br></pre></td></tr></table></figure></article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta"><i class="fas fa-circle-user fa-fw"></i>文章作者: </span><span class="post-copyright-info"><a href="http://example.com">Nahida</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta"><i class="fas fa-square-arrow-up-right fa-fw"></i>文章链接: </span><span class="post-copyright-info"><a href="http://example.com/2025/03/28/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9ASeurat%E7%9A%84%E5%B8%B8%E8%A7%81%E5%88%86%E6%9E%90%E5%B7%A5%E4%BD%9C%E6%B5%81%E7%A8%8B%EF%BC%9A%E4%BB%A5PBMC-3K%E6%95%B0%E6%8D%AE%E9%9B%86%E4%B8%BA%E4%BE%8B/">http://example.com/2025/03/28/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9ASeurat%E7%9A%84%E5%B8%B8%E8%A7%81%E5%88%86%E6%9E%90%E5%B7%A5%E4%BD%9C%E6%B5%81%E7%A8%8B%EF%BC%9A%E4%BB%A5PBMC-3K%E6%95%B0%E6%8D%AE%E9%9B%86%E4%B8%BA%E4%BE%8B/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta"><i class="fas fa-circle-exclamation fa-fw"></i>版权声明: </span><span class="post-copyright-info">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0</a> 许可协议。转载请注明来源 <a href="http://example.com" target="_blank">Nahida的净善宫</a>！</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/">生信分析</a><a class="post-meta__tags" href="/tags/R%E8%AF%AD%E8%A8%80/">R语言</a><a class="post-meta__tags" href="/tags/Seurat/">Seurat</a></div><div class="post-share"><div class="social-share" data-image="/images/Seurat.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><a class="pagination-related" href="/2025/03/24/%E3%80%90%E5%8C%BB%E5%AD%A6%E5%BD%B1%E5%83%8F%E5%AD%A6%E3%80%91%E5%BE%AA%E7%8E%AF%E7%B3%BB%E7%BB%9F%E5%B8%B8%E8%A7%81%E7%96%BE%E7%97%85/" title="【医学影像学】循环系统常见疾病"><div class="cover" style="background: var(--default-bg-color)"></div><div class="info"><div class="info-1"><div class="info-item-1">上一篇</div><div class="info-item-2">【医学影像学】循环系统常见疾病</div></div><div class="info-2"><div class="info-item-1">...</div></div></div></a><a class="pagination-related" href="/2025/03/28/%E3%80%90%E5%8C%BB%E5%AD%A6%E5%BD%B1%E5%83%8F%E5%AD%A6%E3%80%91%E6%B6%88%E5%8C%96%E7%B3%BB%E7%BB%9F%E6%A3%80%E6%9F%A5%E6%96%B9%E6%B3%95%E3%80%81%E6%AD%A3%E5%B8%B8%E8%A1%A8%E7%8E%B0%E5%92%8C%E5%9F%BA%E6%9C%AC%E7%97%85%E5%BE%81/" title="【医学影像学】消化系统检查方法、正常表现和基本病征"><div class="cover" style="background: var(--default-bg-color)"></div><div class="info text-right"><div class="info-1"><div class="info-item-1">下一篇</div><div class="info-item-2">【医学影像学】消化系统检查方法、正常表现和基本病征</div></div><div class="info-2"><div class="info-item-1">学习要求 掌握：食管和胃肠道有关的正常影像解剖 掌握：食管和胃肠道的基本异常影像学征象 了解：常见胃肠道疾病的影像学检查方法的比较和选择  消化道影像学检查方法X线 造影检查按检查方法及范围分类 口服钡餐造影（oral barium opacification）： 食管吞钡检查（至贲门）、上消化道钡剂检查、小肠钡剂造影   钡剂灌肠造影（barium enema） ：结肠造影 空气灌肠造影：肠套叠 经内镜逆行胆胰管造影 （Endoscopic Retrograde Cholangio-Pancreatography, ERCP）：胆道病变  引入对比剂的种类 传统钡剂造影法: 钡剂单对比造影...</div></div></div></a></nav><div class="relatedPosts"><div class="headline"><i class="fas fa-thumbs-up fa-fw"></i><span>相关推荐</span></div><div class="relatedPosts-list"><a class="pagination-related" href="/2025/07/10/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E5%85%B3%E4%BA%8EAzimuth%E5%8C%85%E6%97%A0%E6%B3%95%E6%AD%A3%E5%B8%B8%E4%B8%8B%E8%BD%BD%E7%9A%84%E8%A7%A3%E5%86%B3%E6%96%B9%E6%B3%95/" title="生信小记：关于Azimuth包无法正常下载的解决方法"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-07-10</div><div class="info-item-2">生信小记：关于Azimuth包无法正常下载的解决方法</div></div><div class="info-2"><div class="info-item-1">问题尚未解决……攻克ing 问题描述运用install.packages(&quot;Azimuth&quot;)安装Azimuth包时，报错如下： 1234567&gt; install.packages(&quot;Azimuth&quot;)Warning in install.packages :  package ‘Azimuth’ is not available for this version of RA version of this package for your version of R might be available elsewhere,see the ideas...</div></div></div></a><a class="pagination-related" href="/2025/07/10/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E5%85%B3%E4%BA%8ESeuratWrappers%E5%8C%85%E6%97%A0%E6%B3%95%E6%AD%A3%E5%B8%B8%E4%B8%8B%E8%BD%BD%E7%9A%84%E8%A7%A3%E5%86%B3%E6%96%B9%E6%B3%95/" title="生信小记：关于SeuratWrappers包无法正常下载的解决方法"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-07-10</div><div class="info-item-2">生信小记：关于SeuratWrappers包无法正常下载的解决方法</div></div><div class="info-2"><div class="info-item-1">问题描述在运行install.packages(&quot;SeuratWrappers&quot;)时，报错如下： 123456Warning in install.packages :  package ‘SeuratWrappers’ is not available for this version of RA version of this package for your version of R might be available elsewhere,see the ideas athttps://cran.r-project.org/doc/manuals/r-patched/R-admin.html#Installing-packages  解决方案下载R包至本地使用电脑浏览器直接访问这个网站，会自动开始下载相应的R包至本地。 安装SeuratWrappers包的所有必需依赖项在Rstudio上运行以下代码： 123install.packages(c(&quot;R.utils&quot;, &quot;rsvd&quot;,...</div></div></div></a><a class="pagination-related" href="/2025/07/09/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9ASeurat-v5%E4%B8%AD%E7%9A%84%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90/" title="生信小记：Seurat_v5中的整合分析"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-07-09</div><div class="info-item-2">生信小记：Seurat_v5中的整合分析</div></div><div class="info-2"><div class="info-item-1">前言单细胞测序数据的整合通常是scRNA-seq测序的重要步骤之一。通过跨实验批次、供体和受体条件的数据的整合，不同数据集之间的共享细胞类型和状态得以匹配，因而大大提高了统计效力……等等，我R包好像下载不了……等我调试完再继续写。 1234567library(Seurat)library(SeuratData)library(SeuratWrappers)library(Azimuth)library(ggplot2)library(patchwork)options(future.globals.maxSize = 1e9) </div></div></div></a><a class="pagination-related" href="/2025/04/29/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%84%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90%E4%BB%8B%E7%BB%8D/" title="生信小记：单细胞转录组数据整合分析介绍"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-04-29</div><div class="info-item-2">生信小记：单细胞转录组数据整合分析介绍</div></div><div class="info-2"><div class="info-item-1">...</div></div></div></a><a class="pagination-related" href="/2025/03/31/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E4%BD%BF%E7%94%A8Seurat%E5%8C%85%E5%A4%84%E7%90%86%E5%A4%9A%E6%A8%A1%E6%80%81%E6%95%B0%E6%8D%AE/" title="生信小记：使用Seurat包处理多模态数据"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-03-31</div><div class="info-item-2">生信小记：使用Seurat包处理多模态数据</div></div><div class="info-2"><div class="info-item-1">在一次实验中，我们可以对同一个细胞测量不同分子层面的数据（如基因表达、蛋白质表达、表观遗传修饰、空间分布等等），所测得的数据我们称之为多模态数据（multimodal data）。例如，CITE-seq可以同时检测单细胞的转录组RNA和表面蛋白表达水平；10X Multiome可以同时测量单细胞的转录组RNA和ATAC（染色质可及性）。本篇文章将以一个包括8617个脐带血单个核细胞（CBMC）的CITE-seq结果（包含转录组RNA测序和表面蛋白丰度估计值）的数据集，介绍如何使用Seurat包对多模态数据进行分析。 导入数据，创建Seurat对象，并加入RNA和表面蛋白数据官方教程中介绍的方法如下： 12345678910111213141516171819202122232425262728library(Seurat)library(ggplot2)library(patchwork)# 注意：这份数据同时存在人和小鼠的细胞，分别冠有&quot;HUMAN_&quot;和&quot;MOUSE_&quot;的前缀。# 导入RNA数据cbmc.rna &lt;-...</div></div></div></a><a class="pagination-related" href="/2025/03/31/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E4%BD%BF%E7%94%A8SCTransform%E8%BF%9B%E8%A1%8C%E6%A0%87%E5%87%86%E5%8C%96/" title="生信小记：使用SCTransform进行标准化"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-03-31</div><div class="info-item-2">生信小记：使用SCTransform进行标准化</div></div><div class="info-2"><div class="info-item-1">...</div></div></div></a></div></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info text-center"><div class="avatar-img"><img src="/img/avatar.jpg" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info-name">Nahida</div><div class="author-info-description">智慧树上智慧果，智慧树下你和我</div><div class="site-data"><a href="/archives/"><div class="headline">文章</div><div class="length-num">83</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">48</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">20</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/xxxxxx"><i class="fab fa-github"></i><span>Follow Me</span></a></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>公告</span></div><div class="announcement_content">暂无公告哦~</div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>目录</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%9B%B4%E6%96%B0%E6%97%A5%E5%BF%97"><span class="toc-number">1.</span> <span class="toc-text">更新日志</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#2025-4-26"><span class="toc-number">1.1.</span> <span class="toc-text">2025&#x2F;4&#x2F;26</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%88%9B%E5%BB%BASeurat%E5%AF%B9%E8%B1%A1"><span class="toc-number">2.</span> <span class="toc-text">创建Seurat对象</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%95%B0%E6%8D%AE%E7%9A%84%E8%AF%BB%E5%8F%96"><span class="toc-number">2.1.</span> <span class="toc-text">数据的读取</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Seurat%E5%AF%B9%E8%B1%A1%E7%9A%84%E5%88%9B%E5%BB%BA"><span class="toc-number">2.2.</span> <span class="toc-text">Seurat对象的创建</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%A0%87%E5%87%86%E9%A2%84%E5%A4%84%E7%90%86%E5%B7%A5%E4%BD%9C%E6%B5%81%E7%A8%8B"><span class="toc-number">3.</span> <span class="toc-text">标准预处理工作流程</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E8%B4%A8%E9%87%8F%E6%8E%A7%E5%88%B6%EF%BC%88QC%EF%BC%89%E5%92%8C%E9%80%89%E6%8B%A9%E7%BB%86%E8%83%9E%E8%BF%9B%E8%A1%8C%E8%BF%9B%E4%B8%80%E6%AD%A5%E5%88%86%E6%9E%90"><span class="toc-number">3.1.</span> <span class="toc-text">质量控制（QC）和选择细胞进行进一步分析</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%95%B0%E6%8D%AE%E6%A0%87%E5%87%86%E5%8C%96"><span class="toc-number">4.</span> <span class="toc-text">数据标准化</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E8%AF%86%E5%88%AB%E9%AB%98%E5%8F%98%E5%9F%BA%E5%9B%A0"><span class="toc-number">5.</span> <span class="toc-text">识别高变基因</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%95%B0%E6%8D%AE%E7%BC%A9%E6%94%BE%EF%BC%88Z-score%E6%A0%87%E5%87%86%E5%8C%96%EF%BC%89"><span class="toc-number">6.</span> <span class="toc-text">数据缩放（Z-score标准化）</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%89%A7%E8%A1%8C%E7%BA%BF%E6%80%A7%E9%99%8D%E7%BB%B4"><span class="toc-number">7.</span> <span class="toc-text">执行线性降维</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E7%A1%AE%E8%AE%A4%E6%95%B0%E6%8D%AE%E9%9B%86%E7%9A%84%E7%BB%B4%E5%BA%A6"><span class="toc-number">8.</span> <span class="toc-text">确认数据集的维度</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E7%BB%86%E8%83%9E%E8%81%9A%E7%B1%BB"><span class="toc-number">9.</span> <span class="toc-text">细胞聚类</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%89%A7%E8%A1%8C%E9%9D%9E%E7%BA%BF%E6%80%A7%E9%99%8D%E7%BB%B4%EF%BC%88UMAP-t-SNE%EF%BC%89"><span class="toc-number">10.</span> <span class="toc-text">执行非线性降维（UMAP&#x2F;$t$-SNE）</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E9%89%B4%E5%AE%9A%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%9F%BA%E5%9B%A0%EF%BC%88DEGs%EF%BC%89"><span class="toc-number">11.</span> <span class="toc-text">鉴定差异表达基因（DEGs）</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%B3%A8%E9%87%8A%E7%BB%86%E8%83%9E%E7%B1%BB%E5%9E%8B"><span class="toc-number">12.</span> <span class="toc-text">注释细胞类型</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%AE%8C%E6%95%B4%E4%BB%A3%E7%A0%81"><span class="toc-number">13.</span> <span class="toc-text">完整代码</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>最新文章</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2025/07/14/%E3%80%90%E6%96%AF%E6%B3%A2%E7%B4%A2%E5%AE%BE%E5%92%8C%E5%A3%B0%E5%AD%A6%E3%80%91%E5%9B%9B%E9%83%A8%E5%92%8C%E5%A3%B0%E5%8F%8A%E5%85%B6%E7%9B%B8%E5%85%B3%E6%A6%82%E5%BF%B5/" title="【斯波索宾和声学】四部和声及其相关概念"><img src="/images/%E5%92%8C%E5%A3%B0.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="【斯波索宾和声学】四部和声及其相关概念"/></a><div class="content"><a class="title" href="/2025/07/14/%E3%80%90%E6%96%AF%E6%B3%A2%E7%B4%A2%E5%AE%BE%E5%92%8C%E5%A3%B0%E5%AD%A6%E3%80%91%E5%9B%9B%E9%83%A8%E5%92%8C%E5%A3%B0%E5%8F%8A%E5%85%B6%E7%9B%B8%E5%85%B3%E6%A6%82%E5%BF%B5/" title="【斯波索宾和声学】四部和声及其相关概念">【斯波索宾和声学】四部和声及其相关概念</a><time datetime="2025-07-14T15:09:21.000Z" title="发表于 2025-07-14 23:09:21">2025-07-14</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2025/07/14/%E3%80%90%E4%B9%90%E5%99%A8%E6%B3%95%E4%B8%8E%E9%85%8D%E5%99%A8%E6%B3%95%E3%80%91%E5%B0%8F%E5%9E%8B%E4%B9%90%E9%98%9F%E7%9A%84%E7%BC%96%E9%85%8D/" title="【乐器法与配器法】小型乐队的编配"><img src="/images/%E4%B9%90%E5%99%A8%E6%B3%95%E4%B8%8E%E9%85%8D%E5%99%A8%E6%B3%95.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="【乐器法与配器法】小型乐队的编配"/></a><div class="content"><a class="title" href="/2025/07/14/%E3%80%90%E4%B9%90%E5%99%A8%E6%B3%95%E4%B8%8E%E9%85%8D%E5%99%A8%E6%B3%95%E3%80%91%E5%B0%8F%E5%9E%8B%E4%B9%90%E9%98%9F%E7%9A%84%E7%BC%96%E9%85%8D/" title="【乐器法与配器法】小型乐队的编配">【乐器法与配器法】小型乐队的编配</a><time datetime="2025-07-14T05:33:04.000Z" title="发表于 2025-07-14 13:33:04">2025-07-14</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2025/07/14/%E3%80%90%E4%B9%90%E5%99%A8%E6%B3%95%E4%B8%8E%E9%85%8D%E5%99%A8%E6%B3%95%E3%80%91%E9%93%9C%E7%AE%A1%E7%BB%84/" title="【乐器法与配器法】铜管组"><img src="/images/%E4%B9%90%E5%99%A8%E6%B3%95%E4%B8%8E%E9%85%8D%E5%99%A8%E6%B3%95.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="【乐器法与配器法】铜管组"/></a><div class="content"><a class="title" href="/2025/07/14/%E3%80%90%E4%B9%90%E5%99%A8%E6%B3%95%E4%B8%8E%E9%85%8D%E5%99%A8%E6%B3%95%E3%80%91%E9%93%9C%E7%AE%A1%E7%BB%84/" title="【乐器法与配器法】铜管组">【乐器法与配器法】铜管组</a><time datetime="2025-07-14T03:40:06.000Z" title="发表于 2025-07-14 11:40:06">2025-07-14</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2025/07/13/%E3%80%90%E4%B9%90%E5%99%A8%E6%B3%95%E4%B8%8E%E9%85%8D%E5%99%A8%E6%B3%95%E3%80%91%E6%9C%A8%E7%AE%A1%E7%BB%84/" title="【乐器法与配器法】木管组"><img src="/images/%E4%B9%90%E5%99%A8%E6%B3%95%E4%B8%8E%E9%85%8D%E5%99%A8%E6%B3%95.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="【乐器法与配器法】木管组"/></a><div class="content"><a class="title" href="/2025/07/13/%E3%80%90%E4%B9%90%E5%99%A8%E6%B3%95%E4%B8%8E%E9%85%8D%E5%99%A8%E6%B3%95%E3%80%91%E6%9C%A8%E7%AE%A1%E7%BB%84/" title="【乐器法与配器法】木管组">【乐器法与配器法】木管组</a><time datetime="2025-07-13T05:41:03.000Z" title="发表于 2025-07-13 13:41:03">2025-07-13</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2025/07/11/%E3%80%90%E4%B9%90%E5%99%A8%E6%B3%95%E4%B8%8E%E9%85%8D%E5%99%A8%E6%B3%95%E3%80%91%E5%BC%A6%E4%B9%90%E7%BB%84/" title="【乐器法与配器法】弦乐组"><img src="/images/%E4%B9%90%E5%99%A8%E6%B3%95%E4%B8%8E%E9%85%8D%E5%99%A8%E6%B3%95.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="【乐器法与配器法】弦乐组"/></a><div class="content"><a class="title" href="/2025/07/11/%E3%80%90%E4%B9%90%E5%99%A8%E6%B3%95%E4%B8%8E%E9%85%8D%E5%99%A8%E6%B3%95%E3%80%91%E5%BC%A6%E4%B9%90%E7%BB%84/" title="【乐器法与配器法】弦乐组">【乐器法与配器法】弦乐组</a><time datetime="2025-07-11T03:57:22.000Z" title="发表于 2025-07-11 11:57:22">2025-07-11</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2019 - 2025 By Nahida</div><div class="framework-info"><span>框架 </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo 7.3.0</a><span class="footer-separator">|</span><span>主题 </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly 5.3.5</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="阅读模式"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="日间和夜间模式切换"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="单栏和双栏切换"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside-config" type="button" title="设置"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="目录"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="回到顶部"><span class="scroll-percent"></span><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><div class="js-pjax"><script>(() => {
  const loadMathjax = () => {
    if (!window.MathJax) {
      window.MathJax = {
        tex: {
          inlineMath: [['$', '$'], ['\\(', '\\)']],
          tags: 'none',
        },
        chtml: {
          scale: 1.1
        },
        options: {
          enableMenu: true,
          renderActions: {
            findScript: [10, doc => {
              for (const node of document.querySelectorAll('script[type^="math/tex"]')) {
                const display = !!node.type.match(/; *mode=display/)
                const math = new doc.options.MathItem(node.textContent, doc.inputJax[0], display)
                const text = document.createTextNode('')
                node.parentNode.replaceChild(text, node)
                math.start = {node: text, delim: '', n: 0}
                math.end = {node: text, delim: '', n: 0}
                doc.math.push(math)
              }
            }, '']
          }
        }
      }

      const script = document.createElement('script')
      script.src = 'https://cdn.jsdelivr.net/npm/mathjax/es5/tex-mml-chtml.min.js'
      script.id = 'MathJax-script'
      script.async = true
      document.head.appendChild(script)
    } else {
      MathJax.startup.document.state(0)
      MathJax.texReset()
      MathJax.typesetPromise()
    }
  }

  btf.addGlobalFn('encrypt', loadMathjax, 'mathjax')
  window.pjax ? loadMathjax() : window.addEventListener('load', loadMathjax)
})()</script></div><script type="text/javascript" src="/js/memos/waterfall.min.js"></script><script type="text/javascript" src="/js/memos/imgStatus.min.js"></script><script type="text/javascript" src="/js/memos/lately.min.js"></script><script type="text/javascript" src="/js/memos/photo.js"></script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>