<!DOCTYPE html><html lang="zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0,viewport-fit=cover"><title>生信小记：单细胞转录组数据整合分析介绍 | Nahida的净善宫</title><meta name="author" content="Nahida"><meta name="copyright" content="Nahida"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="整合多组单细胞测序数据集（如跨实验批次、供体或条件），是scRNA-seq的重要步骤之一。整合分析可以匹配不同细胞集之间相同的细胞类型和细胞状态，提高统计效能，有利于跨数据集的准确比较分析。本篇文章将对利用Seurat集成分析进行复杂细胞类型间的比较分析进行简要概述。我们进行比较分析的主要目的主要在于以下三点：  确认在不同细胞数据集中均存在的细胞簇； 确认在实验组（经过某种特定刺激）和对照组中保">
<meta property="og:type" content="article">
<meta property="og:title" content="生信小记：单细胞转录组数据整合分析介绍">
<meta property="og:url" content="http://example.com/2025/04/29/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%84%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90%E4%BB%8B%E7%BB%8D/index.html">
<meta property="og:site_name" content="Nahida的净善宫">
<meta property="og:description" content="整合多组单细胞测序数据集（如跨实验批次、供体或条件），是scRNA-seq的重要步骤之一。整合分析可以匹配不同细胞集之间相同的细胞类型和细胞状态，提高统计效能，有利于跨数据集的准确比较分析。本篇文章将对利用Seurat集成分析进行复杂细胞类型间的比较分析进行简要概述。我们进行比较分析的主要目的主要在于以下三点：  确认在不同细胞数据集中均存在的细胞簇； 确认在实验组（经过某种特定刺激）和对照组中保">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="http://example.com/images/Seurat.png">
<meta property="article:published_time" content="2025-04-29T12:17:29.000Z">
<meta property="article:modified_time" content="2025-07-10T07:29:40.508Z">
<meta property="article:author" content="Nahida">
<meta property="article:tag" content="生信分析">
<meta property="article:tag" content="R语言">
<meta property="article:tag" content="Seurat">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://example.com/images/Seurat.png"><script type="application/ld+json">{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "生信小记：单细胞转录组数据整合分析介绍",
  "url": "http://example.com/2025/04/29/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%84%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90%E4%BB%8B%E7%BB%8D/",
  "image": "http://example.com/images/Seurat.png",
  "datePublished": "2025-04-29T12:17:29.000Z",
  "dateModified": "2025-07-10T07:29:40.508Z",
  "author": [
    {
      "@type": "Person",
      "name": "Nahida",
      "url": "http://example.com/"
    }
  ]
}</script><link rel="shortcut icon" href="/img/avatar.jpg"><link rel="canonical" href="http://example.com/2025/04/29/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%84%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90%E4%BB%8B%E7%BB%8D/index.html"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css"><script>
    (() => {
      
    const saveToLocal = {
      set: (key, value, ttl) => {
        if (!ttl) return
        const expiry = Date.now() + ttl * 86400000
        localStorage.setItem(key, JSON.stringify({ value, expiry }))
      },
      get: key => {
        const itemStr = localStorage.getItem(key)
        if (!itemStr) return undefined
        const { value, expiry } = JSON.parse(itemStr)
        if (Date.now() > expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return value
      }
    }

    window.btf = {
      saveToLocal,
      getScript: (url, attr = {}) => new Promise((resolve, reject) => {
        const script = document.createElement('script')
        script.src = url
        script.async = true
        Object.entries(attr).forEach(([key, val]) => script.setAttribute(key, val))
        script.onload = script.onreadystatechange = () => {
          if (!script.readyState || /loaded|complete/.test(script.readyState)) resolve()
        }
        script.onerror = reject
        document.head.appendChild(script)
      }),
      getCSS: (url, id) => new Promise((resolve, reject) => {
        const link = document.createElement('link')
        link.rel = 'stylesheet'
        link.href = url
        if (id) link.id = id
        link.onload = link.onreadystatechange = () => {
          if (!link.readyState || /loaded|complete/.test(link.readyState)) resolve()
        }
        link.onerror = reject
        document.head.appendChild(link)
      }),
      addGlobalFn: (key, fn, name = false, parent = window) => {
        if (!false && key.startsWith('pjax')) return
        const globalFn = parent.globalFn || {}
        globalFn[key] = globalFn[key] || {}
        globalFn[key][name || Object.keys(globalFn[key]).length] = fn
        parent.globalFn = globalFn
      }
    }
  
      
      const activateDarkMode = () => {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      const activateLightMode = () => {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }

      btf.activateDarkMode = activateDarkMode
      btf.activateLightMode = activateLightMode

      const theme = saveToLocal.get('theme')
    
          theme === 'dark' ? activateDarkMode() : theme === 'light' ? activateLightMode() : null
        
      
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        document.documentElement.classList.toggle('hide-aside', asideStatus === 'hide')
      }
    
      
    const detectApple = () => {
      if (/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)) {
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
  
    })()
  </script><script>const GLOBAL_CONFIG = {
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  highlight: {"plugin":"highlight.js","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false,"highlightFullpage":false,"highlightMacStyle":false},
  copy: {
    success: '复制成功',
    error: '复制失败',
    noSupport: '浏览器不支持'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  dateSuffix: {
    just: '刚刚',
    min: '分钟前',
    hour: '小时前',
    day: '天前',
    month: '个月前'
  },
  copyright: undefined,
  lightbox: 'null',
  Snackbar: undefined,
  infinitegrid: {
    js: 'https://cdn.jsdelivr.net/npm/@egjs/infinitegrid/dist/infinitegrid.min.js',
    buttonText: '加载更多'
  },
  isPhotoFigcaption: false,
  islazyloadPlugin: false,
  isAnchor: false,
  percent: {
    toc: true,
    rightside: false,
  },
  autoDarkmode: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: '生信小记：单细胞转录组数据整合分析介绍',
  isHighlightShrink: false,
  isToc: true,
  pageType: 'post'
}</script><meta name="generator" content="Hexo 7.3.0"><link rel="alternate" href="/atom.xml" title="Nahida的净善宫" type="application/atom+xml">
</head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img text-center"><img src="/img/avatar.jpg" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="site-data text-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">97</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">58</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">22</div></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> 时间轴</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> 标签</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> 分类</span></a></div><div class="menus_item"><a class="site-page" href="/photos/"><i class="fa-fw fas fa-images"></i><span> 相册</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> 友链</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url(/images/Seurat.png);"><nav id="nav"><span id="blog-info"><a class="nav-site-title" href="/"><img class="site-icon" src="/image/test.gif" alt="Logo"><span class="site-name">Nahida的净善宫</span></a><a class="nav-page-title" href="/"><span class="site-name">生信小记：单细胞转录组数据整合分析介绍</span></a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> 时间轴</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> 标签</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> 分类</span></a></div><div class="menus_item"><a class="site-page" href="/photos/"><i class="fa-fw fas fa-images"></i><span> 相册</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> 友链</span></a></div></div><div id="toggle-menu"><span class="site-page"><i class="fas fa-bars fa-fw"></i></span></div></div></nav><div id="post-info"><h1 class="post-title">生信小记：单细胞转录组数据整合分析介绍</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">发表于</span><time class="post-meta-date-created" datetime="2025-04-29T12:17:29.000Z" title="发表于 2025-04-29 20:17:29">2025-04-29</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">更新于</span><time class="post-meta-date-updated" datetime="2025-07-10T07:29:40.508Z" title="更新于 2025-07-10 15:29:40">2025-07-10</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/">生信分析</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title=""><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">浏览量:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="container post-content" id="article-container"><p>整合多组单细胞测序数据集（如跨实验批次、供体或条件），是scRNA-seq的重要步骤之一。整合分析可以匹配不同细胞集之间相同的细胞类型和细胞状态，提高统计效能，有利于跨数据集的准确比较分析。本篇文章将对利用Seurat集成分析进行复杂细胞类型间的比较分析进行简要概述。我们进行比较分析的主要目的主要在于以下三点：</p>
<ol>
<li>确认在不同细胞数据集中均存在的细胞簇；</li>
<li>确认在实验组（经过某种特定刺激）和对照组中保守的细胞类型标记；</li>
<li>对比并确认细胞对该刺激所作出的特定反应。</li>
</ol>
<h1 id="创建Seurat对象"><a href="#创建Seurat对象" class="headerlink" title="创建Seurat对象"></a>创建Seurat对象</h1><p>首先我们安装并加载所需要的R包，并通过<code>SeuratData</code>包下载本篇文章所需要的原始数据。这份数据包括以下两组细胞：①经干扰素刺激的人PBMC细胞；②对照组人PBMC细胞。在接下来的分析中，整合上述两组细胞数据集，以识别共同的细胞亚群并比较其在不同处理条件下的差异。</p>
<blockquote>
<p>实操过程中，由于网络情况及文件大小等原因，可能会出现因超时而导致下载中止并报错的情况。在安装数据之前运行<code>options(timeout=10000)</code>，将超时限制放宽即可正常下载。</p>
</blockquote>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">library(Seurat)</span><br><span class="line">library(SeuratData)</span><br><span class="line">library(patchwork)</span><br><span class="line">options(timeout=10000)</span><br><span class="line">InstallData(&quot;ifnb&quot;)</span><br></pre></td></tr></table></figure>

<p>接下来我们加载上述原始数据集。由于不同组的细胞数据集均存放于同一个Seurat对象中，我们在分析之前需要先将其分为两层。分层之后我们可以看到CTRL（对照，control）和STIM（刺激，stimulate）两组数据。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">ifnb &lt;- LoadData(&quot;ifnb&quot;)</span><br><span class="line">ifnb[[&quot;RNA&quot;]] &lt;- split(ifnb[[&quot;RNA&quot;]], f = ifnb$stim)</span><br><span class="line">ifnb</span><br></pre></td></tr></table></figure>
<ul>
<li>运行结果</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">&gt; ifnb</span><br><span class="line">An object of class Seurat </span><br><span class="line">14053 features across 13999 samples within 1 assay </span><br><span class="line">Active assay: RNA (14053 features, 0 variable features)</span><br><span class="line"> 4 layers present: counts.CTRL, counts.STIM, data.CTRL, data.STIM</span><br></pre></td></tr></table></figure>
<h1 id="进行不整合的简单分析"><a href="#进行不整合的简单分析" class="headerlink" title="进行不整合的简单分析"></a>进行不整合的简单分析</h1><p>为了更直观地感受分别分析给两组比较带来的挑战，我们先用传统的方法对两组数据分别进行单独分析。从得到的UMAP图上我们很容易看到，两个数据集所分出的细胞簇总数超级多，这是因为两个数据集中每一种类型的细胞都按照数据集的类型被分为两类，而没有整合为一类细胞。这很显然对我们对比分析两组细胞不太有利。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">ifnb &lt;- NormalizeData(ifnb)</span><br><span class="line">ifnb &lt;- FindVariableFeatures(ifnb)</span><br><span class="line">ifnb &lt;- ScaleData(ifnb)</span><br><span class="line">ifnb &lt;- RunPCA(ifnb)</span><br><span class="line">ifnb &lt;- FindNeighbors(ifnb, dims = 1:30, reduction = &quot;pca&quot;)</span><br><span class="line">ifnb &lt;- FindClusters(ifnb, resolution = 2, cluster.name = &quot;unintegrated_clusters&quot;)</span><br><span class="line">ifnb &lt;- RunUMAP(ifnb, dims = 1:30, reduction = &quot;pca&quot;, reduction.name = &quot;umap.unintegrated&quot;)</span><br><span class="line">DimPlot(ifnb, reduction = &quot;umap.unintegrated&quot;, group.by = c(&quot;stim&quot;, &quot;seurat_clusters&quot;))</span><br></pre></td></tr></table></figure>
<p><img src="/images/integration/30220371-ec41-4958-9f33-8847863ec108.png"></p>
<h1 id="整合两组数据集（CCA降维）"><a href="#整合两组数据集（CCA降维）" class="headerlink" title="整合两组数据集（CCA降维）"></a>整合两组数据集（CCA降维）</h1><p>利用<code>IntegrateLayer()</code>函数，可以通过典型相关分析（CCA），识别共享或同源基因来确定共享细胞类型或状态（也就是在不同组别中共享的细胞类型），进而建立跨数据集的关联，从而实现两组数据的整合。这一过程不会消除不同组别细胞之间基因表达的生物学差异，因此不会影响组间分析。整合后的数据将会存放在一个低维嵌入空间（<code>integrated.cca</code>）中。</p>
<p>此时两组数据仍然存储在两层中，因此我们使用<code>JoinLayer()</code>函数合并两组数据，该过程将会覆盖原始分层存储的数据。</p>
<p>随后，我们对CCA降维后的数据进行细胞聚类和UMAP降维。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">ifnb &lt;- IntegrateLayers(object = ifnb, method = CCAIntegration, orig.reduction = &quot;pca&quot;, new.reduction = &quot;integrated.cca&quot;, verbose = FALSE)</span><br><span class="line">ifnb[[&quot;RNA&quot;]] &lt;- JoinLayers(ifnb[[&quot;RNA&quot;]])</span><br><span class="line"></span><br><span class="line">ifnb &lt;- FindNeighbors(ifnb, reduction = &quot;integrated.cca&quot;, dims = 1:30)</span><br><span class="line">ifnb &lt;- FindClusters(ifnb, resolution = 1)</span><br><span class="line">ifnb &lt;- RunUMAP(ifnb, dims = 1:30, reduction = &quot;integrated.cca&quot;)</span><br></pre></td></tr></table></figure>

<p>随后我们可以进行可视化。下图中左图为按照实验条件分组的图像，右图为按照CCA整合后聚类的人工注释结果分类的图像。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">DimPlot(ifnb, reduction = &quot;umap&quot;, group.by = c(&quot;stim&quot;, &quot;seurat_annotations&quot;))</span><br></pre></td></tr></table></figure>
<p><img src="/images/integration/07211de9-c49b-4867-9477-4bb4530e54d7.png"></p>
<p>当然，通过调整<code>split.by</code>参数，我们可以将两组细胞数据分开，分别显示CCA整合后聚类的人工注释结果分类的图像。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">DimPlot(ifnb, reduction = &quot;umap&quot;, split.by = &quot;stim&quot;)</span><br></pre></td></tr></table></figure>
<p><img src="/images/integration/c0bec9e8-fad3-4f45-9a3c-e013afa5859a.png"></p>
<h1 id="识别保守的细胞类型标记"><a href="#识别保守的细胞类型标记" class="headerlink" title="识别保守的细胞类型标记"></a>识别保守的细胞类型标记</h1><p><code>FindConservedMarker()</code>函数，人如其名，可以帮助我们识别在不同处理条件下同一类型细胞中不随处理条件变化的稳定标记基因。其原理为：①在每个实验条件中，分别比较目标细胞群与其他所有细胞的差异表达基因，并计算每个条件的$p$值和表达量；②使用MadeDE包中的​元分析，将不同实验条件的统计值（如$p$值）合并，得到​跨条件统一的显著性评估。</p>
<p>以下述代码为例，我们确认了细胞簇6（即NK细胞）的保守标记基因。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">install.packages(&#x27;BiocManager&#x27;)</span><br><span class="line">BiocManager::install(&#x27;multtest&#x27;)</span><br><span class="line">install.packages(&#x27;metap&#x27;)</span><br><span class="line"></span><br><span class="line">Idents(ifnb) &lt;- &quot;seurat_annotations&quot;</span><br><span class="line">nk.markers &lt;- FindConservedMarkers(ifnb, ident.1 = &quot;NK&quot;, grouping.var = &quot;stim&quot;, verbose = FALSE)</span><br><span class="line">head(nk.markers)</span><br></pre></td></tr></table></figure>
<ul>
<li>运行结果</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">&gt; head(nk.markers)</span><br><span class="line">      CTRL_p_val CTRL_avg_log2FC CTRL_pct.1 CTRL_pct.2 CTRL_p_val_adj</span><br><span class="line">GNLY           0        6.854586      0.943      0.046              0</span><br><span class="line">NKG7           0        5.358881      0.953      0.085              0</span><br><span class="line">GZMB           0        5.078135      0.839      0.044              0</span><br><span class="line">CLIC3          0        5.765314      0.601      0.024              0</span><br><span class="line">CTSW           0        5.307246      0.537      0.030              0</span><br><span class="line">KLRD1          0        5.261553      0.507      0.019              0</span><br><span class="line">      STIM_p_val STIM_avg_log2FC STIM_pct.1 STIM_pct.2 STIM_p_val_adj</span><br><span class="line">GNLY           0        6.435910      0.956      0.059              0</span><br><span class="line">NKG7           0        4.971397      0.950      0.081              0</span><br><span class="line">GZMB           0        5.151924      0.897      0.060              0</span><br><span class="line">CLIC3          0        5.505208      0.623      0.031              0</span><br><span class="line">CTSW           0        5.240729      0.592      0.035              0</span><br><span class="line">KLRD1          0        4.852457      0.555      0.027              0</span><br><span class="line">      max_pval minimump_p_val</span><br><span class="line">GNLY         0              0</span><br><span class="line">NKG7         0              0</span><br><span class="line">GZMB         0              0</span><br><span class="line">CLIC3        0              0</span><br><span class="line">CTSW         0              0</span><br><span class="line">KLRD1        0              0</span><br></pre></td></tr></table></figure>

<p>利用<code>DotPlot()</code>函数，且通过调整<code>split.by</code>参数，我们可以查看不同处理条件下的保守细胞基因，并展示其表达水平和在细胞簇中的表达百分比。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">Idents(ifnb) &lt;- factor(Idents(ifnb), levels = c(&quot;pDC&quot;, &quot;Eryth&quot;, &quot;Mk&quot;, &quot;DC&quot;, &quot;CD14 Mono&quot;, &quot;CD16 Mono&quot;, &quot;B Activated&quot;, &quot;B&quot;, &quot;CD8 T&quot;, &quot;NK&quot;, &quot;T activated&quot;, &quot;CD4 Naive T&quot;, &quot;CD4 Memory T&quot;))</span><br><span class="line">markers.to.plot &lt;- c(&quot;CD3D&quot;, &quot;CREM&quot;, &quot;HSPH1&quot;, &quot;SELL&quot;, &quot;GIMAP5&quot;, &quot;CACYBP&quot;, &quot;GNLY&quot;, &quot;NKG7&quot;, &quot;CCL5&quot;, &quot;CD8A&quot;, &quot;MS4A1&quot;, &quot;CD79A&quot;, &quot;MIR155HG&quot;, &quot;NME1&quot;, &quot;FCGR3A&quot;, &quot;VMO1&quot;, &quot;CCL2&quot;, &quot;S100A9&quot;, &quot;HLA-DQA1&quot;, &quot;GPR183&quot;, &quot;PPBP&quot;, &quot;GNG11&quot;, &quot;HBA2&quot;, &quot;HBB&quot;, &quot;TSPAN13&quot;, &quot;IL3RA&quot;, &quot;IGJ&quot;, &quot;PRSS57&quot;)</span><br><span class="line">DotPlot(ifnb, features = markers.to.plot, cols = c(&quot;blue&quot;, &quot;red&quot;), dot.scale = 8, split.by = &quot;stim&quot;) + RotatedAxis()</span><br></pre></td></tr></table></figure>
<p><img src="/images/integration/ee507161-e244-4d6b-b28e-6ad0e20f7091.png"></p>
<h1 id="识别不同处理条件下细胞的差异表达基因"><a href="#识别不同处理条件下细胞的差异表达基因" class="headerlink" title="识别不同处理条件下细胞的差异表达基因"></a>识别不同处理条件下细胞的差异表达基因</h1><p>这一步我们将分析出在不同处理条件下，在同一细胞类型中表达存在显著差异的基因。但在这之前，我们需要把现在的数据转化为“伪批量”数据。这是因为，在单细胞测序中，单个细胞的基因表达量差异很大，某个基因可以在某个或某些细胞偶然性地高表达或低表达，这和细胞环境及所受刺激没有直接关联。伪批量（pseudo-bulk）是将同一实验条件下同一类型的多个细胞数据合并并计算的平均值。伪批量的稳定性和可靠性较原始数据（这里指没有转化为伪批量的数据）强，更容易发现不同刺激条件下基因表达的差异。</p>
<p>下面代码将以CD4 Naive T cells和CD14 Monocyte为例进行比较。如下图所示，很多相同的、可能与干扰素保守反应途径相关的基因在这两种细胞中都被上调，尽管CD14 Monocyte的转录水平差异更大。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">library(ggplot2)</span><br><span class="line">library(cowplot)</span><br><span class="line">theme_set(theme_cowplot())</span><br><span class="line"></span><br><span class="line">aggregate_ifnb &lt;- AggregateExpression(ifnb, group.by = c(&quot;seurat_annotations&quot;, &quot;stim&quot;), return.seurat = TRUE)</span><br><span class="line">genes.to.label = c(&quot;ISG15&quot;, &quot;LY6E&quot;, &quot;IFI6&quot;, &quot;ISG20&quot;, &quot;MX1&quot;, &quot;IFIT2&quot;, &quot;IFIT1&quot;, &quot;CXCL10&quot;, &quot;CCL8&quot;)</span><br><span class="line"></span><br><span class="line">p1 &lt;- CellScatter(aggregate_ifnb, &quot;CD14 Mono_CTRL&quot;, &quot;CD14 Mono_STIM&quot;, highlight = genes.to.label)</span><br><span class="line">p2 &lt;- LabelPoints(plot = p1, points = genes.to.label, repel = TRUE)</span><br><span class="line"></span><br><span class="line">p3 &lt;- CellScatter(aggregate_ifnb, &quot;CD4 Naive T_CTRL&quot;, &quot;CD4 Naive T_STIM&quot;, highlight = genes.to.label)</span><br><span class="line">p4 &lt;- LabelPoints(plot = p3, points = genes.to.label, repel = TRUE)</span><br><span class="line"></span><br><span class="line">p1 +p2 + p3 + p4</span><br></pre></td></tr></table></figure>

<ul>
<li>运行结果：<br><img src="/images/integration/9f214fbe-ca90-4b36-9cff-2302ef0744d4.png"></li>
</ul>
<p>上述代码可以比较我们所指定的基因在不同刺激条件下的表达差异。然而，当我们不清楚在不同处理条件下哪些基因表达水平会发生显著变化时，如果一个一个的调出与比较，很显然效率会过于低下。这时候你可能会想到（没想到也没关系），在<a target="_blank" rel="noopener" href="https://nahida23333.github.io/2025/03/28/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9ASeurat%E7%9A%84%E5%B8%B8%E8%A7%81%E5%88%86%E6%9E%90%E5%B7%A5%E4%BD%9C%E6%B5%81%E7%A8%8B%EF%BC%9A%E4%BB%A5PBMC-3K%E6%95%B0%E6%8D%AE%E9%9B%86%E4%B8%BA%E4%BE%8B/#%E9%89%B4%E5%AE%9A%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%9F%BA%E5%9B%A0%EF%BC%88DEGs%EF%BC%89">Seurat标准分析工作流程的教程</a>中用到的<code>FindMarkers()</code>和<code>FindAllMarkers()</code>函数可以用来识别并列出不同细胞簇之间的差异表达基因，那么在这里能不能用呢？</p>
<p>答案是可以的。在这里，我们在meta.data这个”槽”中新建一列用以存储细胞类型和存储信息，然后将<code>ident</code>（默认分组标识）切换到这个列；也就是告诉机器，细胞按照这个标准分组。然后我们就可以上<code>FindMarkers()</code>函数啦，这样就可以区分出不同刺激条件下的差异表达基因。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">ifnb$celltype.stim &lt;- paste(ifnb$seurat_annotations, ifnb$stim, sep = &quot;_&quot;)</span><br><span class="line"># 将每个细胞的所属细胞类型和处理条件合并、并用“_”分开，从而进行重新分组，比如刺激组的B细胞将会重新分组为“B_STIM”。</span><br><span class="line">Idents(ifnb) &lt;- &quot;celltype.stim&quot;</span><br><span class="line"># 将Seurat对象的默认分组标识（Idents）设为新生成的celltype.stim列。</span><br><span class="line">b.interferon.response &lt;- FindMarkers(ifnb, ident.1 = &quot;B_STIM&quot;, ident.2 = &quot;B_CTRL&quot;, verbose = FALSE)</span><br><span class="line"># 相对于对照组的B细胞，寻找刺激组的B细胞的差异表达基因</span><br><span class="line">head(b.interferon.response, n = 15)</span><br><span class="line"># 显示前15个差异表达基因</span><br></pre></td></tr></table></figure>
<ul>
<li>运行结果：</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">&gt; head(b.interferon.response, n = 15)</span><br><span class="line">                p_val avg_log2FC pct.1 pct.2     p_val_adj</span><br><span class="line">ISG15   5.387767e-159  5.0588481 0.998 0.233 7.571429e-155</span><br><span class="line">IFIT3   1.945114e-154  6.1124940 0.965 0.052 2.733468e-150</span><br><span class="line">IFI6    2.503565e-152  5.4933132 0.965 0.076 3.518260e-148</span><br><span class="line">ISG20   6.492570e-150  3.0549593 1.000 0.668 9.124009e-146</span><br><span class="line">IFIT1   1.951022e-139  6.2320388 0.907 0.029 2.741772e-135</span><br><span class="line">MX1     6.897626e-123  3.9798482 0.905 0.115 9.693234e-119</span><br><span class="line">LY6E    2.825649e-120  3.7907800 0.898 0.150 3.970885e-116</span><br><span class="line">TNFSF10 4.007285e-112  6.5802175 0.786 0.020 5.631437e-108</span><br><span class="line">IFIT2   2.672552e-108  5.5525558 0.786 0.037 3.755738e-104</span><br><span class="line">B2M      5.283684e-98  0.6104044 1.000 1.000  7.425161e-94</span><br><span class="line">PLSCR1   4.634658e-96  3.8010721 0.793 0.113  6.513085e-92</span><br><span class="line">IRF7     2.411149e-94  3.1992949 0.835 0.187  3.388388e-90</span><br><span class="line">CXCL10   3.708508e-86  8.0906108 0.651 0.010  5.211566e-82</span><br><span class="line">UBE2L6   5.547472e-83  2.5167981 0.851 0.297  7.795863e-79</span><br><span class="line">PSMB9    1.716262e-77  1.7715351 0.937 0.568  2.411863e-73</span><br></pre></td></tr></table></figure>
<p>我们需要注意的是，在解释上述结果中所显示的$p$值要慎之又慎。原因主要在于忽略了各个细胞间的共同来源性（非独立性）。同一批次的各个细胞之间的遗传背景、环境暴露等因素相同或类似，且存在相互影响。而在统计学分析过程中，则默认每个细胞对于刺激所作出的反应是独立的。讲人话就是，假设有样本A、B各1000个细胞，如果我们直接将2000个细胞作为独立的“观测值”，分析时程序会误认为有2000个样本（但实质上只有A和B两个样本）。由于同一样本的细胞间变异通常小于样本间细胞的变异性，此时变异性会被稀释，从而导致更容易得到统计差异显著的结果，也就是“假阳性”。</p>
<p>如何解决上述问题呢？答案就是伪批量分析。通过计算样本内所有基因表达量的平均值，将同一样本内的多个细胞合并为“一个”细胞，最后在比较时就是两个“细胞”之间的比较啦！就可以顺利地解决问题！</p>
<p>我们还可以利用<code>FeaturePlot()</code>和<code>VlnPlot()</code>函数，通过调整<code>split.by</code>参数进行可视化。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">FeaturePlot(ifnb, features = c(&quot;CD3D&quot;, &quot;GNLY&quot;, &quot;IFI6&quot;), split.by = &quot;stim&quot;, max.cutoff = 3, cols = c(&quot;grey&quot;, &quot;red&quot;), reduction = &quot;umap&quot;)</span><br></pre></td></tr></table></figure>
<ul>
<li>运行结果：<img src="/images/integration/a69ec383-eb00-4edc-9326-bffd7d9add77.png"></li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">plots &lt;- VlnPlot(ifnb, features = c(&quot;CD14&quot;, &quot;ISG15&quot;, &quot;CXCL10&quot;), split.by = &quot;stim&quot;, group.by = &quot;seurat_annotations&quot;, pt.size = 0, combine = FALSE)</span><br><span class="line">wrap_plots(plots = plots, ncol = 1)</span><br><span class="line"># wrap_plots()函数可以将不同的图像组合起来，合并为一张图。</span><br></pre></td></tr></table></figure>
<ul>
<li>运行结果：<img src="/images/integration/Rplot01.png"></li>
</ul>
<p><code>CD3D</code>基因是T细胞的特征基因，而<code>GNLY</code>基因则是NK&#x2F;CD8 T细胞的特征基因。从我们得到的图像中不难看出，其表达在两组细胞中差异并不大，说明这两个基因几乎不受干扰素刺激的影响，并且在对照组和刺激组中表现出相似的基因表达模式。<code>IF16</code>和<code>ISG15</code>基因是干扰素反应的核心基因，在刺激组中的表达明显上调。<code>CD14</code>和<code>CXCL10</code>基因是在干扰素（IFN）应答中具有细胞类型特异性的基因。在刺激组中，我们发现刺激组CD14 Monocyte的<code>CD14</code>基因反而出现下调，<code>CXCL10</code>基因在单核细胞和B细胞中表现出明显的上调。</p>
<p>这里值得一提的是，如果我们将<code>CD14</code>作为识别CD14 Monocyte的唯一标志基因的话，由于刺激组<code>CD14</code>基因表达下调，在分析的时候容易将这些细胞误认为非CD14 Monocyte而排除，就会造成分类的错误。这凸显了整合分析的重要性。</p>
<h1 id="执行与SCTransform规范化数据集的集成"><a href="#执行与SCTransform规范化数据集的集成" class="headerlink" title="执行与SCTransform规范化数据集的集成"></a>执行与SCTransform规范化数据集的集成</h1><p>LogNormalize太差了怎么办！可以试试它！SCTransform标准化，更适合单细胞数据宝宝体质！（划掉）</p>
<p>还是和之前一样，用SCTransform标准化替代LogNormalize、FindVaribleFeatures、ScaleData这三步。在这里不做过多解释啦！</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">ifnb &lt;- LoadData(&quot;ifnb&quot;)</span><br><span class="line">ifnb[[&quot;RNA&quot;]] &lt;- split(ifnb[[&quot;RNA&quot;]], f = ifnb$stim)</span><br><span class="line"></span><br><span class="line">ifnb &lt;- SCTransform(ifnb)</span><br><span class="line">ifnb &lt;- RunPCA(ifnb)</span><br><span class="line">ifnb &lt;- RunUMAP(ifnb, dims = 1:30)</span><br><span class="line">DimPlot(ifnb, reduction = &quot;umap&quot;, group.by = c(&quot;stim&quot;, &quot;seurat_annotations&quot;))</span><br></pre></td></tr></table></figure>
<ul>
<li>运行结果：<img src="/images/integration/a719dc30-4647-42aa-9fcc-dec5c573508e.png"></li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">ifnb &lt;- IntegrateLayers(object = ifnb, method = CCAIntegration, normalization.method = &quot;SCT&quot;, verbose = F)</span><br><span class="line">ifnb &lt;- FindNeighbors(ifnb, reduction = &quot;integrated.dr&quot;, dims = 1:30)</span><br><span class="line">ifnb &lt;- FindClusters(ifnb, resolution = 0.6)</span><br><span class="line"></span><br><span class="line">ifnb &lt;- RunUMAP(ifnb, dims = 1:30, reduction = &quot;integrated.dr&quot;)</span><br><span class="line">DimPlot(ifnb, reduction = &quot;umap&quot;, group.by = c(&quot;stim&quot;, &quot;seurat_annotations&quot;))</span><br></pre></td></tr></table></figure>
<ul>
<li>运行结果：<img src="/images/integration/96a6d733-d842-4525-9592-c678e81dc536.png"></li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">ifnb &lt;- PrepSCTFindMarkers(ifnb)</span><br><span class="line">ifnb$celltype.stim &lt;- paste(ifnb$seurat_annotations, ifnb$stim, sep = &quot;_&quot;)</span><br><span class="line">Idents(ifnb) &lt;- &quot;celltype.stim&quot;</span><br><span class="line">b.interferon.response &lt;- FindMarkers(ifnb, ident.1 = &quot;B_STIM&quot;, ident.2 = &quot;B_CTRL&quot;, verbose = FALSE)</span><br></pre></td></tr></table></figure>
<p>再往下就和前面一样啦，所以到这里！本篇完结撒花！</p>
<h1 id="完整代码"><a href="#完整代码" class="headerlink" title="完整代码"></a>完整代码</h1><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br></pre></td><td class="code"><pre><span class="line">library(Seurat)</span><br><span class="line">library(SeuratData)</span><br><span class="line">library(patchwork)</span><br><span class="line">options(timeout=10000)</span><br><span class="line">InstallData(&quot;ifnb&quot;)</span><br><span class="line"></span><br><span class="line">ifnb &lt;- LoadData(&quot;ifnb&quot;)</span><br><span class="line">ifnb[[&quot;RNA&quot;]] &lt;- split(ifnb[[&quot;RNA&quot;]], f = ifnb$stim)</span><br><span class="line">ifnb</span><br><span class="line"></span><br><span class="line">ifnb &lt;- NormalizeData(ifnb)</span><br><span class="line">ifnb &lt;- FindVariableFeatures(ifnb)</span><br><span class="line">ifnb &lt;- ScaleData(ifnb)</span><br><span class="line">ifnb &lt;- RunPCA(ifnb)</span><br><span class="line">ifnb &lt;- FindNeighbors(ifnb, dims = 1:30, reduction = &quot;pca&quot;)</span><br><span class="line">ifnb &lt;- FindClusters(ifnb, resolution = 2, cluster.name = &quot;unintegrated_clusters&quot;)</span><br><span class="line">ifnb &lt;- RunUMAP(ifnb, dims = 1:30, reduction = &quot;pca&quot;, reduction.name = &quot;umap.unintegrated&quot;)</span><br><span class="line">DimPlot(ifnb, reduction = &quot;umap.unintegrated&quot;, group.by = c(&quot;stim&quot;, &quot;seurat_clusters&quot;))</span><br><span class="line"></span><br><span class="line">ifnb &lt;- IntegrateLayers(object = ifnb, method = CCAIntegration, orig.reduction = &quot;pca&quot;, new.reduction = &quot;integrated.cca&quot;, verbose = FALSE)</span><br><span class="line">ifnb[[&quot;RNA&quot;]] &lt;- JoinLayers(ifnb[[&quot;RNA&quot;]])</span><br><span class="line"></span><br><span class="line">ifnb &lt;- FindNeighbors(ifnb, reduction = &quot;integrated.cca&quot;, dims = 1:30)</span><br><span class="line">ifnb &lt;- FindClusters(ifnb, resolution = 1)</span><br><span class="line">ifnb &lt;- RunUMAP(ifnb, dims = 1:30, reduction = &quot;integrated.cca&quot;)</span><br><span class="line"></span><br><span class="line">DimPlot(ifnb, reduction = &quot;umap&quot;, group.by = c(&quot;stim&quot;, &quot;seurat_annotations&quot;))</span><br><span class="line">DimPlot(ifnb, reduction = &quot;umap&quot;, split.by = &quot;stim&quot;)</span><br><span class="line"></span><br><span class="line">install.packages(&#x27;BiocManager&#x27;)</span><br><span class="line">BiocManager::install(&#x27;multtest&#x27;)</span><br><span class="line">install.packages(&#x27;metap&#x27;)</span><br><span class="line">Idents(ifnb) &lt;- &quot;seurat_annotations&quot;</span><br><span class="line">nk.markers &lt;- FindConservedMarkers(ifnb, ident.1 = &quot;NK&quot;, grouping.var = &quot;stim&quot;, verbose = FALSE)</span><br><span class="line">head(nk.markers)</span><br><span class="line"></span><br><span class="line">Idents(ifnb) &lt;- factor(Idents(ifnb), levels = c(&quot;pDC&quot;, &quot;Eryth&quot;, &quot;Mk&quot;, &quot;DC&quot;, &quot;CD14 Mono&quot;, &quot;CD16 Mono&quot;, &quot;B Activated&quot;, &quot;B&quot;, &quot;CD8 T&quot;, &quot;NK&quot;, &quot;T activated&quot;, &quot;CD4 Naive T&quot;, &quot;CD4 Memory T&quot;))</span><br><span class="line"></span><br><span class="line">markers.to.plot &lt;- c(&quot;CD3D&quot;, &quot;CREM&quot;, &quot;HSPH1&quot;, &quot;SELL&quot;, &quot;GIMAP5&quot;, &quot;CACYBP&quot;, &quot;GNLY&quot;, &quot;NKG7&quot;, &quot;CCL5&quot;, &quot;CD8A&quot;, &quot;MS4A1&quot;, &quot;CD79A&quot;, &quot;MIR155HG&quot;, &quot;NME1&quot;, &quot;FCGR3A&quot;, &quot;VMO1&quot;, &quot;CCL2&quot;, &quot;S100A9&quot;, &quot;HLA-DQA1&quot;, &quot;GPR183&quot;, &quot;PPBP&quot;, &quot;GNG11&quot;, &quot;HBA2&quot;, &quot;HBB&quot;, &quot;TSPAN13&quot;, &quot;IL3RA&quot;, &quot;IGJ&quot;, &quot;PRSS57&quot;)</span><br><span class="line">DotPlot(ifnb, features = markers.to.plot, cols = c(&quot;blue&quot;, &quot;red&quot;), dot.scale = 8, split.by = &quot;stim&quot;) + RotatedAxis()</span><br><span class="line"></span><br><span class="line">install.packages(&quot;ggplot2&quot;)</span><br><span class="line">library(ggplot2)</span><br><span class="line">library(cowplot)</span><br><span class="line">theme_set(theme_cowplot())</span><br><span class="line"></span><br><span class="line">aggregate_ifnb &lt;- AggregateExpression(ifnb, group.by = c(&quot;seurat_annotations&quot;, &quot;stim&quot;), return.seurat = TRUE)</span><br><span class="line">genes.to.label = c(&quot;ISG15&quot;, &quot;LY6E&quot;, &quot;IFI6&quot;, &quot;ISG20&quot;, &quot;MX1&quot;, &quot;IFIT2&quot;, &quot;IFIT1&quot;, &quot;CXCL10&quot;, &quot;CCL8&quot;)</span><br><span class="line"></span><br><span class="line">p1 &lt;- CellScatter(aggregate_ifnb, &quot;CD14 Mono_CTRL&quot;, &quot;CD14 Mono_STIM&quot;, highlight = genes.to.label)</span><br><span class="line">p2 &lt;- LabelPoints(plot = p1, points = genes.to.label, repel = TRUE)</span><br><span class="line"></span><br><span class="line">p3 &lt;- CellScatter(aggregate_ifnb, &quot;CD4 Naive T_CTRL&quot;, &quot;CD4 Naive T_STIM&quot;, highlight = genes.to.label)</span><br><span class="line">p4 &lt;- LabelPoints(plot = p3, points = genes.to.label, repel = TRUE)</span><br><span class="line"></span><br><span class="line">p1 +p2 + p3 + p4</span><br><span class="line"></span><br><span class="line">ifnb$celltype.stim &lt;- paste(ifnb$seurat_annotations, ifnb$stim, sep = &quot;_&quot;)</span><br><span class="line">Idents(ifnb) &lt;- &quot;celltype.stim&quot;</span><br><span class="line">b.interferon.response &lt;- FindMarkers(ifnb, ident.1 = &quot;B_STIM&quot;, ident.2 = &quot;B_CTRL&quot;, verbose = FALSE)</span><br><span class="line">head(b.interferon.response, n = 15)</span><br><span class="line"></span><br><span class="line">FeaturePlot(ifnb, features = c(&quot;CD3D&quot;, &quot;GNLY&quot;, &quot;IFI6&quot;), split.by = &quot;stim&quot;, max.cutoff = 3, cols = c(&quot;grey&quot;, &quot;red&quot;), reduction = &quot;umap&quot;)</span><br><span class="line">plots &lt;- VlnPlot(ifnb, features = c(&quot;CD14&quot;, &quot;ISG15&quot;, &quot;CXCL10&quot;), split.by = &quot;stim&quot;, group.by = &quot;seurat_annotations&quot;, pt.size = 0, combine = FALSE)</span><br><span class="line">wrap_plots(plots = plots, ncol = 1)</span><br><span class="line"></span><br><span class="line">ifnb &lt;- LoadData(&quot;ifnb&quot;)</span><br><span class="line"></span><br><span class="line"># split datasets and process without integration</span><br><span class="line">ifnb[[&quot;RNA&quot;]] &lt;- split(ifnb[[&quot;RNA&quot;]], f = ifnb$stim)</span><br><span class="line">ifnb &lt;- SCTransform(ifnb)</span><br><span class="line">ifnb &lt;- RunPCA(ifnb)</span><br><span class="line">ifnb &lt;- RunUMAP(ifnb, dims = 1:30)</span><br><span class="line">DimPlot(ifnb, reduction = &quot;umap&quot;, group.by = c(&quot;stim&quot;, &quot;seurat_annotations&quot;))</span><br><span class="line"></span><br><span class="line"># integrate datasets</span><br><span class="line">ifnb &lt;- IntegrateLayers(object = ifnb, method = CCAIntegration, normalization.method = &quot;SCT&quot;, verbose = F)</span><br><span class="line">ifnb &lt;- FindNeighbors(ifnb, reduction = &quot;integrated.dr&quot;, dims = 1:30)</span><br><span class="line">ifnb &lt;- FindClusters(ifnb, resolution = 0.6)</span><br><span class="line"></span><br><span class="line">ifnb &lt;- RunUMAP(ifnb, dims = 1:30, reduction = &quot;integrated.dr&quot;)</span><br><span class="line">DimPlot(ifnb, reduction = &quot;umap&quot;, group.by = c(&quot;stim&quot;, &quot;seurat_annotations&quot;))</span><br><span class="line"></span><br><span class="line"># perform differential expression</span><br><span class="line">ifnb &lt;- PrepSCTFindMarkers(ifnb)</span><br><span class="line">ifnb$celltype.stim &lt;- paste(ifnb$seurat_annotations, ifnb$stim, sep = &quot;_&quot;)</span><br><span class="line">Idents(ifnb) &lt;- &quot;celltype.stim&quot;</span><br><span class="line">b.interferon.response &lt;- FindMarkers(ifnb, ident.1 = &quot;B_STIM&quot;, ident.2 = &quot;B_CTRL&quot;, verbose = FALSE)</span><br></pre></td></tr></table></figure></article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta"><i class="fas fa-circle-user fa-fw"></i>文章作者: </span><span class="post-copyright-info"><a href="http://example.com">Nahida</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta"><i class="fas fa-square-arrow-up-right fa-fw"></i>文章链接: </span><span class="post-copyright-info"><a href="http://example.com/2025/04/29/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%84%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90%E4%BB%8B%E7%BB%8D/">http://example.com/2025/04/29/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%84%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90%E4%BB%8B%E7%BB%8D/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta"><i class="fas fa-circle-exclamation fa-fw"></i>版权声明: </span><span class="post-copyright-info">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0</a> 许可协议。转载请注明来源 <a href="http://example.com" target="_blank">Nahida的净善宫</a>！</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/">生信分析</a><a class="post-meta__tags" href="/tags/R%E8%AF%AD%E8%A8%80/">R语言</a><a class="post-meta__tags" href="/tags/Seurat/">Seurat</a></div><div class="post-share"><div class="social-share" data-image="/images/Seurat.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><a class="pagination-related" href="/2025/04/29/%E3%80%90%E5%8C%BB%E5%AD%A6%E5%BD%B1%E5%83%8F%E5%AD%A6%E3%80%91%E5%85%B3%E8%8A%82%E7%9A%84%E6%A3%80%E6%9F%A5%E6%96%B9%E6%B3%95%E5%8F%8A%E5%B8%B8%E8%A7%81%E7%97%85%E5%8F%98/" title="【医学影像学】关节的检查方法及常见病变"><div class="cover" style="background: var(--default-bg-color)"></div><div class="info"><div class="info-1"><div class="info-item-1">上一篇</div><div class="info-item-2">【医学影像学】关节的检查方法及常见病变</div></div><div class="info-2"><div class="info-item-1">关节的正常影像学解剖骨与骨之间的连结 直接连接：韧带连结、软骨连结、骨性结合 间接连结：滑膜关节（synovial joints），如四肢各关节  基本结构：覆盖关节软骨的关节骨端、关节囊、关节腔及其附属结构 关节正常结构及其X线表现 关节间隙：两侧骨端的骨性关节面之间的透亮间隙 骨性关节面：为关节软骨下方的钙化带及薄层致密骨质 关节囊及韧带：一般不能显示 关节内外脂肪层：稍低密度，可衬托出关节囊的轮廓。  下图为正常膝关节的X线表现，其中*为关节间隙；↑为骨性关节面；◊为髌上囊及髌下囊。  儿童骨关节X线解剖特点： 可见骨骺、干骺端、骨骺线（骨骺板）、关节间隙。儿童骺软骨尚未完全骨化，因而“关节软骨”较厚，关节间隙较成人宽。 关节病变的基本影像学征象关节肿胀多为关节局部炎症、急性创伤性关节积液或关节囊及周围软组织充血、水肿、出血所致平片：关节周围软组织肿胀，密度增高；软组织层次欠清晰；大量积液：关节间隙增宽 影像学图解双膝关节肿胀：软组织肿胀、密度增高 关节间隙增宽 普遍性骨质疏松 关节间隙异常 关节破坏定义： 关节软骨及其下方骨质为病理组织侵犯、取代 病因：...</div></div></div></a><a class="pagination-related" href="/2025/05/06/%E3%80%90%E5%8C%BB%E5%AD%A6%E5%BD%B1%E5%83%8F%E5%AD%A6%E3%80%91%E8%84%8A%E6%9F%B1%E5%85%B3%E8%8A%82%E7%9A%84%E6%AD%A3%E5%B8%B8%E5%BD%B1%E5%83%8F%E5%AD%A6%E8%A1%A8%E7%8E%B0%E3%80%81%E5%9F%BA%E6%9C%AC%E7%97%85%E5%8F%98%E4%B8%8E%E5%B8%B8%E8%A7%81%E7%96%BE%E7%97%85/" title="【医学影像学】脊柱关节的正常影像学表现、基本病变与常见疾病"><div class="cover" style="background: var(--default-bg-color)"></div><div class="info text-right"><div class="info-1"><div class="info-item-1">下一篇</div><div class="info-item-2">【医学影像学】脊柱关节的正常影像学表现、基本病变与常见疾病</div></div><div class="info-2"><div class="info-item-1">...</div></div></div></a></nav><div class="relatedPosts"><div class="headline"><i class="fas fa-thumbs-up fa-fw"></i><span>相关推荐</span></div><div class="relatedPosts-list"><a class="pagination-related" href="/2025/09/20/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E6%98%A0%E5%B0%84%E5%92%8C%E6%B3%A8%E9%87%8A%E6%9F%A5%E8%AF%A2%E6%95%B0%E6%8D%AE%E9%9B%86/" title="生信小记：映射和注释查询数据集"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-09-20</div><div class="info-item-2">生信小记：映射和注释查询数据集</div></div><div class="info-2"><div class="info-item-1">在前面的教程中，我们比较系统地学习了处理单细胞测序数据的分析流程。分析流程严谨而精细，可以帮助我们获得很多有价值的信息。然而，对于长期专注于某一领域的研究者来说，每拿到一批数据都要走一遍这样的分析流程，其计算量大、耗费的时间精力多、效率低下，不太能满足研究者的效率需求。为了提高效率，研究者们开辟出了一条更为方便的捷径，即使用多个可靠的数据集构建一个整合的参考框架，在拿到新的数据之后不再需要从头开始分析，只需要通过算法将数据映射到参考数据集上，即可自动获得较高质量的注释。这个过程很像人工智能大模型的开发，即使用高质量的标注标注数据​“训练”​​一个模型，学习其中的模式、特征和关联，让其获得充分的知识储备和逻辑分析能力，在遇到新问题时可以利用内部知识进行推理并生成答案。 数据集的预处理我们先引入示例数据集，这份数据集包括由四种不同测序技术得到的人胰岛细胞数据，即CelSeq（GSE81076）、CelSeq2（GSE85241）、Fluidigm...</div></div></div></a><a class="pagination-related" href="/2025/08/20/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9Ascvi-tools%E7%9A%84%E5%AE%89%E8%A3%85/" title="生信小记：scvi_tools的安装"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-08-20</div><div class="info-item-2">生信小记：scvi_tools的安装</div></div><div class="info-2"><div class="info-item-1">使用scvi算法进行单细胞测序数据的整合，一方面需要reticulate包，另一方面则需要安装在conda环境中的scvi-tools包及其依赖项。前者可以通过install.packages()语句从CRAN中下载，而本文章将介绍如何从零开始安装scvi-tools。 参考教程如下：  Anaconda安装教程（2025年保姆级超详解）【附安装包+环境玩转指南】   安装Anaconda打开Anaconda官网，点击右上角的“Free Download”，注册并登录后即可下载。 下载完成后打开安装包，可见如下界面。点击Next &gt;。 点击I Agree。 点击All users (requires admin privileges)。 选择存储位置时，强烈不建议存储在C盘，因为它安装完占用的空间不小。更改完存储路径后点击Next &gt;。 全勾选上再点Next...</div></div></div></a><a class="pagination-related" href="/2025/07/09/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9ASeurat-v5%E4%B8%AD%E7%9A%84%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90/" title="生信小记：Seurat_v5中的整合分析"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-07-09</div><div class="info-item-2">生信小记：Seurat_v5中的整合分析</div></div><div class="info-2"><div class="info-item-1">前言单细胞测序中，实验和数据的收集往往不是一次性完成的，而是分为多个批次进行的。在不同批次之间，操作者、操作流程等诸多因素存在差异，可能给实验引入一些与真实生物学差异无关的系统性偏差，称为 批次效应 。 单细胞测序数据的整合通常是scRNA-seq测序的重要步骤之一。通过匹配不同数据集之间的共享细胞类型和状态，我们可以消除批次效应所引入的偏差，从而提高了统计功效。我们常使用的整合数据的方法有多种，包括Seurat早期版本发布的基于锚点的整合流程，其他团队发布的整合算法（如Harmony、scVI等）。Seurat...</div></div></div></a><a class="pagination-related" href="/2025/03/31/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E4%BD%BF%E7%94%A8Seurat%E5%8C%85%E5%A4%84%E7%90%86%E5%A4%9A%E6%A8%A1%E6%80%81%E6%95%B0%E6%8D%AE/" title="生信小记：使用Seurat包处理多模态数据"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-03-31</div><div class="info-item-2">生信小记：使用Seurat包处理多模态数据</div></div><div class="info-2"><div class="info-item-1">在一次实验中，我们可以对同一个细胞测量不同分子层面的数据（如基因表达、蛋白质表达、表观遗传修饰、空间分布等等），所测得的数据我们称之为多模态数据（multimodal data）。例如，CITE-seq可以同时检测单细胞的转录组RNA和表面蛋白表达水平；10X Multiome可以同时测量单细胞的转录组RNA和ATAC（染色质可及性）。本篇文章将以一个包括8617个脐带血单个核细胞（CBMC）的CITE-seq结果（包含转录组RNA测序和表面蛋白丰度估计值）的数据集，介绍如何使用Seurat包对多模态数据进行分析。 导入数据，创建Seurat对象，并加入RNA和表面蛋白数据官方教程中介绍的方法如下： 12345678910111213141516171819202122232425262728library(Seurat)library(ggplot2)library(patchwork)# 注意：这份数据同时存在人和小鼠的细胞，分别冠有&quot;HUMAN_&quot;和&quot;MOUSE_&quot;的前缀。# 导入RNA数据cbmc.rna &lt;-...</div></div></div></a><a class="pagination-related" href="/2025/03/31/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E4%BD%BF%E7%94%A8SCTransform%E8%BF%9B%E8%A1%8C%E6%A0%87%E5%87%86%E5%8C%96/" title="生信小记：使用SCTransform进行标准化"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-03-31</div><div class="info-item-2">生信小记：使用SCTransform进行标准化</div></div><div class="info-2"><div class="info-item-1">...</div></div></div></a><a class="pagination-related" href="/2025/03/29/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9ASeurat%E5%8C%85%E7%9A%84%E6%95%B0%E6%8D%AE%E5%8F%AF%E8%A7%86%E5%8C%96%E6%96%B9%E6%B3%95/" title="生信小记：Seurat包的数据可视化方法"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-03-29</div><div class="info-item-2">生信小记：Seurat包的数据可视化方法</div></div><div class="info-2"><div class="info-item-1">在这篇教程中，我们会沿用上一篇教程处理好的数据来进行演示。 先做一些前期准备，安装并加载好R包，下载并加载好数据（或者直接运用上一篇教程中处理好的数据）。 12345678910111213141516171819202122install.packages(c(&quot;Seurat&quot;,&quot;patchwork&quot;,&quot;ggplot2&quot;,&quot;SeuratData&quot;))#下载SeuratData包时报错显示：package ‘SeuratData’ is not available for this version of...</div></div></div></a></div></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info text-center"><div class="avatar-img"><img src="/img/avatar.jpg" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info-name">Nahida</div><div class="author-info-description">智慧树上智慧果，智慧树下你和我</div><div class="site-data"><a href="/archives/"><div class="headline">文章</div><div class="length-num">97</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">58</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">22</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/xxxxxx"><i class="fab fa-github"></i><span>Follow Me</span></a></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>公告</span></div><div class="announcement_content">暂无公告哦~</div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>目录</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%88%9B%E5%BB%BASeurat%E5%AF%B9%E8%B1%A1"><span class="toc-number">1.</span> <span class="toc-text">创建Seurat对象</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E8%BF%9B%E8%A1%8C%E4%B8%8D%E6%95%B4%E5%90%88%E7%9A%84%E7%AE%80%E5%8D%95%E5%88%86%E6%9E%90"><span class="toc-number">2.</span> <span class="toc-text">进行不整合的简单分析</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%95%B4%E5%90%88%E4%B8%A4%E7%BB%84%E6%95%B0%E6%8D%AE%E9%9B%86%EF%BC%88CCA%E9%99%8D%E7%BB%B4%EF%BC%89"><span class="toc-number">3.</span> <span class="toc-text">整合两组数据集（CCA降维）</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E8%AF%86%E5%88%AB%E4%BF%9D%E5%AE%88%E7%9A%84%E7%BB%86%E8%83%9E%E7%B1%BB%E5%9E%8B%E6%A0%87%E8%AE%B0"><span class="toc-number">4.</span> <span class="toc-text">识别保守的细胞类型标记</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E8%AF%86%E5%88%AB%E4%B8%8D%E5%90%8C%E5%A4%84%E7%90%86%E6%9D%A1%E4%BB%B6%E4%B8%8B%E7%BB%86%E8%83%9E%E7%9A%84%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%9F%BA%E5%9B%A0"><span class="toc-number">5.</span> <span class="toc-text">识别不同处理条件下细胞的差异表达基因</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%89%A7%E8%A1%8C%E4%B8%8ESCTransform%E8%A7%84%E8%8C%83%E5%8C%96%E6%95%B0%E6%8D%AE%E9%9B%86%E7%9A%84%E9%9B%86%E6%88%90"><span class="toc-number">6.</span> <span class="toc-text">执行与SCTransform规范化数据集的集成</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%AE%8C%E6%95%B4%E4%BB%A3%E7%A0%81"><span class="toc-number">7.</span> <span class="toc-text">完整代码</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>最新文章</span></div><div class="aside-list"><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2025/09/28/%E3%80%90%E7%9C%BC%E7%A7%91%E5%AD%A6%E3%80%91%E8%91%A1%E8%90%84%E8%86%9C%E7%82%8E/" title="【眼科学】葡萄膜炎">【眼科学】葡萄膜炎</a><time datetime="2025-09-28T07:41:33.000Z" title="发表于 2025-09-28 15:41:33">2025-09-28</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2025/09/28/%E3%80%90%E7%9C%BC%E7%A7%91%E5%AD%A6%E3%80%91%E5%B9%B2%E7%9C%BC%E7%97%87/" title="【眼科学】干眼症">【眼科学】干眼症</a><time datetime="2025-09-28T06:37:52.000Z" title="发表于 2025-09-28 14:37:52">2025-09-28</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2025/09/25/%E3%80%90%E7%9C%BC%E7%A7%91%E5%AD%A6%E3%80%91%E7%BB%93%E8%86%9C%E7%96%BE%E7%97%85/" title="【眼科学】结膜疾病">【眼科学】结膜疾病</a><time datetime="2025-09-25T13:58:52.000Z" title="发表于 2025-09-25 21:58:52">2025-09-25</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2025/09/25/%E3%80%90%E7%9C%BC%E7%A7%91%E5%AD%A6%E3%80%91%E8%A7%92%E8%86%9C%E7%96%BE%E7%97%85/" title="【眼科学】角膜疾病">【眼科学】角膜疾病</a><time datetime="2025-09-25T11:39:15.000Z" title="发表于 2025-09-25 19:39:15">2025-09-25</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2025/09/25/%E3%80%90%E7%9C%BC%E7%A7%91%E5%AD%A6%E3%80%91%E7%9C%BC%E7%A7%91%E5%AD%A6%E5%9F%BA%E7%A1%80/" title="【眼科学】眼科学基础">【眼科学】眼科学基础</a><time datetime="2025-09-25T08:23:24.000Z" title="发表于 2025-09-25 16:23:24">2025-09-25</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2019 - 2025 By Nahida</div><div class="framework-info"><span>框架 </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo 7.3.0</a><span class="footer-separator">|</span><span>主题 </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly 5.3.5</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="阅读模式"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="日间和夜间模式切换"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="单栏和双栏切换"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside-config" type="button" title="设置"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="目录"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="回到顶部"><span class="scroll-percent"></span><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><div class="js-pjax"><script>(() => {
  const loadMathjax = () => {
    if (!window.MathJax) {
      window.MathJax = {
        tex: {
          inlineMath: [['$', '$'], ['\\(', '\\)']],
          tags: 'none',
        },
        chtml: {
          scale: 1.1
        },
        options: {
          enableMenu: true,
          renderActions: {
            findScript: [10, doc => {
              for (const node of document.querySelectorAll('script[type^="math/tex"]')) {
                const display = !!node.type.match(/; *mode=display/)
                const math = new doc.options.MathItem(node.textContent, doc.inputJax[0], display)
                const text = document.createTextNode('')
                node.parentNode.replaceChild(text, node)
                math.start = {node: text, delim: '', n: 0}
                math.end = {node: text, delim: '', n: 0}
                doc.math.push(math)
              }
            }, '']
          }
        }
      }

      const script = document.createElement('script')
      script.src = 'https://cdn.jsdelivr.net/npm/mathjax/es5/tex-mml-chtml.min.js'
      script.id = 'MathJax-script'
      script.async = true
      document.head.appendChild(script)
    } else {
      MathJax.startup.document.state(0)
      MathJax.texReset()
      MathJax.typesetPromise()
    }
  }

  btf.addGlobalFn('encrypt', loadMathjax, 'mathjax')
  window.pjax ? loadMathjax() : window.addEventListener('load', loadMathjax)
})()</script></div><script type="text/javascript" src="/js/memos/waterfall.min.js"></script><script type="text/javascript" src="/js/memos/imgStatus.min.js"></script><script type="text/javascript" src="/js/memos/lately.min.js"></script><script type="text/javascript" src="/js/memos/photo.js"></script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>