<!DOCTYPE html><html lang="zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0,viewport-fit=cover"><title>生信小记：Seurat_v5中的整合分析 | Nahida的净善宫</title><meta name="author" content="Nahida"><meta name="copyright" content="Nahida"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="前言单细胞测序中，实验和数据的收集往往不是一次性完成的，而是分为多个批次进行的。在不同批次之间，操作者、操作流程等诸多因素存在差异，可能给实验引入一些与真实生物学差异无关的系统性偏差，称为 批次效应 。 单细胞测序数据的整合通常是scRNA-seq测序的重要步骤之一。通过匹配不同数据集之间的共享细胞类型和状态，我们可以消除批次效应所引入的偏差，从而提高了统计功效。我们常使用的整合数据的方法有多种，">
<meta property="og:type" content="article">
<meta property="og:title" content="生信小记：Seurat_v5中的整合分析">
<meta property="og:url" content="http://example.com/2025/07/09/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9ASeurat-v5%E4%B8%AD%E7%9A%84%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90/index.html">
<meta property="og:site_name" content="Nahida的净善宫">
<meta property="og:description" content="前言单细胞测序中，实验和数据的收集往往不是一次性完成的，而是分为多个批次进行的。在不同批次之间，操作者、操作流程等诸多因素存在差异，可能给实验引入一些与真实生物学差异无关的系统性偏差，称为 批次效应 。 单细胞测序数据的整合通常是scRNA-seq测序的重要步骤之一。通过匹配不同数据集之间的共享细胞类型和状态，我们可以消除批次效应所引入的偏差，从而提高了统计功效。我们常使用的整合数据的方法有多种，">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="http://example.com/images/Seurat.png">
<meta property="article:published_time" content="2025-07-09T04:46:16.000Z">
<meta property="article:modified_time" content="2025-08-20T15:25:18.567Z">
<meta property="article:author" content="Nahida">
<meta property="article:tag" content="生信分析">
<meta property="article:tag" content="R语言">
<meta property="article:tag" content="Seurat">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://example.com/images/Seurat.png"><script type="application/ld+json">{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "生信小记：Seurat_v5中的整合分析",
  "url": "http://example.com/2025/07/09/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9ASeurat-v5%E4%B8%AD%E7%9A%84%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90/",
  "image": "http://example.com/images/Seurat.png",
  "datePublished": "2025-07-09T04:46:16.000Z",
  "dateModified": "2025-08-20T15:25:18.567Z",
  "author": [
    {
      "@type": "Person",
      "name": "Nahida",
      "url": "http://example.com/"
    }
  ]
}</script><link rel="shortcut icon" href="/img/avatar.jpg"><link rel="canonical" href="http://example.com/2025/07/09/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9ASeurat-v5%E4%B8%AD%E7%9A%84%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90/index.html"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css"><script>
    (() => {
      
    const saveToLocal = {
      set: (key, value, ttl) => {
        if (!ttl) return
        const expiry = Date.now() + ttl * 86400000
        localStorage.setItem(key, JSON.stringify({ value, expiry }))
      },
      get: key => {
        const itemStr = localStorage.getItem(key)
        if (!itemStr) return undefined
        const { value, expiry } = JSON.parse(itemStr)
        if (Date.now() > expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return value
      }
    }

    window.btf = {
      saveToLocal,
      getScript: (url, attr = {}) => new Promise((resolve, reject) => {
        const script = document.createElement('script')
        script.src = url
        script.async = true
        Object.entries(attr).forEach(([key, val]) => script.setAttribute(key, val))
        script.onload = script.onreadystatechange = () => {
          if (!script.readyState || /loaded|complete/.test(script.readyState)) resolve()
        }
        script.onerror = reject
        document.head.appendChild(script)
      }),
      getCSS: (url, id) => new Promise((resolve, reject) => {
        const link = document.createElement('link')
        link.rel = 'stylesheet'
        link.href = url
        if (id) link.id = id
        link.onload = link.onreadystatechange = () => {
          if (!link.readyState || /loaded|complete/.test(link.readyState)) resolve()
        }
        link.onerror = reject
        document.head.appendChild(link)
      }),
      addGlobalFn: (key, fn, name = false, parent = window) => {
        if (!false && key.startsWith('pjax')) return
        const globalFn = parent.globalFn || {}
        globalFn[key] = globalFn[key] || {}
        globalFn[key][name || Object.keys(globalFn[key]).length] = fn
        parent.globalFn = globalFn
      }
    }
  
      
      const activateDarkMode = () => {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      const activateLightMode = () => {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }

      btf.activateDarkMode = activateDarkMode
      btf.activateLightMode = activateLightMode

      const theme = saveToLocal.get('theme')
    
          theme === 'dark' ? activateDarkMode() : theme === 'light' ? activateLightMode() : null
        
      
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        document.documentElement.classList.toggle('hide-aside', asideStatus === 'hide')
      }
    
      
    const detectApple = () => {
      if (/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)) {
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
  
    })()
  </script><script>const GLOBAL_CONFIG = {
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  highlight: {"plugin":"highlight.js","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false,"highlightFullpage":false,"highlightMacStyle":false},
  copy: {
    success: '复制成功',
    error: '复制失败',
    noSupport: '浏览器不支持'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  dateSuffix: {
    just: '刚刚',
    min: '分钟前',
    hour: '小时前',
    day: '天前',
    month: '个月前'
  },
  copyright: undefined,
  lightbox: 'null',
  Snackbar: undefined,
  infinitegrid: {
    js: 'https://cdn.jsdelivr.net/npm/@egjs/infinitegrid/dist/infinitegrid.min.js',
    buttonText: '加载更多'
  },
  isPhotoFigcaption: false,
  islazyloadPlugin: false,
  isAnchor: false,
  percent: {
    toc: true,
    rightside: false,
  },
  autoDarkmode: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: '生信小记：Seurat_v5中的整合分析',
  isHighlightShrink: false,
  isToc: true,
  pageType: 'post'
}</script><meta name="generator" content="Hexo 7.3.0"><link rel="alternate" href="/atom.xml" title="Nahida的净善宫" type="application/atom+xml">
</head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img text-center"><img src="/img/avatar.jpg" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="site-data text-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">90</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">51</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">21</div></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> 时间轴</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> 标签</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> 分类</span></a></div><div class="menus_item"><a class="site-page" href="/photos/"><i class="fa-fw fas fa-images"></i><span> 相册</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> 友链</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url(/images/Seurat.png);"><nav id="nav"><span id="blog-info"><a class="nav-site-title" href="/"><img class="site-icon" src="/image/test.gif" alt="Logo"><span class="site-name">Nahida的净善宫</span></a><a class="nav-page-title" href="/"><span class="site-name">生信小记：Seurat_v5中的整合分析</span></a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> 时间轴</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> 标签</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> 分类</span></a></div><div class="menus_item"><a class="site-page" href="/photos/"><i class="fa-fw fas fa-images"></i><span> 相册</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> 友链</span></a></div></div><div id="toggle-menu"><span class="site-page"><i class="fas fa-bars fa-fw"></i></span></div></div></nav><div id="post-info"><h1 class="post-title">生信小记：Seurat_v5中的整合分析</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">发表于</span><time class="post-meta-date-created" datetime="2025-07-09T04:46:16.000Z" title="发表于 2025-07-09 12:46:16">2025-07-09</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">更新于</span><time class="post-meta-date-updated" datetime="2025-08-20T15:25:18.567Z" title="更新于 2025-08-20 23:25:18">2025-08-20</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/">生信分析</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title=""><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">浏览量:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="container post-content" id="article-container"><h1 id="前言"><a href="#前言" class="headerlink" title="前言"></a>前言</h1><p>单细胞测序中，实验和数据的收集往往不是一次性完成的，而是分为多个批次进行的。在不同批次之间，操作者、操作流程等诸多因素存在差异，可能给实验引入一些与真实生物学差异无关的系统性偏差，称为 <strong>批次效应</strong> 。</p>
<p>单细胞测序数据的整合通常是scRNA-seq测序的重要步骤之一。通过匹配不同数据集之间的共享细胞类型和状态，我们可以消除批次效应所引入的偏差，从而提高了统计功效。我们常使用的整合数据的方法有多种，包括Seurat早期版本发布的基于锚点的整合流程，其他团队发布的整合算法（如Harmony、scVI等）。Seurat v5版本的突破在于为这些不同的算法提供了一个通用的调用接口；只需要一行代码即可用不同的算法整合数据，为数据分析提供了极大的便利。</p>
<p>这里先加载需要用到的一系列R包。需要注意的是：</p>
<ol>
<li><code>SeuratWrappers</code>和<code>Azimuth</code>这两个包的过程中遇到了下载失败的问题，这是因为这两个包并没有在CRAN上发布，因此无法通过<code>install.packages()</code>语句直接下载。你可以尝试在安装remotes包后，用<code>remotes::install.github()</code>语句下载。如果仍然下载失败，可以Win+R后输入<code>inetcpl.cpl</code>后确认，打开Internet属性界面；然后在“高级”选项下找到并勾选“使用TLS 1.0”、“使用TLS 1.1”、“使用TLS 1.2”。最后重新运行安装语句即可下载。</li>
<li>建议检查自己所使用的Seurat包的版本。如果为Seurat v5.3.0版本，建议回退5.2及之前的版本（<code>install.packages(&quot;Seurat&quot;, version = &quot;5.2.0&quot;)</code>），或安装Github发布的最新开发版本（<code>devtools::install_github(&quot;satijalab/seurat&quot;)</code>）。原因见下述。</li>
</ol>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">library(Seurat)</span><br><span class="line">library(SeuratData)</span><br><span class="line">library(SeuratWrappers)</span><br><span class="line">library(Azimuth)</span><br><span class="line">library(ggplot2)</span><br><span class="line">library(patchwork)</span><br><span class="line">options(future.globals.maxSize = 1e9)</span><br></pre></td></tr></table></figure>

<h1 id="Seurat-v5对象中的layer"><a href="#Seurat-v5对象中的layer" class="headerlink" title="Seurat v5对象中的layer"></a>Seurat v5对象中的layer</h1><p>首先我们先介绍assay和layer的概念。Assay是存储同种分子类型数据的容器，多模态数据可通过多个独立的assay进行存储。layer是存储在同一assay内部的数据处理状态版本。​如果将Seurat对象比喻为一座图书馆，每一个assay即为这个图书馆内的一类藏书；而assay中的每一个layer则代表这类藏书的每一个版本。除此之外，这个图书馆还有一套“图书索引系统”，也就是meta.data，用来记录细胞的批次、干预信息。</p>
<p>Seurat早期版本中，每一个assay中能且仅能存储一个layer。这个限制会给我们带来了麻烦：我们在对数据处理的时候，处理后的数据不得不覆盖原先的数据，删除了原数据而导致难以回溯。而Seurat v5的一个突破点在于解除了这一限制，一个assay可以储存多个layer，可以储存多种处理状态下的实验数据（原始数据、未标准化的数据、标准化后的数据、Z-score标准化后的数据等）。</p>
<p><code>Azimuth</code>包是一款自动化单细胞注释工具，通过预训练的参考模型将细胞精准映射到已知细胞类型，无需手动标注。在这里，我们加载PBMC数据集（<code>pbmcsca</code>）并质量控制后，使用<code>Azimuth</code>包，以<code>pbmcsca</code>（经过预先训练的参考数据集）为对照，获取细胞类型的预测结果。SeuratData提供的参考数据集似乎需要连接位于纽约的服务器下载，因此下载超级超级慢，在RStudio中容易超时报错。故在下载前先修改了timeout时间再进行下载。或者在浏览器中打开下载网址下载<a target="_blank" rel="noopener" href="http://seurat.nygenome.org/src/contrib/pbmcsca.SeuratData_3.0.0.tar.gz">pbmcsca</a>和<a target="_blank" rel="noopener" href="http://seurat.nygenome.org/src/contrib/pbmcref.SeuratData_1.0.0.tar.gz">pbmcref</a>。尽管如此，下载速度慢的问题仍然难以解决，建议预留充足时间。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">options<span class="punctuation">(</span>timeout <span class="operator">=</span> <span class="number">600000000</span><span class="punctuation">)</span></span><br><span class="line">InstallData<span class="punctuation">(</span><span class="string">&quot;pbmcsca&quot;</span><span class="punctuation">)</span></span><br><span class="line">obj <span class="operator">&lt;-</span> LoadData<span class="punctuation">(</span><span class="string">&quot;pbmcsca&quot;</span><span class="punctuation">)</span></span><br><span class="line">obj <span class="operator">&lt;-</span> subset<span class="punctuation">(</span>obj<span class="punctuation">,</span> nFeature_RNA <span class="operator">&gt;</span> <span class="number">1000</span><span class="punctuation">)</span></span><br><span class="line">obj <span class="operator">&lt;-</span> RunAzimuth<span class="punctuation">(</span>obj<span class="punctuation">,</span> reference <span class="operator">=</span> <span class="string">&quot;pbmcref&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>另外，如果你遇到了以下报错：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">Error in ValidateParams_FindTransferAnchors(reference = reference, query = query, : </span><br><span class="line">Reference assay is SCT, but query assay is RNA. Mixing SCT and non-SCT in FindTransferAnchors is not supported.</span><br></pre></td></tr></table></figure>

<p>有可能是因为Seurat v5.3.0版本中FindTransferAnchors函数对分析数据类型(SCT、RNA)的严格检查机制。Azimuth包提供的参照数据集<code>pbmcsca</code>仅包含SCTransform标准化后的版本；因此在运行时，如果所处理的数据仅包含未标准化的数据时，会报告SCT缺失；如果对数据进行SCTransform转换，又会报告RNA缺失。网络上推荐的方法为更换Seurat包的版本，可以回退5.2及之前的版本（<code>install.packages(&quot;Seurat&quot;, version = &quot;5.2.0&quot;)</code>），也可以安装Github发布的最新开发版本（<code>devtools::install_github(&quot;satijalab/seurat&quot;)</code>）。我选择安装Github上的最新版本。</p>
<ul>
<li>输出结果：</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">&gt; obj</span><br><span class="line">An object of class Seurat </span><br><span class="line">33789 features across 10434 samples within 4 assays </span><br><span class="line">Active assay: RNA (33694 features, 0 variable features)</span><br><span class="line"> 2 layers present: counts, data</span><br><span class="line"> 3 other assays present: prediction.score.celltype.l1, prediction.score.celltype.l2, prediction.score.celltype.l3</span><br><span class="line"> 2 dimensional reductions calculated: integrated_dr, ref.umap</span><br></pre></td></tr></table></figure>

<p>这个数据集中一共有9个批次的数据，通过7个不同的技术平台得到。在Seurat早期版本中，我们需要将这些数据保存在9个不同的Seurat对象中，非常不方便。但在Seurat v5中，我们可以将这些数据全部塞进一个Seurat对象中，只需要拆分层次进行区分即可。每批数据分为<code>counts</code>层和<code>data</code>层，因此在拆分之后一共产生了18个layer。我们可以在不整合的情况下对数据进行分析，但需注意：此时各批次标准化和高变基因筛选是独立计算的，最终再形成跨批次的统一输出。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">obj[[&quot;RNA&quot;]] &lt;- split(obj[[&quot;RNA&quot;]], f = obj$Method)</span><br><span class="line">obj</span><br></pre></td></tr></table></figure>

<ul>
<li>运行结果：</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">&gt; obj[[&quot;RNA&quot;]] &lt;- split(obj[[&quot;RNA&quot;]], f = obj$Method)</span><br><span class="line">&gt; obj</span><br><span class="line">An object of class Seurat </span><br><span class="line">33789 features across 10434 samples within 4 assays </span><br><span class="line">Active assay: RNA (33694 features, 0 variable features)</span><br><span class="line"> 18 layers present: counts.Smart-seq2, counts.CEL-Seq2, counts.10x_Chromium_v2_A, counts.10x_Chromium_v2_B, counts.10x_Chromium_v3, counts.Drop-seq, counts.Seq-Well, counts.inDrops, counts.10x_Chromium_v2, data.Smart-seq2, data.CEL-Seq2, data.10x_Chromium_v2_A, data.10x_Chromium_v2_B, data.10x_Chromium_v3, data.Drop-seq, data.Seq-Well, data.inDrops, data.10x_Chromium_v2</span><br><span class="line"> 3 other assays present: prediction.score.celltype.l1, prediction.score.celltype.l2, prediction.score.celltype.l3</span><br><span class="line"> 2 dimensional reductions calculated: integrated_dr, ref.umap</span><br></pre></td></tr></table></figure>

<p>现在我们来对未整合的数据进行可视化。我们的细胞数据主要根据细胞类型和实验批次分类，然而在UMAP降维后的结果中，细胞簇主要按照其所使用的技术平台聚集，而不是按照真正的生物学差异聚集。这是由于不同的技术平台之间存在技术性偏差。如果此时直接聚类，我们将会得到以不同技术平台归类形成的假性细胞簇，而非我们所需要的按照真实生物学差异分类的真细胞簇。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">obj &lt;- FindNeighbors(obj, dims = 1:30, reduction = &quot;pca&quot;)</span><br><span class="line">obj &lt;- FindClusters(obj, resolution = 2, cluster.name = &quot;unintegrated_clusters&quot;)</span><br><span class="line">obj &lt;- RunUMAP(obj, dims = 1:30, reduction = &quot;pca&quot;, reduction.name = &quot;umap.unintegrated&quot;)</span><br><span class="line"># 根据细胞分类的结果是预置于Azimuth包内的数据</span><br><span class="line">DimPlot(obj, reduction = &quot;umap.unintegrated&quot;, group.by = c(&quot;Method&quot;, &quot;predicted.celltype.l2&quot;))</span><br></pre></td></tr></table></figure>

<p>运行结果：<br><img src="/images/integrative_analysis/803dc9f8-43c3-499b-be32-17c1bf24af3f.png"></p>
<p>通过上图不难发现，同一技术平台得出的细胞会聚集在一起，但同一细胞类型的细胞反而相距较远，呈碎片化分布。这两张图即批次效应的具象化显现。因此，我们需要通过整合分析来去除批次效应，才能得出有意义、站得住脚的结论。</p>
<h1 id="通过单行代码实现不同批次数据的整合分析"><a href="#通过单行代码实现不同批次数据的整合分析" class="headerlink" title="通过单行代码实现不同批次数据的整合分析"></a>通过单行代码实现不同批次数据的整合分析</h1><p>Seurat v5支持通过<code>IntegrateLayers()</code>函数，实现不同批次数据的整合分析。其支持的算法有以下五种：</p>
<ul>
<li>基于锚点的CCA整合（设置<code>method=CCAIntegration</code>）：适用于小型数据集（≤5万个细胞），可以精细地保留细胞亚群结构。但计算大数据集时计算速度慢、内存要求高。</li>
<li>基于锚点的RPCA整合（设置<code>method=RPCAIntegration</code>）：适用于存在较强技术噪声的小型数据集（≤10万个细胞），具有抗干扰能力强的特点。</li>
<li>Harmony（设置<code>method=HarmonyIntegration</code>）：兼顾了精度和速度，适用于中型和大型数据集（5—100万个细胞），可满足大多数需求。</li>
<li>FastMNN（设置<code>method=FastMNNIntegration</code>）：适用于超大规模数据集（＞50万个细胞），但处理小型数据集时可能过度平滑。</li>
<li>scVI（设置<code>method=scVIIntegration</code>）：适用于超大规模的数据集，尤其适用于处理复杂批次和多模态数据等复杂数据集。</li>
</ul>
<p>以下是使用不同算法进行整合的代码示例：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line">obj &lt;- IntegrateLayers(</span><br><span class="line">  object = obj, method = CCAIntegration,</span><br><span class="line">  orig.reduction = &quot;pca&quot;, new.reduction = &quot;integrated.cca&quot;,</span><br><span class="line">  verbose = FALSE</span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">obj &lt;- IntegrateLayers(</span><br><span class="line">  object = obj, method = RPCAIntegration,</span><br><span class="line">  orig.reduction = &quot;pca&quot;, new.reduction = &quot;integrated.rpca&quot;,</span><br><span class="line">  verbose = FALSE</span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">obj &lt;- IntegrateLayers(</span><br><span class="line">  object = obj, method = HarmonyIntegration,</span><br><span class="line">  orig.reduction = &quot;pca&quot;, new.reduction = &quot;harmony&quot;,</span><br><span class="line">  verbose = FALSE</span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">obj &lt;- IntegrateLayers(</span><br><span class="line">  object = obj, method = FastMNNIntegration,</span><br><span class="line">  new.reduction = &quot;integrated.mnn&quot;,</span><br><span class="line">  verbose = FALSE</span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">obj &lt;- IntegrateLayers(</span><br><span class="line">  object = obj, method = scVIIntegration,</span><br><span class="line">  new.reduction = &quot;integrated.scvi&quot;,</span><br><span class="line">  conda_env = &quot;../miniconda3/envs/scvi-env&quot;, verbose = FALSE</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p>随后，我们可以进一步进行可视化和聚类，这里以CCA和scVI为例：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">obj &lt;- FindNeighbors(obj, reduction = &quot;integrated.cca&quot;, dims = 1:30)</span><br><span class="line">obj &lt;- FindClusters(obj, resolution = 2, cluster.name = &quot;cca_clusters&quot;)</span><br><span class="line"></span><br><span class="line">obj &lt;- RunUMAP(obj, reduction = &quot;integrated.cca&quot;, dims = 1:30, reduction.name = &quot;umap.cca&quot;)</span><br><span class="line">p1 &lt;- DimPlot(</span><br><span class="line">  obj,</span><br><span class="line">  reduction = &quot;umap.cca&quot;,</span><br><span class="line">  group.by = c(&quot;Method&quot;, &quot;predicted.celltype.l2&quot;, &quot;cca_clusters&quot;),</span><br><span class="line">  combine = FALSE, label.size = 2</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">obj &lt;- FindNeighbors(obj, reduction = &quot;integrated.scvi&quot;, dims = 1:30)</span><br><span class="line">obj &lt;- FindClusters(obj, resolution = 2, cluster.name = &quot;scvi_clusters&quot;)</span><br><span class="line"></span><br><span class="line">obj &lt;- RunUMAP(obj, reduction = &quot;integrated.scvi&quot;, dims = 1:30, reduction.name = &quot;umap.scvi&quot;)</span><br><span class="line">p2 &lt;- DimPlot(</span><br><span class="line">  obj,</span><br><span class="line">  reduction = &quot;umap.scvi&quot;,</span><br><span class="line">  group.by = c(&quot;Method&quot;, &quot;predicted.celltype.l2&quot;, &quot;scvi_clusters&quot;),</span><br><span class="line">  combine = FALSE, label.size = 2</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">wrap_plots(c(p1, p2), ncol = 2, byrow = F)</span><br></pre></td></tr></table></figure>




</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta"><i class="fas fa-circle-user fa-fw"></i>文章作者: </span><span class="post-copyright-info"><a href="http://example.com">Nahida</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta"><i class="fas fa-square-arrow-up-right fa-fw"></i>文章链接: </span><span class="post-copyright-info"><a href="http://example.com/2025/07/09/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9ASeurat-v5%E4%B8%AD%E7%9A%84%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90/">http://example.com/2025/07/09/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9ASeurat-v5%E4%B8%AD%E7%9A%84%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta"><i class="fas fa-circle-exclamation fa-fw"></i>版权声明: </span><span class="post-copyright-info">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0</a> 许可协议。转载请注明来源 <a href="http://example.com" target="_blank">Nahida的净善宫</a>！</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/">生信分析</a><a class="post-meta__tags" href="/tags/R%E8%AF%AD%E8%A8%80/">R语言</a><a class="post-meta__tags" href="/tags/Seurat/">Seurat</a></div><div class="post-share"><div class="social-share" data-image="/images/Seurat.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><a class="pagination-related" href="/2025/07/09/%E3%80%90%E4%B9%90%E7%90%86%E3%80%91%E8%B0%83%E5%BC%8F/" title="【乐理】调式"><img class="cover" src="/images/%E5%9F%BA%E7%A1%80%E4%B9%90%E7%90%86.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="info"><div class="info-1"><div class="info-item-1">上一篇</div><div class="info-item-2">【乐理】调式</div></div><div class="info-2"><div class="info-item-1">是根据@三分钟音乐社的基础乐理课程所做的笔记！ 由若干个高低不同的乐音，围绕某一有稳定感的中心音（即主音），按照一定的音程关系组织在一起，成为一个有机的体系，称为 调式 。 大调式 自然大调：七个音按照“全全半全全全半”的规则组织在一起所形成的调式即为 自然大调 。 和声大调：在自然大调的基础上将Ⅵ级音降低一个半音，所形成的调式为 和声大调 ，即“全全半全半增二半”。因其与自然大调的相似性，和声大调通常在自然大调中偶尔出现，使乐曲富于变化和新鲜感。 旋律大调：在自然大调的基础上，将下行的Ⅵ级和Ⅶ级降低半音而形成的调式为 旋律大调 。其通常在自然大调中偶尔出现，使乐曲富于变化和新鲜感。   大调的判断方式:①死记$C$大调和$F$大调。②对于带♯的大调，如果最后一个升号所在位置为$♯X$音，则此调为$𝄪X$大调。在下图中，自左向右为$G,D,A,E,B,F♯,C♯$大调③对于带♭的大调，如果倒数第二个降号所在位置为$♭X$音，则此调为$♭X$大调。  小调式 自然小调：七个音按照“全半全全半全全”的规则组织在一起所形成的调式即为 自然小调...</div></div></div></a><a class="pagination-related" href="/2025/07/10/%E3%80%90%E4%B9%90%E7%90%86%E3%80%91%E4%B8%80%E4%BA%9B%E8%A1%A5%E5%85%85%E5%86%85%E5%AE%B9/" title="【乐理】一些补充内容"><img class="cover" src="/images/%E5%9F%BA%E7%A1%80%E4%B9%90%E7%90%86.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="info text-right"><div class="info-1"><div class="info-item-1">下一篇</div><div class="info-item-2">【乐理】一些补充内容</div></div><div class="info-2"><div class="info-item-1">是根据@三分钟音乐社的基础乐理课程所做的笔记！ 律制律制 是指为了使音乐规范化，人们有意选择的一组不同音高的音符所组成的体系，以及这些音符之间的相互关系。律制的核心在于研究不同音音高的大小。常见的律制主要有十二平均律、五度相生律和纯律，除此之外还有中庸全音律、威克麦斯特律、基恩伯格律、尤格律等。 音分距离指的是两个音高之间的差异程度。 通常可通过将两个音高的频率比值转换为以2为底的对数，然后取绝对值得到，即$音分距离&#x3D;\frac{1200}{lg 2}×lg(\frac{高音频率}{低音频率})$。 十二平均律将一个纯八度分为十二等份的半音，使得每两个相邻的半音之间的频率之比相等，此即 十二平均律 。但由于人工生硬均分，导致半音之间的频率比过于复杂（$\sqrt[12]{2}$）。不符合人体听觉美感，因此听感上并不如下述五度相生律和纯律谐和。但由于十二平均律存在等音，因此其在转调上具有一定的优势。 五度相生律五度相生律是根据复合音的纯五度关系（频率比为$2:3$），即由某一音开始推一纯五度，产生次一律，再由次一律推一纯五度，产生再次一律，如此继续相生所定出的音律叫做...</div></div></div></a></nav><div class="relatedPosts"><div class="headline"><i class="fas fa-thumbs-up fa-fw"></i><span>相关推荐</span></div><div class="relatedPosts-list"><a class="pagination-related" href="/2025/08/20/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9Ascvi-tools%E7%9A%84%E5%AE%89%E8%A3%85/" title="生信小记：scvi_tools的安装"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-08-20</div><div class="info-item-2">生信小记：scvi_tools的安装</div></div><div class="info-2"><div class="info-item-1">使用scvi算法进行单细胞测序数据的整合，一方面需要reticulate包，另一方面则需要安装在conda环境中的scvi-tools包及其依赖项。前者可以通过install.packages()语句从CRAN中下载，而本文章将介绍如何从零开始安装scvi-tools。 参考教程如下：  Anaconda安装教程（2025年保姆级超详解）【附安装包+环境玩转指南】   安装Anaconda打开Anaconda官网，点击右上角的“Free Download”，注册并登录后即可下载。 下载完成后打开安装包，可见如下界面。点击Next &gt;。 点击I Agree。 点击All users (requires admin privileges)。 选择存储位置时，强烈不建议存储在C盘，因为它安装完占用的空间不小。更改完存储路径后点击Next &gt;。 全勾选上再点Next...</div></div></div></a><a class="pagination-related" href="/2025/04/29/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%84%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90%E4%BB%8B%E7%BB%8D/" title="生信小记：单细胞转录组数据整合分析介绍"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-04-29</div><div class="info-item-2">生信小记：单细胞转录组数据整合分析介绍</div></div><div class="info-2"><div class="info-item-1">...</div></div></div></a><a class="pagination-related" href="/2025/03/31/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E4%BD%BF%E7%94%A8Seurat%E5%8C%85%E5%A4%84%E7%90%86%E5%A4%9A%E6%A8%A1%E6%80%81%E6%95%B0%E6%8D%AE/" title="生信小记：使用Seurat包处理多模态数据"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-03-31</div><div class="info-item-2">生信小记：使用Seurat包处理多模态数据</div></div><div class="info-2"><div class="info-item-1">在一次实验中，我们可以对同一个细胞测量不同分子层面的数据（如基因表达、蛋白质表达、表观遗传修饰、空间分布等等），所测得的数据我们称之为多模态数据（multimodal data）。例如，CITE-seq可以同时检测单细胞的转录组RNA和表面蛋白表达水平；10X Multiome可以同时测量单细胞的转录组RNA和ATAC（染色质可及性）。本篇文章将以一个包括8617个脐带血单个核细胞（CBMC）的CITE-seq结果（包含转录组RNA测序和表面蛋白丰度估计值）的数据集，介绍如何使用Seurat包对多模态数据进行分析。 导入数据，创建Seurat对象，并加入RNA和表面蛋白数据官方教程中介绍的方法如下： 12345678910111213141516171819202122232425262728library(Seurat)library(ggplot2)library(patchwork)# 注意：这份数据同时存在人和小鼠的细胞，分别冠有&quot;HUMAN_&quot;和&quot;MOUSE_&quot;的前缀。# 导入RNA数据cbmc.rna &lt;-...</div></div></div></a><a class="pagination-related" href="/2025/03/31/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9A%E4%BD%BF%E7%94%A8SCTransform%E8%BF%9B%E8%A1%8C%E6%A0%87%E5%87%86%E5%8C%96/" title="生信小记：使用SCTransform进行标准化"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-03-31</div><div class="info-item-2">生信小记：使用SCTransform进行标准化</div></div><div class="info-2"><div class="info-item-1">...</div></div></div></a><a class="pagination-related" href="/2025/03/29/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9ASeurat%E5%8C%85%E7%9A%84%E6%95%B0%E6%8D%AE%E5%8F%AF%E8%A7%86%E5%8C%96%E6%96%B9%E6%B3%95/" title="生信小记：Seurat包的数据可视化方法"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-03-29</div><div class="info-item-2">生信小记：Seurat包的数据可视化方法</div></div><div class="info-2"><div class="info-item-1">在这篇教程中，我们会沿用上一篇教程处理好的数据来进行演示。 先做一些前期准备，安装并加载好R包，下载并加载好数据（或者直接运用上一篇教程中处理好的数据）。 12345678910111213141516171819202122install.packages(c(&quot;Seurat&quot;,&quot;patchwork&quot;,&quot;ggplot2&quot;,&quot;SeuratData&quot;))#下载SeuratData包时报错显示：package ‘SeuratData’ is not available for this version of...</div></div></div></a><a class="pagination-related" href="/2025/03/28/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9ASeurat%E7%9A%84%E5%B8%B8%E8%A7%81%E5%88%86%E6%9E%90%E5%B7%A5%E4%BD%9C%E6%B5%81%E7%A8%8B%EF%BC%9A%E4%BB%A5PBMC-3K%E6%95%B0%E6%8D%AE%E9%9B%86%E4%B8%BA%E4%BE%8B/" title="生信小记：Seurat的常见分析工作流程：以PBMC_3K数据集为例"><img class="cover" src="/images/Seurat.png" alt="cover"><div class="info text-center"><div class="info-1"><div class="info-item-1"><i class="far fa-calendar-alt fa-fw"></i> 2025-03-28</div><div class="info-item-2">生信小记：Seurat的常见分析工作流程：以PBMC_3K数据集为例</div></div><div class="info-2"><div class="info-item-1">更新日志2025&#x2F;4&#x2F;26 更正了“确认数据集维度”中，对肘部图意义解释的错误。 修改了“执行非线性降维”引语中，对于PCA降维不足之处的表述。 在“创建Seurat对象”中，新增对于读入文件的解释。 修正了一些排版错误。   在网上找了好几篇Seurat包的教程发现对新手都不太友好……遂自己写了一篇…… 创建Seurat对象我们将对10X Genomics免费提供的PBMC数据集进行分析，包括2700个由Illumina NextSeq测序平台测序的单细胞。我们的原始数据包括3个文件：barcodes.tsv、genes.tsv、matrix.mtx。分别用于存储每个单细胞的细胞条形码、基因的标识符和名称、存储基因表达数据。 数据的读取Read10X()函数可以从10X cellranger pipeline的输出中读取数据，并形成一个唯一的分子识别（UMI）计数矩阵。矩阵中的每个值代表所测序的每个细胞中的每个特征（如基因）。对于h5文件格式的文件，可以用Read10X_h5()读取。   10X Cell Ranger Pipeline 是 10X...</div></div></div></a></div></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info text-center"><div class="avatar-img"><img src="/img/avatar.jpg" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info-name">Nahida</div><div class="author-info-description">智慧树上智慧果，智慧树下你和我</div><div class="site-data"><a href="/archives/"><div class="headline">文章</div><div class="length-num">90</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">51</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">21</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/xxxxxx"><i class="fab fa-github"></i><span>Follow Me</span></a></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>公告</span></div><div class="announcement_content">暂无公告哦~</div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>目录</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%89%8D%E8%A8%80"><span class="toc-number">1.</span> <span class="toc-text">前言</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#Seurat-v5%E5%AF%B9%E8%B1%A1%E4%B8%AD%E7%9A%84layer"><span class="toc-number">2.</span> <span class="toc-text">Seurat v5对象中的layer</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E9%80%9A%E8%BF%87%E5%8D%95%E8%A1%8C%E4%BB%A3%E7%A0%81%E5%AE%9E%E7%8E%B0%E4%B8%8D%E5%90%8C%E6%89%B9%E6%AC%A1%E6%95%B0%E6%8D%AE%E7%9A%84%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90"><span class="toc-number">3.</span> <span class="toc-text">通过单行代码实现不同批次数据的整合分析</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>最新文章</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2025/08/20/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9Ascvi-tools%E7%9A%84%E5%AE%89%E8%A3%85/" title="生信小记：scvi_tools的安装"><img src="/images/Seurat.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="生信小记：scvi_tools的安装"/></a><div class="content"><a class="title" href="/2025/08/20/%E7%94%9F%E4%BF%A1%E5%B0%8F%E8%AE%B0%EF%BC%9Ascvi-tools%E7%9A%84%E5%AE%89%E8%A3%85/" title="生信小记：scvi_tools的安装">生信小记：scvi_tools的安装</a><time datetime="2025-08-20T10:34:51.000Z" title="发表于 2025-08-20 18:34:51">2025-08-20</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2025/08/13/%E3%80%90%E5%A6%87%E4%BA%A7%E5%AD%A6%E3%80%91%E5%A5%B3%E6%80%A7%E7%94%9F%E6%AE%96%E7%B3%BB%E7%BB%9F%E7%94%9F%E7%90%86/" title="【妇产学】女性生殖系统生理">【妇产学】女性生殖系统生理</a><time datetime="2025-08-12T19:35:33.000Z" title="发表于 2025-08-13 03:35:33">2025-08-13</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2025/08/12/%E3%80%90%E5%A6%87%E4%BA%A7%E5%AD%A6%E3%80%91%E5%A5%B3%E6%80%A7%E7%94%9F%E6%AE%96%E7%B3%BB%E7%BB%9F%E8%A7%A3%E5%89%96/" title="【妇产学】女性生殖系统解剖">【妇产学】女性生殖系统解剖</a><time datetime="2025-08-12T11:32:47.000Z" title="发表于 2025-08-12 19:32:47">2025-08-12</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2025/07/27/%E3%80%90%E6%96%AF%E6%B3%A2%E7%B4%A2%E5%AE%BE%E5%92%8C%E5%A3%B0%E5%AD%A6%E3%80%91%E4%B9%90%E5%8F%A5%E3%80%81%E4%B9%90%E6%AE%B5%E4%B8%8E%E7%BB%88%E6%AD%A2%E5%BC%8F/" title="【斯波索宾和声学】乐句、乐段与终止式"><img src="/images/%E5%92%8C%E5%A3%B0.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="【斯波索宾和声学】乐句、乐段与终止式"/></a><div class="content"><a class="title" href="/2025/07/27/%E3%80%90%E6%96%AF%E6%B3%A2%E7%B4%A2%E5%AE%BE%E5%92%8C%E5%A3%B0%E5%AD%A6%E3%80%91%E4%B9%90%E5%8F%A5%E3%80%81%E4%B9%90%E6%AE%B5%E4%B8%8E%E7%BB%88%E6%AD%A2%E5%BC%8F/" title="【斯波索宾和声学】乐句、乐段与终止式">【斯波索宾和声学】乐句、乐段与终止式</a><time datetime="2025-07-27T07:19:46.000Z" title="发表于 2025-07-27 15:19:46">2025-07-27</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2025/07/22/%E3%80%90%E6%96%AF%E6%B3%A2%E7%B4%A2%E5%AE%BE%E5%92%8C%E5%A3%B0%E5%AD%A6%E3%80%91%E6%AD%A3%E4%B8%89%E5%92%8C%E5%BC%A6%E7%9A%84%E5%85%AD%E5%92%8C%E5%BC%A6/" title="【斯波索宾和声学】正三和弦的六和弦"><img src="/images/%E5%92%8C%E5%A3%B0.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="【斯波索宾和声学】正三和弦的六和弦"/></a><div class="content"><a class="title" href="/2025/07/22/%E3%80%90%E6%96%AF%E6%B3%A2%E7%B4%A2%E5%AE%BE%E5%92%8C%E5%A3%B0%E5%AD%A6%E3%80%91%E6%AD%A3%E4%B8%89%E5%92%8C%E5%BC%A6%E7%9A%84%E5%85%AD%E5%92%8C%E5%BC%A6/" title="【斯波索宾和声学】正三和弦的六和弦">【斯波索宾和声学】正三和弦的六和弦</a><time datetime="2025-07-22T03:53:38.000Z" title="发表于 2025-07-22 11:53:38">2025-07-22</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2019 - 2025 By Nahida</div><div class="framework-info"><span>框架 </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo 7.3.0</a><span class="footer-separator">|</span><span>主题 </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly 5.3.5</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="阅读模式"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="日间和夜间模式切换"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="单栏和双栏切换"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside-config" type="button" title="设置"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="目录"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="回到顶部"><span class="scroll-percent"></span><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><div class="js-pjax"><script>(() => {
  const loadMathjax = () => {
    if (!window.MathJax) {
      window.MathJax = {
        tex: {
          inlineMath: [['$', '$'], ['\\(', '\\)']],
          tags: 'none',
        },
        chtml: {
          scale: 1.1
        },
        options: {
          enableMenu: true,
          renderActions: {
            findScript: [10, doc => {
              for (const node of document.querySelectorAll('script[type^="math/tex"]')) {
                const display = !!node.type.match(/; *mode=display/)
                const math = new doc.options.MathItem(node.textContent, doc.inputJax[0], display)
                const text = document.createTextNode('')
                node.parentNode.replaceChild(text, node)
                math.start = {node: text, delim: '', n: 0}
                math.end = {node: text, delim: '', n: 0}
                doc.math.push(math)
              }
            }, '']
          }
        }
      }

      const script = document.createElement('script')
      script.src = 'https://cdn.jsdelivr.net/npm/mathjax/es5/tex-mml-chtml.min.js'
      script.id = 'MathJax-script'
      script.async = true
      document.head.appendChild(script)
    } else {
      MathJax.startup.document.state(0)
      MathJax.texReset()
      MathJax.typesetPromise()
    }
  }

  btf.addGlobalFn('encrypt', loadMathjax, 'mathjax')
  window.pjax ? loadMathjax() : window.addEventListener('load', loadMathjax)
})()</script></div><script type="text/javascript" src="/js/memos/waterfall.min.js"></script><script type="text/javascript" src="/js/memos/imgStatus.min.js"></script><script type="text/javascript" src="/js/memos/lately.min.js"></script><script type="text/javascript" src="/js/memos/photo.js"></script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>